{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# XPRESSpipe methodology analysis\n",
    "\n",
    "Copyright (C) 2019  Jordan A. Berg   \n",
    "jordan.berg@biochem.utah.edu   \n",
    "   \n",
    "This program is free software: you can redistribute it and/or modify it under\n",
    "the terms of the GNU General Public License as published by the Free Software\n",
    "Foundation, either version 3 of the License, or (at your option) any later\n",
    "version.    \n",
    "    \n",
    "This program is distributed in the hope that it will be useful, but WITHOUT ANY\n",
    "WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A\n",
    "PARTICULAR PURPOSE. See the GNU General Public License for more details.    \n",
    "    \n",
    "You should have received a copy of the GNU General Public License along with\n",
    "this program.  If not, see <https://www.gnu.org/licenses/>."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import print_function\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as patches\n",
    "matplotlib.rcParams['font.sans-serif'] = 'Arial'\n",
    "from matplotlib_venn import venn2, venn3\n",
    "import seaborn as sns\n",
    "from xpressplot import convert_names, rpm, check_samples, threshold\n",
    "from scipy import stats\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "%matplotlib inline\n",
    "import random\n",
    "import ast\n",
    "from urllib.request import Request, urlopen\n",
    "import plotly\n",
    "import plotly.offline as py\n",
    "import plotly_express as px"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set global file names and directories\n",
    "- Make sure the above dependencies have been installed on your computer\n",
    "- Change paths to files and directories as appropriate for use on your computer\n",
    "- Execute a code block by clicking on the code block and pressing `Shift + Enter`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# XPRESSpipe-processed data \n",
    "file = '/Users/jordan/Desktop/xpressyourself_manuscript/isrib_analysis/isrib_comp_test/isrib_comp_v98_truncated_count_table.tsv'\n",
    "\n",
    "# Location and filename of original count data \n",
    "original = '/Users/jordan/Desktop/xpressyourself_manuscript/isrib_analysis/ingolia_counts_table.txt'\n",
    "\n",
    "# Define location to save plots to\n",
    "plot_dir = '/Users/jordan/Desktop/xpressyourself_manuscript/isrib_analysis/plots/'\n",
    "\n",
    "# Directory where data for DESeq2 should be output\n",
    "deseq_directory = '/Users/jordan/Desktop/xpressyourself_manuscript/isrib_analysis/isrib_de/xpresspipe_data_deseq2/'\n",
    "\n",
    "# Define location to save DESeq2 plots to\n",
    "deseq_plot_directory = '/Users/jordan/Desktop/xpressyourself_manuscript/isrib_analysis/isrib_de/plots/'\n",
    "\n",
    "# Path and filename for original data reprocessed with DESeq2\n",
    "dir_og2 = '/Users/jordan/Desktop/xpressyourself_manuscript/isrib_analysis/isrib_de/original_data_deseq2/'\n",
    "\n",
    "# GTF reference file \n",
    "gtf_file = '/Users/jordan/Desktop/reference/Homo_sapiens.GRCh38.98.gtf'\n",
    "\n",
    "# SRA sample table info\n",
    "sra_file = '/Users/jordan/Desktop/xpressyourself_manuscript/isrib_analysis/GSE65778_table.txt'\n",
    "\n",
    "# Location of comparison count tables between methods and with multiple approaches\n",
    "comparison_tables_directory = '/Users/jordan/Desktop/xpressyourself_manuscript/isrib_analysis/isrib_comp_test/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set global gene lists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ingolia_tm = [\n",
    "    'FAM50B','CHST2','LNP1','AX747756','ACRC',\n",
    "    'ATF5','ANKRD23','ATF4','C19orf48','TEX14',\n",
    "    'SLC35A4','DDIT3','RAB35','AK293147','PTP4A1',\n",
    "    'ANKRD36C','NPIP','ANKMY1','UCP2','NPIPL1',\n",
    "    'PNRC2','C7orf31','PPP1R15A','IFRD1','BCL2L11',\n",
    "    'OBSCN','SLC24A1','NPIPL1','NPIPL1','ZNF22',\n",
    "    'NAIP','NPIP','DNAH1','KIAA1377','FMO4',\n",
    "    'CCDC150','NPIP','AK310228','ZSCAN5A','FAM13B',\n",
    "    'NPIPL1','ZNF79','HOXA6','NPIPL2','SLMO1',\n",
    "    'ARMCX4','AK310228','LOC100288332','SAT1','FLJ00285',\n",
    "    'AK097143','AX748249','AGAP4','FLJ00322','GOLGA8A',\n",
    "    'GOLGA8B','ZNF432','FSBP','CPS1','MATK',\n",
    "    'EPB41L4B','HCN2','GLRB','KIAA1841','MYO5B',\n",
    "    'EHBP1L1','MAPK8IP1','SH3RF3','DET1','PTPN21',\n",
    "    'VANGL2','MMP15','MAP3K10','PAQR5','GAL',\n",
    "    'RPP25','FLJ44955','DQ576756']\n",
    "\n",
    "ingolia_tmisrib = [\n",
    "    'AK310228','AK310228','NPIP','AK302511','NPIP',\n",
    "    'LOC100288332','STARD9','PKD1','ANKRD36']\n",
    "\n",
    "ingolia_isrib = [\n",
    "    'AK310228','AK310228','AK302511','NPIP']\n",
    "\n",
    "# Initialize lists of Ingolia highlights\n",
    "isr = [\n",
    "    'ATF4',\n",
    "    'ATF5',\n",
    "    'PPP1R15A',\n",
    "    'DDIT3']\n",
    "\n",
    "uorf_targets = [\n",
    "    'SLC35A4',\n",
    "    'PTP4A1',\n",
    "    'UCP2',\n",
    "    'C7orf31',\n",
    "    'BCL2L11',\n",
    "    'PNRC2',\n",
    "    'SAT1']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define functions needed for processing and plotting"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Define methods for correlation analysis within and between methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Collapse technical replicates and rename samples\n",
    "\"\"\"\n",
    "def name_map(\n",
    "        data,\n",
    "        map):\n",
    "\n",
    "    name_dict = pd.Series(map.ingolia_name.values,index=map.Run).to_dict()\n",
    "    data_c = data.copy()\n",
    "    data_c.columns = data_c.columns.to_series().map(name_dict)\n",
    "    data_sum = data_c.groupby(data_c.columns, axis=1).sum()\n",
    "\n",
    "    return data_sum\n",
    "\n",
    "\"\"\"Make supplemental figure 4\n",
    "\"\"\"\n",
    "def make_paralogs_analysis(\n",
    "    data,\n",
    "    original_data,\n",
    "    title):\n",
    "\n",
    "    # Init plotting space\n",
    "    fig, axes = plt.subplots(\n",
    "        nrows = 2,\n",
    "        ncols = 2,\n",
    "        figsize = (20, 20),\n",
    "        subplot_kw = {\n",
    "            'facecolor':'none'},\n",
    "        sharey=True,\n",
    "        sharex=True) # Create shared axis for cleanliness\n",
    "    plt.subplots_adjust(\n",
    "        bottom = 0.1)\n",
    "    plt.yticks([0,1,2,3,4,5]) # Limit axis labels to ints\n",
    "    plt.xticks([0,1,2,3,4,5])\n",
    "\n",
    "    x = 0\n",
    "    file_number = 0\n",
    "    file_list = [\n",
    "        'ribo_untr_a',\n",
    "        'ribo_tm_a',\n",
    "        'untr_a_hek',\n",
    "        'tm_a_hek'] # Designate sample order\n",
    "\n",
    "    for y in range(4):\n",
    "\n",
    "        # Get data as array-like for samples being compared\n",
    "        data_c1 = data.copy()\n",
    "        data_c2 = original_data.copy()\n",
    "\n",
    "        sample_a = data_c1[file_list[file_number]].values.tolist()\n",
    "        sample_a = [x + 1 for x in sample_a]\n",
    "        sample_a = np.array(sample_a).astype(np.float)\n",
    "        sample_a = np.ndarray.tolist(sample_a)\n",
    "\n",
    "        sample_b = data_c2[file_list[file_number]].values.tolist()\n",
    "        sample_b = [x + 1 for x in sample_b]\n",
    "        sample_b = np.array(sample_b).astype(np.float)\n",
    "        sample_b = np.ndarray.tolist(sample_b)\n",
    "\n",
    "        # Determine subplot location\n",
    "        if file_number in [0,1]:\n",
    "            ax_x = file_number % 2\n",
    "            ax_y = 0\n",
    "        elif file_number in [2,3]:\n",
    "            ax_x = file_number % 2\n",
    "            ax_y = 1\n",
    "        else:\n",
    "            print('oops')\n",
    "\n",
    "        # Plot data\n",
    "        axes[ax_y, ax_x].tick_params(axis='both', labelsize=32)\n",
    "        axes[ax_y, ax_x].scatter(np.log10(sample_a), np.log10(sample_b), s=5,c='grey')\n",
    "\n",
    "        genes = ['EIF3C','NOMO2','KCNQ2','RPL39','TUBA1A']\n",
    "        colors = ['red','blue','green','purple','orange']\n",
    "        for x in range(5):\n",
    "\n",
    "            s1 = (data_c1 + 1).at[genes[x], file_list[file_number]]\n",
    "            s2 = (data_c2 + 1).at[genes[x], file_list[file_number]]\n",
    "\n",
    "            axes[ax_y, ax_x].scatter(np.log10(s1), np.log10(s2), s=100,c=colors[x])\n",
    "\n",
    "        axes[ax_y, ax_x].axhline(0, ls='-', color='black', xmin=0.0457, xmax=1) # Create axis lines\n",
    "        axes[ax_y, ax_x].axvline(0, ls='-', color='black', ymin=0.0457, ymax=1)\n",
    "        file_number += 1 # Plot counter\n",
    "\n",
    "\n",
    "    # Create shared row/column titles\n",
    "    for ax in axes[:,0]:\n",
    "        ax.set_ylabel('log$_1$$_0$(counts)', fontsize=32)\n",
    "\n",
    "    cols = ['Untreated','Tm']\n",
    "    for ax, col in zip(axes[0], cols):\n",
    "        ax.set_xlabel(col, fontsize=40)\n",
    "        ax.xaxis.set_label_position('top')\n",
    "\n",
    "    rows = ['Ribo-seq','RNA-seq']\n",
    "    for ax, rows in zip(axes[:,1], rows):\n",
    "        ax.set_ylabel(rows, fontsize=40, labelpad=40, rotation=270)\n",
    "        ax.yaxis.set_label_position('right')\n",
    "\n",
    "    for ax in axes[1]:\n",
    "        ax.set_xlabel('log$_1$$_0$(counts)', fontsize=32)\n",
    "        ax.xaxis.set_label_position('bottom')\n",
    "\n",
    "    fig.savefig(\n",
    "        str(plot_dir) + str(title),\n",
    "        dpi = 600,\n",
    "        bbox_inches = 'tight')\n",
    "\n",
    "# Make figure 2B\n",
    "def make_external_correlations(\n",
    "        data,\n",
    "        original_data,\n",
    "        title):\n",
    "\n",
    "    fig, axes = plt.subplots(\n",
    "        nrows = 2,\n",
    "        ncols = 2,\n",
    "        figsize = (20, 20),\n",
    "        subplot_kw = {\n",
    "            'facecolor':'none'},\n",
    "        sharey=True) # Create shared axis for cleanliness\n",
    "    plt.subplots_adjust(\n",
    "        bottom = 0.1)\n",
    "    plt.yticks([0,1,2,3,4,5]) # Limit axis labels to ints\n",
    "    plt.xticks([0,1,2,3,4,5])\n",
    "    plt.axis('equal')\n",
    "\n",
    "    x = 0\n",
    "    file_number = 0\n",
    "    file_list = [\n",
    "        'ribo_tm_a',\n",
    "        'isrib_b_hek'] # Designate sample order\n",
    "\n",
    "    for y in range(2):\n",
    "\n",
    "        # Get data as array-like for samples being compared\n",
    "        data_c1 = data.copy()\n",
    "        data_c2 = original_data.copy()\n",
    "\n",
    "        sample_a = data_c1[file_list[file_number]].values.tolist()\n",
    "        sample_a = [x + 1 for x in sample_a]\n",
    "        sample_a = np.array(sample_a).astype(np.float)\n",
    "        sample_a = np.ndarray.tolist(sample_a)\n",
    "\n",
    "        sample_b = data_c2[file_list[file_number]].values.tolist()\n",
    "        sample_b = [x + 1 for x in sample_b]\n",
    "        sample_b = np.array(sample_b).astype(np.float)\n",
    "        sample_b = np.ndarray.tolist(sample_b)\n",
    "\n",
    "        # Run Spearman R linreg for non-normal data\n",
    "        rho, p_value = stats.spearmanr(sample_a, sample_b)\n",
    "\n",
    "        # Determine subplot location\n",
    "        if file_number in [0,1]:\n",
    "            ax_x = file_number % 2\n",
    "            ax_y = 0\n",
    "        else:\n",
    "            print('oops')\n",
    "\n",
    "        # Format p value\n",
    "        if p_value < 0.001:\n",
    "            p_val = '< 0.001'\n",
    "        else:\n",
    "            p_val = '= {:.3f}'.format(round(p_value, 3))\n",
    "\n",
    "        rho = '{:.3f}'.format(round(rho, 3))\n",
    "\n",
    "        # Plot data\n",
    "        axes[ax_y, ax_x].tick_params(axis='both', labelsize=32)\n",
    "        axes[ax_y, ax_x].scatter(np.log10(sample_a), np.log10(sample_b), s=1,c='black')\n",
    "        axes[ax_y, ax_x].set_title(r\"$\\rho$\" + ' = ' + str(rho) + '\\nP ' + p_val, y=0.1, x=0.9, fontsize=32) # Format titles\n",
    "        axes[ax_y, ax_x].axhline(0, ls='-', color='black', xmin=0.0457, xmax=1) # Create axis lines\n",
    "        axes[ax_y, ax_x].axvline(0, ls='-', color='black', ymin=0.0457, ymax=1)\n",
    "        file_number += 1 # Plot counter\n",
    "        x += 2\n",
    "        print(rho)\n",
    "\n",
    "    # Create shared row/column titles\n",
    "    for ax in axes[:,0]:\n",
    "        ax.set_ylabel('log$_1$$_0$(counts)', fontsize=32)\n",
    "\n",
    "    cols = ['RPF Tm RepA','mRNA ISRIB RepB']\n",
    "    for ax, col in zip(axes[0], cols):\n",
    "        ax.set_xlabel(col, fontsize=40)\n",
    "        ax.xaxis.set_label_position('top')\n",
    "    cols = ['log$_1$$_0$(counts)','log$_1$$_0$(counts)']\n",
    "    for ax, col in zip(axes[1], cols):\n",
    "        ax.set_xlabel(col, fontsize=32, labelpad=30)\n",
    "        ax.xaxis.set_label_position('top')\n",
    "\n",
    "    fig.savefig(\n",
    "        str(plot_dir) + str(title),\n",
    "        dpi = 600,\n",
    "        bbox_inches = 'tight')\n",
    "\n",
    "def make_external_correlations_supplement(\n",
    "        data,\n",
    "        original_data,\n",
    "        title,\n",
    "        interactive=False):\n",
    "\n",
    "    fig, axes = plt.subplots(\n",
    "        nrows = 4,\n",
    "        ncols = 4,\n",
    "        figsize = (20, 20),\n",
    "        subplot_kw = {\n",
    "            'facecolor':'none'},\n",
    "        sharex=True, sharey=True) # Create shared axis for cleanliness\n",
    "    plt.subplots_adjust(\n",
    "        bottom = 0.1)\n",
    "    plt.yticks([0,1,2,3,4,5]) # Limit axis labels to ints\n",
    "    plt.xticks([0,1,2,3,4,5])\n",
    "\n",
    "    file_number = 0\n",
    "    file_list = [\n",
    "        'ribo_untr_a',\n",
    "        'ribo_untr_b',\n",
    "        'untr_a_hek',\n",
    "        'untr_b_hek',\n",
    "        'ribo_tm_a',\n",
    "        'ribo_tm_b',\n",
    "        'tm_a_hek',\n",
    "        'tm_b_hek',\n",
    "        'ribo_tmisrib_a',\n",
    "        'ribo_tmisrib_b',\n",
    "        'tmisrib_a_hek',\n",
    "        'tmisrib_b_hek',\n",
    "        'ribo_isrib_a',\n",
    "        'ribo_isrib_b',\n",
    "        'isrib_a_hek',\n",
    "        'isrib_b_hek'] # Designate sample order\n",
    "\n",
    "    for x in file_list:\n",
    "\n",
    "        if interactive == True:\n",
    "            merged_best = pd.concat([data[[str(x)]], original_data[[str(x)]]], axis=1, sort=False)\n",
    "            merged_best.columns = ['xpresspipe', 'original']\n",
    "            merged_best['genes'] = merged_best.index\n",
    "            sc = px.scatter(\n",
    "                merged_best,\n",
    "                x='xpresspipe',\n",
    "                y='original',\n",
    "                hover_name='genes',\n",
    "                log_x=True,\n",
    "                log_y=True,\n",
    "                opacity=0.4,\n",
    "                width=1400,\n",
    "                height=1000,\n",
    "                title=str(x))\n",
    "\n",
    "            py.offline.plot(\n",
    "                sc, \n",
    "                filename=str(plot_dir) + 'interactive_' + str(x) + '.html')\n",
    "\n",
    "        # Get data as array-like for samples being compared\n",
    "        data_c1 = data.copy()\n",
    "        data_c2 = original_data.copy()\n",
    "\n",
    "        sample_a = data_c1[x].values.tolist()\n",
    "        sample_a = [x + 1 for x in sample_a]\n",
    "        sample_a = np.array(sample_a).astype(np.float)\n",
    "        sample_a = np.ndarray.tolist(sample_a)\n",
    "\n",
    "        sample_b = data_c2[x].values.tolist()\n",
    "        sample_b = [x + 1 for x in sample_b]\n",
    "        sample_b = np.array(sample_b).astype(np.float)\n",
    "        sample_b = np.ndarray.tolist(sample_b)\n",
    "\n",
    "        # Run Spearman R linreg for non-normal data\n",
    "        rho, p_value = stats.spearmanr(sample_a, sample_b)\n",
    "\n",
    "        # Determine subplot location\n",
    "        if file_number in [0,1,2,3]:\n",
    "            ax_x = file_number % 4\n",
    "            ax_y = 0\n",
    "        elif file_number in [4,5,6,7]:\n",
    "            ax_x = file_number - 4\n",
    "            ax_y = 1\n",
    "        elif file_number in [8,9,10,11]:\n",
    "            ax_x = file_number - 8\n",
    "            ax_y = 2\n",
    "        elif file_number in [12,13,14,15]:\n",
    "            ax_x = file_number - 12\n",
    "            ax_y = 3\n",
    "        else:\n",
    "            print('oops')\n",
    "\n",
    "        # Format p value\n",
    "        if p_value < 0.001:\n",
    "            p_val = '< 0.001'\n",
    "        else:\n",
    "            p_val = '= {:.3f}'.format(round(p_value, 3))\n",
    "\n",
    "        rho = '{:.3f}'.format(round(rho, 3))\n",
    "\n",
    "        # Plot data\n",
    "        axes[ax_y, ax_x].tick_params(axis='both', labelsize=32)\n",
    "        axes[ax_y, ax_x].scatter(np.log10(sample_a), np.log10(sample_b), s=1,c='black')\n",
    "        axes[ax_y, ax_x].set_title(r\"$\\rho$\" + ' = ' + str(rho) + '\\nP ' + p_val, y=0.1, x=0.8, fontsize=20) # Format titles\n",
    "        axes[ax_y, ax_x].axhline(0, ls='-', color='black', xmin=0.05, xmax=1) # Create axis lines\n",
    "        axes[ax_y, ax_x].axvline(0, ls='-', color='black', ymin=0.05, ymax=0.88)\n",
    "        file_number += 1 # Plot counter\n",
    "        print(rho)\n",
    "\n",
    "    # Create shared row/column titles\n",
    "    count_label = ['log$_1$$_0$(counts)','log$_1$$_0$(counts)','log$_1$$_0$(counts)','log$_1$$_0$(counts)']\n",
    "\n",
    "    cols = ['RPF RepA','RPF RepB','mRNA RepA','mRNA RepB']\n",
    "    for ax, col in zip(axes[0], cols):\n",
    "        ax.set_xlabel(col, fontsize=40)\n",
    "        ax.xaxis.set_label_position('top')\n",
    "\n",
    "    for ax in axes[3]:\n",
    "        ax.set_xlabel('log$_1$$_0$(counts)', fontsize=32)\n",
    "        ax.xaxis.set_label_position('bottom')\n",
    "\n",
    "    for ax in axes[:,0]:\n",
    "        ax.set_ylabel('log$_1$$_0$(counts)', fontsize=32)\n",
    "\n",
    "    rows = ['Untreated','Tm','Tm + ISRIB','ISRIB']\n",
    "    for ax, rows in zip(axes[:,3], rows):\n",
    "        ax.set_ylabel(rows, fontsize=40, labelpad=40, rotation=270)\n",
    "        ax.yaxis.set_label_position('right')\n",
    "\n",
    "    fig.savefig(\n",
    "        str(plot_dir) + str(title),\n",
    "        dpi = 600,\n",
    "        bbox_inches = 'tight')\n",
    "\n",
    "\n",
    "def make_spearman_pearson_correlations(\n",
    "        data,\n",
    "        original_data,\n",
    "        title):\n",
    "\n",
    "    data_t = np.log10(data + 1)\n",
    "    original_data_t = np.log10(original_data + 1)\n",
    "\n",
    "    file_list = [\n",
    "        'ribo_untr_a',\n",
    "        'ribo_untr_b',\n",
    "        'untr_a_hek',\n",
    "        'untr_b_hek',\n",
    "        'ribo_tm_a',\n",
    "        'ribo_tm_b',\n",
    "        'tm_a_hek',\n",
    "        'tm_b_hek',\n",
    "        'ribo_tmisrib_a',\n",
    "        'ribo_tmisrib_b',\n",
    "        'tmisrib_a_hek',\n",
    "        'tmisrib_b_hek',\n",
    "        'ribo_isrib_a',\n",
    "        'ribo_isrib_b',\n",
    "        'isrib_a_hek',\n",
    "        'isrib_b_hek'] # Designate sample order\n",
    "\n",
    "    # Run Spearman on raw counts\n",
    "    spear_list = []\n",
    "    for x in file_list:\n",
    "\n",
    "        # Get data as array-like for samples being compared\n",
    "        data_c1 = data.copy()\n",
    "        data_c2 = original_data.copy()\n",
    "\n",
    "        sample_a = data_c1[x].values.tolist()\n",
    "        sample_a = [x + 1 for x in sample_a]\n",
    "        sample_a = np.array(sample_a).astype(np.float)\n",
    "        sample_a = np.ndarray.tolist(sample_a)\n",
    "\n",
    "        sample_b = data_c2[x].values.tolist()\n",
    "        sample_b = [x + 1 for x in sample_b]\n",
    "        sample_b = np.array(sample_b).astype(np.float)\n",
    "        sample_b = np.ndarray.tolist(sample_b)\n",
    "\n",
    "        # Run Spearman R linreg for non-normal data\n",
    "        rho, spear_p_value = stats.spearmanr(sample_a, sample_b)\n",
    "        spear_list.append(rho)\n",
    "\n",
    "    # Run Pearson on log transformed data\n",
    "    pear_list = []\n",
    "    for x in file_list:\n",
    "\n",
    "        # Get data as array-like for samples being compared\n",
    "        data_c1 = data_t.copy()\n",
    "        data_c2 = original_data_t.copy()\n",
    "\n",
    "        sample_a = data_c1[x].values.tolist()\n",
    "        sample_a = [x + 1 for x in sample_a]\n",
    "        sample_a = np.array(sample_a).astype(np.float)\n",
    "        sample_a = np.ndarray.tolist(sample_a)\n",
    "\n",
    "        sample_b = data_c2[x].values.tolist()\n",
    "        sample_b = [x + 1 for x in sample_b]\n",
    "        sample_b = np.array(sample_b).astype(np.float)\n",
    "        sample_b = np.ndarray.tolist(sample_b)\n",
    "\n",
    "        r, pear_p_value = stats.pearsonr(sample_a, sample_b)\n",
    "        pear_list.append(r)\n",
    "\n",
    "    spear_list_reps = []\n",
    "    file_number = 0\n",
    "    for x in range(int(len(file_list)/2)):\n",
    "\n",
    "        # Get data as array-like for samples being compared\n",
    "        data_c1 = data.copy()\n",
    "\n",
    "        sample_a = data_c1[file_list[file_number]].values.tolist()\n",
    "        sample_a = [x + 1 for x in sample_a]\n",
    "        sample_a = np.array(sample_a).astype(np.float)\n",
    "        sample_a = np.ndarray.tolist(sample_a)\n",
    "\n",
    "        sample_b = data_c1[file_list[file_number+1]].values.tolist()\n",
    "        sample_b = [x + 1 for x in sample_b]\n",
    "        sample_b = np.array(sample_b).astype(np.float)\n",
    "        sample_b = np.ndarray.tolist(sample_b)\n",
    "\n",
    "        file_number += 2\n",
    "\n",
    "        # Run Spearman R linreg for non-normal data\n",
    "        rho, spear_p_value = stats.spearmanr(sample_a, sample_b)\n",
    "        spear_list_reps.append(rho)\n",
    "\n",
    "\n",
    "    pear_list_reps = []\n",
    "    file_number = 0\n",
    "    for x in range(int(len(file_list)/2)):\n",
    "\n",
    "        # Get data as array-like for samples being compared\n",
    "        data_c1 = data_t.copy()\n",
    "\n",
    "        sample_a = data_c1[file_list[file_number]].values.tolist()\n",
    "        sample_a = [x + 1 for x in sample_a]\n",
    "        sample_a = np.array(sample_a).astype(np.float)\n",
    "        sample_a = np.ndarray.tolist(sample_a)\n",
    "\n",
    "        sample_b = data_c1[file_list[file_number+1]].values.tolist()\n",
    "        sample_b = [x + 1 for x in sample_b]\n",
    "        sample_b = np.array(sample_b).astype(np.float)\n",
    "        sample_b = np.ndarray.tolist(sample_b)\n",
    "\n",
    "        file_number += 2\n",
    "\n",
    "        r, pear_p_value = stats.pearsonr(sample_a, sample_b)\n",
    "        pear_list_reps.append(r)\n",
    "\n",
    "    data = [\n",
    "        spear_list_reps,\n",
    "        pear_list_reps,\n",
    "        spear_list,\n",
    "        pear_list]\n",
    "\n",
    "    names = [\n",
    "        'Spearman Replicates',\n",
    "        'Pearson Replicates',\n",
    "        'Spearman Method Comparison',\n",
    "        'Pearson Method Comparison']\n",
    "\n",
    "    counter = 0\n",
    "    df = pd.DataFrame(columns = ['Value', 'Type'])\n",
    "\n",
    "    for x in range(4):\n",
    "        l = data[x]\n",
    "        n = names[x]\n",
    "\n",
    "        for y in l:\n",
    "            df.loc[counter] = [float(y), str(n)]\n",
    "            counter += 1\n",
    "\n",
    "    df1 = df.loc[df['Type'].str.contains('Replicates')]\n",
    "    df1['Type'] = df1['Type'].str.split(' ').str[0]\n",
    "    df1.columns = ['Coefficient Value', 'Replicates']\n",
    "    ax1 = sns.violinplot(x=\"Coefficient Value\", y=\"Replicates\", data=df1)\n",
    "    plt.savefig(\n",
    "        str(plot_dir) + 'replicates_' + str(title),\n",
    "        dpi = 1800,\n",
    "        bbox_inches = 'tight')\n",
    "    plt.close()\n",
    "\n",
    "    df2 = df.loc[df['Type'].str.contains('Method')]\n",
    "    df2['Type'] = df2['Type'].str.split(' ').str[0]\n",
    "    df2.columns = ['Coefficient Value', 'Method Comparison']\n",
    "    ax2 = sns.violinplot(x=\"Coefficient Value\", y=\"Method Comparison\", data=df2)\n",
    "    plt.savefig(\n",
    "        str(plot_dir) + 'method_' + str(title),\n",
    "        dpi = 1800,\n",
    "        bbox_inches = 'tight')\n",
    "    plt.close()\n",
    "\n",
    "\"\"\"Make Figure 2A\n",
    "\"\"\"\n",
    "def make_internal_correlations(\n",
    "        data,\n",
    "        title):\n",
    "\n",
    "    fig, axes = plt.subplots(\n",
    "        nrows = 2,\n",
    "        ncols = 2,\n",
    "        figsize = (20, 20),\n",
    "        subplot_kw = {\n",
    "            'facecolor':'none'}) # Create shared axis for cleanliness\n",
    "    plt.subplots_adjust(\n",
    "        bottom = 0.1)\n",
    "    plt.yticks([0,1,2,3,4,5]) # Limit axis labels to ints\n",
    "    plt.xticks([0,1,2,3,4,5])\n",
    "\n",
    "    x = 0\n",
    "    file_number = 0\n",
    "    file_list = [\n",
    "        'ribo_tm_a',\n",
    "        'ribo_tm_b',\n",
    "        'isrib_a_hek',\n",
    "        'isrib_b_hek'] # Designate sample order\n",
    "\n",
    "    for y in range(int(len(file_list)/2)):\n",
    "\n",
    "        # Get data as array-like for samples being compared\n",
    "        data_c1 = data.copy()\n",
    "\n",
    "        sample_a = data_c1[file_list[x]].values.tolist()\n",
    "        sample_a = [x + 1 for x in sample_a]\n",
    "        sample_a = np.array(sample_a).astype(np.float)\n",
    "        sample_a = np.ndarray.tolist(sample_a)\n",
    "\n",
    "        sample_b = data_c1[file_list[x+1]].values.tolist()\n",
    "        sample_b = [x + 1 for x in sample_b]\n",
    "        sample_b = np.array(sample_b).astype(np.float)\n",
    "        sample_b = np.ndarray.tolist(sample_b)\n",
    "\n",
    "        # Run Spearman R linreg for non-normal data\n",
    "        rho, p_value = stats.spearmanr(sample_a, sample_b)\n",
    "\n",
    "        # Determine subplot location\n",
    "        if file_number in [0,1]:\n",
    "            ax_x = file_number % 2\n",
    "            ax_y = 0\n",
    "        else:\n",
    "            print('oops')\n",
    "\n",
    "        # Format p value\n",
    "        if p_value < 0.001:\n",
    "            p_val = '< 0.001'\n",
    "        else:\n",
    "            p_val = '= {:.3f}'.format(round(p_value, 3))\n",
    "\n",
    "        rho = '{:.3f}'.format(round(rho, 3))\n",
    "\n",
    "        # Plot data\n",
    "        axes[ax_y, ax_x].tick_params(axis='both', labelsize=32)\n",
    "        axes[ax_y, ax_x].scatter(np.log10(sample_a), np.log10(sample_b), s=1,c='black')\n",
    "        axes[ax_y, ax_x].set_title(r\"$\\rho$\" + ' = ' + str(rho) + '\\nP ' + p_val, y=0.1, x=0.9, fontsize=32) # Format titles\n",
    "        axes[ax_y, ax_x].axhline(0, ls='-', color='black', xmin=0.0457, xmax=1) # Create axis lines\n",
    "        axes[ax_y, ax_x].axvline(0, ls='-', color='black', ymin=0.0457, ymax=1)\n",
    "        file_number += 1 # Plot counter\n",
    "        x += 2\n",
    "        print(rho)\n",
    "\n",
    "    # Create shared row/column titles\n",
    "\n",
    "    for ax in axes[:,0]:\n",
    "        ax.set_ylabel('log$_1$$_0$(counts)', fontsize=32)\n",
    "\n",
    "    cols = ['RPF Tm','mRNA ISRIB']\n",
    "    for ax, col in zip(axes[0], cols):\n",
    "        ax.set_xlabel(col, fontsize=40)\n",
    "        ax.xaxis.set_label_position('top')\n",
    "    cols = ['log$_1$$_0$(counts)','log$_1$$_0$(counts)']\n",
    "    for ax, col in zip(axes[1], cols):\n",
    "        ax.set_xlabel(col, fontsize=32, labelpad=30)\n",
    "        ax.xaxis.set_label_position('top')\n",
    "\n",
    "    fig.savefig(\n",
    "        str(plot_dir) + str(title),\n",
    "        dpi = 600,\n",
    "        bbox_inches = 'tight')\n",
    "\n",
    "def make_internal_correlations_supplement(\n",
    "        data,\n",
    "        title):\n",
    "\n",
    "    fig, axes = plt.subplots(\n",
    "        nrows = 2,\n",
    "        ncols = 4,\n",
    "        figsize = (20, 10),\n",
    "        subplot_kw = {\n",
    "            'facecolor':'none'},\n",
    "        sharex=True, sharey=True) # Create shared axis for cleanliness\n",
    "    plt.subplots_adjust(\n",
    "        bottom = 0.1)\n",
    "    plt.yticks([0,1,2,3,4,5]) # Limit axis labels to ints\n",
    "    plt.xticks([0,1,2,3,4,5])\n",
    "\n",
    "    x = 0\n",
    "    file_number = 0\n",
    "    file_list = [\n",
    "        'ribo_untr_a',\n",
    "        'ribo_untr_b',\n",
    "        'untr_a_hek',\n",
    "        'untr_b_hek',\n",
    "        'ribo_tm_a',\n",
    "        'ribo_tm_b',\n",
    "        'tm_a_hek',\n",
    "        'tm_b_hek',\n",
    "        'ribo_tmisrib_a',\n",
    "        'ribo_tmisrib_b',\n",
    "        'tmisrib_a_hek',\n",
    "        'tmisrib_b_hek',\n",
    "        'ribo_isrib_a',\n",
    "        'ribo_isrib_b',\n",
    "        'isrib_a_hek',\n",
    "        'isrib_b_hek'] # Designate sample order\n",
    "\n",
    "    for y in range(int(len(file_list)/2)):\n",
    "\n",
    "        # Get data as array-like for samples being compared\n",
    "        data_c1 = data.copy()\n",
    "\n",
    "        sample_a = data_c1[file_list[x]].values.tolist()\n",
    "        sample_a = [x + 1 for x in sample_a]\n",
    "        sample_a = np.array(sample_a).astype(np.float)\n",
    "        sample_a = np.ndarray.tolist(sample_a)\n",
    "\n",
    "        sample_b = data_c1[file_list[x+1]].values.tolist()\n",
    "        sample_b = [x + 1 for x in sample_b]\n",
    "        sample_b = np.array(sample_b).astype(np.float)\n",
    "        sample_b = np.ndarray.tolist(sample_b)\n",
    "\n",
    "        # Run Spearman R linreg for non-normal data\n",
    "        rho, p_value = stats.spearmanr(sample_a, sample_b)\n",
    "\n",
    "        # Determine subplot location\n",
    "        if file_number in [0,1,2,3]:\n",
    "            ax_x = file_number % 4\n",
    "            ax_y = 0\n",
    "        elif file_number in [4,5,6,7]:\n",
    "            ax_x = file_number - 4\n",
    "            ax_y = 1\n",
    "        else:\n",
    "            print('oops')\n",
    "\n",
    "        # Format p value\n",
    "        if p_value < 0.001:\n",
    "            p_val = '< 0.001'\n",
    "        else:\n",
    "            p_val = '= {:.3f}'.format(round(p_value, 3))\n",
    "\n",
    "        rho = '{:.3f}'.format(round(rho, 3))\n",
    "\n",
    "        # Plot data\n",
    "        axes[ax_y, ax_x].tick_params(axis='both', labelsize=32)\n",
    "        axes[ax_y, ax_x].scatter(np.log10(sample_a), np.log10(sample_b), s=1,c='black')\n",
    "        axes[ax_y, ax_x].set_title(r\"$\\rho$\" + ' = ' + str(rho) + '\\nP ' + p_val, y=0.1, x=0.8, fontsize=20) # Format titles\n",
    "        axes[ax_y, ax_x].axhline(0, ls='-', color='black', xmin=0.05, xmax=1) # Create axis lines\n",
    "        axes[ax_y, ax_x].axvline(0, ls='-', color='black', ymin=0.05, ymax=1)\n",
    "        file_number += 1 # Plot counter\n",
    "        x += 2\n",
    "        print(rho)\n",
    "\n",
    "    # Create shared row/column titles\n",
    "\n",
    "    for ax in axes[:,0]:\n",
    "        ax.set_ylabel('log$_1$$_0$(counts)', fontsize=32)\n",
    "\n",
    "    cols = ['RPF Untreated','mRNA Untreated','RPF Tm','mRNA Tm']\n",
    "    for ax, col in zip(axes[0], cols):\n",
    "        ax.set_xlabel(col, fontsize=32)\n",
    "        ax.xaxis.set_label_position('top')\n",
    "\n",
    "    cols = ['RPF Tm + ISRIB','mRNA Tm + ISRIB','RPF ISRIB','mRNA ISRIB']\n",
    "    for ax, col in zip(axes[1], cols):\n",
    "        ax.set_xlabel(col, fontsize=32)\n",
    "        ax.xaxis.set_label_position('top')\n",
    "\n",
    "    fig.savefig(\n",
    "        str(plot_dir) + str(title),\n",
    "        dpi = 600,\n",
    "        bbox_inches = 'tight')\n",
    "\n",
    "# Read in gene dictionary\n",
    "def make_ref(\n",
    "        ref_file,\n",
    "        gene_list):\n",
    "\n",
    "    df = pd.read_csv(\n",
    "        str(ref_file) ,\n",
    "        sep='\\t',\n",
    "        index_col=2)\n",
    "    del df.index.name\n",
    "    df = df[['NCBI gene ID', 'UniProt accession']]\n",
    "    df['NCBI gene ID'] = df['NCBI gene ID'].astype(pd.Int32Dtype())\n",
    "\n",
    "    df['index'] = df.index\n",
    "    df_keep = df.drop_duplicates(subset='index', keep='first')\n",
    "    df_genes = df_keep.reindex(gene_list)\n",
    "    df_genes = df_genes.drop('index', axis=1)\n",
    "\n",
    "    return df_genes\n",
    "\n",
    "\"\"\"Read in single-pass XPRESSpipe processed read data quantified with HTSeq using non-de-deuplicated alignments\n",
    "\"\"\"\n",
    "def get_data(\n",
    "        file,\n",
    "        sample_suffix='_1_Aligned'):\n",
    "\n",
    "    data = pd.read_csv(\n",
    "        file,\n",
    "        sep = '\\t',\n",
    "        index_col = 0)\n",
    "\n",
    "    data.index = data.index.str.split('.').str[0]\n",
    "\n",
    "    # Accessed via:\n",
    "    # $ curl -O ftp://ftp.ensembl.org/pub/release-96/gtf/homo_sapiens/Homo_sapiens.GRCh38.96.gtf.gz\n",
    "    # $ gzip -d Homo_sapiens.GRCh38.96.gtf.gz\n",
    "    data = convert_names(\n",
    "        data,\n",
    "        str(gtf_file))\n",
    "    data.columns = data.columns.str.replace(str(sample_suffix), '')\n",
    "    data.shape\n",
    "\n",
    "    # Combine lanes\n",
    "    sra_info = pd.read_csv(\n",
    "        str(sra_file),\n",
    "        sep = '\\t')\n",
    "    data = name_map(data, sra_info)\n",
    "    data = data.groupby(level=0).sum() # Combine duplicate named genes\n",
    "\n",
    "    return data\n",
    "\n",
    "\"\"\"Run correlations between sample alignments\n",
    "\"\"\"\n",
    "def cycle_external_correlations_supplement(\n",
    "        file,\n",
    "        original):\n",
    "\n",
    "    try:\n",
    "        data = get_data(file, sample_suffix='_1_Aligned')\n",
    "    except:\n",
    "        data = get_data(file, sample_suffix='__Aligned')\n",
    "\n",
    "    data_threshold = data.loc[data[['untr_a_hek', 'untr_b_hek', 'tm_a_hek', 'tm_b_hek', 'tmisrib_a_hek', 'tmisrib_b_hek', 'isrib_a_hek', 'isrib_b_hek']].min(axis=1) >= 10] # Apply threshold to data\n",
    "    data_rpm = rpm(data_threshold)\n",
    "    data_genes = data_rpm.index.tolist()\n",
    "\n",
    "    original_rpm = rpm(original)\n",
    "    original_genes = original_rpm.index.tolist()\n",
    "\n",
    "    common_genes = list(set(data_genes).intersection(original_genes))\n",
    "\n",
    "    data_common = data_rpm.reindex(index = common_genes)\n",
    "    original_common = original_rpm.reindex(index = common_genes)\n",
    "\n",
    "    make_external_correlations_supplement(\n",
    "        data_common,\n",
    "        original_common,\n",
    "        str(file.split('/')[-1][:-4]) + '_external_correlations_summary_htseq_all.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Differential Expression Methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Prep data vectors\n",
    "\"\"\"\n",
    "def prep_data(\n",
    "        data,\n",
    "        y_name,\n",
    "        x_name):\n",
    "\n",
    "    ribo_values = data[[y_name]].values.tolist()\n",
    "    ribo_values = np.array(ribo_values).astype(np.float)\n",
    "    ribo_values = np.ndarray.tolist(ribo_values)\n",
    "\n",
    "    rna_values = data[[x_name]].values.tolist()\n",
    "    rna_values = np.array(rna_values).astype(np.float)\n",
    "    rna_values = np.ndarray.tolist(rna_values)\n",
    "\n",
    "    return ribo_values, rna_values\n",
    "\n",
    "\"\"\"Plot optional data\n",
    "\"\"\"\n",
    "def plot_optional(\n",
    "        ax,\n",
    "        ribo_values,\n",
    "        rna_values,\n",
    "        data,\n",
    "        data_subset,\n",
    "        color,\n",
    "        alpha=1,\n",
    "        regulation_list=None,\n",
    "        significance_list=None,\n",
    "        label=False,\n",
    "        optional_genes_1=[],\n",
    "        optional_genes_2=[]):\n",
    "\n",
    "    # Plot first pass data\n",
    "    ax.scatter(\n",
    "        rna_values,\n",
    "        ribo_values,\n",
    "        s=80,\n",
    "        c=color,\n",
    "        alpha=alpha)\n",
    "\n",
    "    # Plot significance coloring\n",
    "    for index, row in data.iterrows():\n",
    "        if index in regulation_list:\n",
    "            if index in significance_list:\n",
    "                ax.scatter(\n",
    "                    row[1],\n",
    "                    row[0],\n",
    "                    s=20,\n",
    "                    c='black',\n",
    "                    alpha=1)\n",
    "\n",
    "            else:\n",
    "                ax.scatter(\n",
    "                    row[1],\n",
    "                    row[0],\n",
    "                    s=20,\n",
    "                    c='gray',\n",
    "                    alpha=1)\n",
    "\n",
    "        else:\n",
    "            pass\n",
    "\n",
    "    # Add labels\n",
    "    if label == True:\n",
    "\n",
    "        for index, row in data_subset.iterrows():\n",
    "\n",
    "            if index in optional_genes_1 + optional_genes_2:\n",
    "\n",
    "                for x in optional_genes_1:\n",
    "\n",
    "                    if index == x:\n",
    "                        ax.text(\n",
    "                            row[1] + 0.1,\n",
    "                            row[0] - 0.07,\n",
    "                            str(index),\n",
    "                            horizontalalignment='left',\n",
    "                            size='medium',\n",
    "                            color=color,\n",
    "                            weight='semibold')\n",
    "\n",
    "                for y in optional_genes_2:\n",
    "\n",
    "                    if index == y:\n",
    "                        ax.text(\n",
    "                            row[1] - 0.05,\n",
    "                            row[0] - 0.2,\n",
    "                            str(index),\n",
    "                            horizontalalignment='right',\n",
    "                            size='medium',\n",
    "                            color=color,\n",
    "                            weight='semibold')\n",
    "\n",
    "            else:\n",
    "                ax.text(\n",
    "                    row[1] - 0.1,\n",
    "                    row[0] - 0.07,\n",
    "                    str(index),\n",
    "                    horizontalalignment='right',\n",
    "                    size='medium',\n",
    "                    color=color,\n",
    "                    weight='semibold')\n",
    "\n",
    "    return ax\n",
    "\n",
    "\"\"\"Plot ribo-seq data on X/Y figure\n",
    "\"\"\"\n",
    "def rp_plot(\n",
    "        data,\n",
    "        y_fc,\n",
    "        x_fc,\n",
    "        y_p,\n",
    "        x_p,\n",
    "        y_name,\n",
    "        x_name,\n",
    "        te_data,\n",
    "        te_p,\n",
    "        title,\n",
    "        list_up,\n",
    "        list_down,\n",
    "        list_down_custom,\n",
    "        genes=[],\n",
    "        label_up=True,\n",
    "        label_down=True,\n",
    "        label_down_custom=True):\n",
    "\n",
    "    # Get relevant data\n",
    "    data_fig = data[[y_fc, x_fc, y_p, x_p]].copy()\n",
    "\n",
    "    # Remove infinite and missing values\n",
    "    data_fig[[y_fc, x_fc]] = data_fig[[y_fc, x_fc]].replace([np.inf, -np.inf], np.nan)\n",
    "    data_fig = data_fig.dropna(subset=[y_fc, x_fc])\n",
    "\n",
    "\n",
    "    # Initialize plotting space\n",
    "    fig, ax = plt.subplots(1,1, figsize=(7,7))\n",
    "    plt.grid(False)\n",
    "    ax.axhline(0, ls='-', color='black') # Create central axis line\n",
    "    ax.axvline(0, ls='-', color='black')\n",
    "    rect = patches.Rectangle((-1,-1),2,2,linewidth=1.5,edgecolor='gray',facecolor='none') # Create significance zone\n",
    "    ax.add_patch(rect)\n",
    "    ax.set_ylabel('Ribo-Seq log$_2$(' + str(y_name) + '/' + str(x_name) + ')', fontsize=16) # Set axis labels\n",
    "    ax.set_xlabel('mRNA-Seq log$_2$(' + str(y_name) + '/' + str(x_name) + ')', fontsize=16)\n",
    "    ax.set_xlim(-3.5,3.5) # Set axis limits\n",
    "    ax.set_ylim(-3.5,3.5)\n",
    "    x = [-3,-2,-1,0,1,2,3] # Set axis spacing\n",
    "    ticks = [\"-3\",\"-2\",\"-1\",\"0\",\"1\",\"2\",\"3\"]\n",
    "    ax.set_facecolor(\"#FFFFFF\") # Set background color\n",
    "\n",
    "    # Set X and Y axis scales\n",
    "    ax.set_xticks(x)\n",
    "    ax.set_xticklabels(ticks, fontsize=16)\n",
    "    ax.set_xticklabels([\"\"]*len(x), minor=True)\n",
    "\n",
    "    ax.set_yticks(x)\n",
    "    ax.set_yticklabels(ticks, fontsize=16)\n",
    "    ax.set_yticklabels([\"\"]*len(x), minor=True)\n",
    "\n",
    "    # Prep data for plotting\n",
    "    ribo_all, rna_all = prep_data(\n",
    "        data=data_fig,\n",
    "        y_name=y_fc,\n",
    "        x_name=x_fc)\n",
    "\n",
    "    # Prep significant hits for plotting\n",
    "    sig_list = te_data.loc[te_data[te_p] < 0.1].index.tolist()\n",
    "    data_sig = data_fig.reindex(sig_list)\n",
    "    ribo_sig, rna_sig = prep_data(\n",
    "        data=data_sig,\n",
    "        y_name=y_fc,\n",
    "        x_name=x_fc)\n",
    "\n",
    "    # Prep option gene sets\n",
    "    if list_up != None:\n",
    "        data_fig_up = data_fig.reindex(list_up)\n",
    "        ribo_up, rna_up = prep_data(\n",
    "            data=data_fig_up,\n",
    "            y_name=y_fc,\n",
    "            x_name=x_fc)\n",
    "\n",
    "    if list_down != None:\n",
    "        data_fig_down = data_fig.reindex(list_down)\n",
    "        ribo_down, rna_down = prep_data(\n",
    "            data=data_fig_down,\n",
    "            y_name=y_fc,\n",
    "            x_name=x_fc)\n",
    "\n",
    "    if list_down_custom != None:\n",
    "        data_fig_down_custom = data_fig.reindex(list_down_custom)\n",
    "        ribo_down_custom, rna_down_custom = prep_data(\n",
    "            data=data_fig_down_custom,\n",
    "            y_name=y_fc,\n",
    "            x_name=x_fc)\n",
    "\n",
    "    #Plot data\n",
    "    ax.scatter(rna_all, ribo_all, s=2.5,c='gray',alpha=0.5)\n",
    "    ax.scatter(rna_sig, ribo_sig, s=5,c='black',alpha=1)\n",
    "\n",
    "    if list_down != None:\n",
    "\n",
    "        ax = plot_optional(\n",
    "            ax=ax,\n",
    "            ribo_values=ribo_down,\n",
    "            rna_values=rna_down,\n",
    "            data=data_fig,\n",
    "            data_subset=data_fig_down,\n",
    "            color='#1b9e77',\n",
    "            regulation_list=list_down,\n",
    "            significance_list=sig_list,\n",
    "            label=label_down,\n",
    "            optional_genes_1=[\n",
    "                'PNRC2',\n",
    "                'SAT1']\n",
    "            )\n",
    "\n",
    "    if list_down_custom != None:\n",
    "\n",
    "        ax = plot_optional(\n",
    "            ax=ax,\n",
    "            ribo_values=ribo_down_custom,\n",
    "            rna_values=rna_down_custom,\n",
    "            data=data_fig,\n",
    "            data_subset=data_fig_down_custom,\n",
    "            color='#d95f02',\n",
    "            regulation_list=list_down_custom,\n",
    "            significance_list=sig_list,\n",
    "            label=label_down_custom,\n",
    "            optional_genes_1=[\n",
    "                'RPS15A',\n",
    "                'HIST2H3D',\n",
    "                'RPL27'],\n",
    "            optional_genes_2=[\n",
    "                'HSPA8']\n",
    "            )\n",
    "\n",
    "    if list_up != None:\n",
    "\n",
    "        ax = plot_optional(\n",
    "            ax=ax,\n",
    "            ribo_values=ribo_up,\n",
    "            rna_values=rna_up,\n",
    "            data=data_fig,\n",
    "            data_subset=data_fig_up,\n",
    "            color='#7570b3',\n",
    "            regulation_list=list_up,\n",
    "            significance_list=sig_list,\n",
    "            label=label_up,\n",
    "            optional_genes_1=[\n",
    "                'ATF4',\n",
    "                'DDIT3',\n",
    "                'PPP1R15A']\n",
    "            )\n",
    "\n",
    "    plt.savefig(\n",
    "        str(deseq_plot_directory) + str(title),\n",
    "        dpi=3600,\n",
    "        bbox_inches='tight')\n",
    "\n",
    "    plt.close()\n",
    "\n",
    "\"\"\"Import DE data\n",
    "\"\"\"\n",
    "def parse_de_data(\n",
    "        directory,\n",
    "        check_file):\n",
    "\n",
    "    check_data = pd.read_csv(\n",
    "        str(directory) + str(check_file),\n",
    "        sep='\\t',\n",
    "        index_col=0)\n",
    "    check_list = check_data.index.tolist()\n",
    "\n",
    "    # Import DESeq2 TE data\n",
    "    tm_data = pd.read_csv(str(directory) + 'tm_counts_diffx.tsv', sep='\\t')\n",
    "    tmisrib_data = pd.read_csv(str(directory) + 'tmisrib_counts_diffx.tsv', sep='\\t')\n",
    "    isrib_data = pd.read_csv(str(directory) + 'isrib_counts_diffx.tsv', sep='\\t')\n",
    "\n",
    "    ### Clean up this data\n",
    "    tm_data = tm_data.drop(['baseMean','lfcSE','stat'],axis=1)\n",
    "    tm_data.columns = ['tm_log2FC','tm_p','tm_padj']\n",
    "\n",
    "    tmisrib_data = tmisrib_data.drop(['baseMean','lfcSE','stat'],axis=1)\n",
    "    tmisrib_data.columns = ['tmisrib_log2FC','tmisrib_p','tmisrib_padj']\n",
    "\n",
    "    isrib_data = isrib_data.drop(['baseMean','lfcSE','stat'],axis=1)\n",
    "    isrib_data.columns = ['isrib_log2FC','isrib_p','isrib_padj']\n",
    "\n",
    "    merged_data = pd.concat([tm_data, tmisrib_data, isrib_data], axis=1, sort=False)\n",
    "\n",
    "    # Import DESeq2 RNA and RPF data\n",
    "    tm_ribo_data = pd.read_csv(str(directory) + 'tm_ribo_counts_diffx.tsv', sep='\\t')\n",
    "    tm_rna_data = pd.read_csv(str(directory) + 'tm_rna_counts_diffx.tsv', sep='\\t')\n",
    "\n",
    "    tmisrib_ribo_data = pd.read_csv(str(directory) + 'tmisrib_ribo_counts_diffx.tsv', sep='\\t')\n",
    "    tmisrib_rna_data = pd.read_csv(str(directory) + 'tmisrib_rna_counts_diffx.tsv', sep='\\t')\n",
    "\n",
    "    isrib_ribo_data = pd.read_csv(str(directory) + 'isrib_ribo_counts_diffx.tsv', sep='\\t')\n",
    "    isrib_rna_data = pd.read_csv(str(directory) + 'isrib_rna_counts_diffx.tsv', sep='\\t')\n",
    "\n",
    "    ### Clean up this data\n",
    "    tm_ribo_data = tm_ribo_data.drop(['baseMean','lfcSE','stat','pvalue'],axis=1)\n",
    "    tm_ribo_data.columns = ['tm_ribo_log2FC','tm_ribo_padj']\n",
    "    tm_rna_data = tm_rna_data.drop(['baseMean','lfcSE','stat','pvalue'],axis=1)\n",
    "    tm_rna_data.columns = ['tm_rna_log2FC','tm_rna_padj']\n",
    "\n",
    "    tmisrib_ribo_data = tmisrib_ribo_data.drop(['baseMean','lfcSE','stat','pvalue'],axis=1)\n",
    "    tmisrib_ribo_data.columns = ['tmisrib_ribo_log2FC','tmisrib_ribo_padj']\n",
    "    tmisrib_rna_data = tmisrib_rna_data.drop(['baseMean','lfcSE','stat','pvalue'],axis=1)\n",
    "    tmisrib_rna_data.columns = ['tmisrib_rna_log2FC','tmisrib_rna_padj']\n",
    "\n",
    "    isrib_ribo_data = isrib_ribo_data.drop(['baseMean','lfcSE','stat','pvalue'],axis=1)\n",
    "    isrib_ribo_data.columns = ['isrib_ribo_log2FC','isrib_ribo_padj']\n",
    "    isrib_rna_data = isrib_rna_data.drop(['baseMean','lfcSE','stat','pvalue'],axis=1)\n",
    "    isrib_rna_data.columns = ['isrib_rna_log2FC','isrib_rna_padj']\n",
    "\n",
    "    merged_data_split = pd.concat([tm_ribo_data, tm_rna_data, tmisrib_ribo_data, tmisrib_rna_data, isrib_ribo_data, isrib_rna_data], axis=1, sort=False)\n",
    "\n",
    "    return merged_data, merged_data_split, check_data\n",
    "\n",
    "\"\"\"TE sanity check\n",
    "\"\"\"\n",
    "def te_sanity(\n",
    "        data,\n",
    "        check_data,\n",
    "        gene_name):\n",
    "\n",
    "    check_data_rpm = rpm(check_data)\n",
    "\n",
    "    print(gene_name)\n",
    "\n",
    "    print('DESeq2 estimate: ', data.loc[gene_name]['tm_log2FC'])\n",
    "\n",
    "    # Get gene info for all samples\n",
    "    gene = check_data_rpm.loc[gene_name][['ribo_tm_a', 'ribo_tm_b', 'ribo_untr_a', 'ribo_untr_b','tm_a_hek','tm_b_hek','untr_a_hek','untr_b_hek']]\n",
    "\n",
    "    # Calculate Tm TE by summing replicates and calculating RPF / RNA\n",
    "    gene['tm_te'] = (gene['ribo_tm_a'] + gene['ribo_tm_b']) / (gene['tm_a_hek'] + gene['tm_b_hek'])\n",
    "    gene['untr_te'] = (gene['ribo_untr_a'] + gene['ribo_untr_b']) / (gene['untr_a_hek'] + gene['untr_b_hek'])\n",
    "\n",
    "    # Calculate FC of TEs for Treatment / not-treated and take the log2 value of that outpu\n",
    "    gene['fold_change'] = np.log2(gene['tm_te'] / gene['untr_te'])\n",
    "    print('Sanity estimate: ', gene['fold_change'])\n",
    "\n",
    "\"\"\"Plot TE for genes of interest\n",
    "\"\"\"\n",
    "def te_plot(\n",
    "        data,\n",
    "        list_up,\n",
    "        list_down,\n",
    "        directory,\n",
    "        plot_all=False,\n",
    "        dpi=3600):\n",
    "\n",
    "    # Init plotting space\n",
    "    fig, ax = plt.subplots()\n",
    "    plt.yticks([-4,-3,-2,-1,0,1,2,3,4,5,6,7,8,9,10])\n",
    "    ax.tick_params(axis='x', which='minor',length=0)\n",
    "\n",
    "    ax.set_facecolor('white')\n",
    "    ax.grid(color='grey', axis='y')\n",
    "\n",
    "    # Plot all genes' TE\n",
    "    if plot_all == True:\n",
    "        data_de_plot.T.plot.line(\n",
    "            legend=False,\n",
    "            color='lightgrey',\n",
    "            linewidth=0.5,\n",
    "            ax=ax)\n",
    "\n",
    "    # Set grid lines\n",
    "    ax.axvline(0.01, ls='-', color='black')\n",
    "    ax.axvline(1, ls='-', color='black')\n",
    "    ax.axvline(2, ls='-', color='black')\n",
    "    ax.axvline(2.99, ls='-', color='black')\n",
    "    ax.axhline(0, ls='-', color='black')\n",
    "    ax.axhline(1, ls='--', color='black')\n",
    "    ax.axhline(-1, ls='--', color='black')\n",
    "\n",
    "    # Plot genes of interest\n",
    "    data_de_plot.loc[list_up].T.plot.line(\n",
    "        legend=False,\n",
    "        color='#7570b3',\n",
    "        linewidth=1,\n",
    "        ax=ax)\n",
    "    data_de_plot.loc[list_down].T.plot.line(\n",
    "        legend=False,\n",
    "        color='#d95f02',\n",
    "        linewidth=1,\n",
    "        ax=ax)\n",
    "    ax.set_ylabel(u'log$_2$(TE)')\n",
    "\n",
    "    # Add legend\n",
    "    from matplotlib.lines import Line2D\n",
    "    legend_elements = [Line2D([0], [0], color='gray', lw=2, label='All'),\n",
    "                       Line2D([0], [0], color='#7570b3', lw=2, label='ISR'),\n",
    "                       Line2D([0], [0], color='#d95f02', lw=2, label='Other')]\n",
    "    ax.legend(handles=legend_elements, loc='upper right')\n",
    "\n",
    "    plt.savefig(\n",
    "        str(directory) + 'te_analysis_down_strict.png',\n",
    "        dpi=dpi,\n",
    "        bbox_inches='tight')\n",
    "\n",
    "    plt.close()\n",
    "\n",
    "\"\"\"Plot Venn diagrams\n",
    "\"\"\"\n",
    "def venn_plot(\n",
    "        directory,\n",
    "        figure_title,\n",
    "        output_name,\n",
    "        gene_list_A,\n",
    "        gene_list_B,\n",
    "        label_A,\n",
    "        label_B,\n",
    "        color_A='#d95f02',\n",
    "        color_B='#7570b3',\n",
    "        dpi=3600):\n",
    "\n",
    "    plt.title(\n",
    "        figure_title,\n",
    "        fontsize = 28)\n",
    "\n",
    "    c = venn2(\n",
    "        [set(gene_list_A), set(gene_list_B)],\n",
    "        set_labels = (label_A, label_B))\n",
    "    c.get_patch_by_id('10').set_color(color_A)\n",
    "    c.get_patch_by_id('01').set_color(color_B)\n",
    "    c.get_patch_by_id('10').set_edgecolor('black')\n",
    "    c.get_patch_by_id('01').set_edgecolor('black')\n",
    "\n",
    "    for text in c.set_labels:\n",
    "\n",
    "        text.set_fontsize(22)\n",
    "\n",
    "    try:\n",
    "        for text in c.subset_labels:\n",
    "\n",
    "            text.set_fontsize(18)\n",
    "\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "    plt.savefig(\n",
    "        directory + output_name,\n",
    "        dpi=dpi,\n",
    "        bbox_inches='tight')\n",
    "\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Process Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Read in XPRESSpipe-processed data and process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input file is protein coding only and truncated, not parsed for longest transcript only\n",
    "data = get_data(file, sample_suffix='_1_Aligned')\n",
    "\n",
    "# Clean up data\n",
    "data_threshold_deseq = data.loc[\n",
    "    data[[\n",
    "        'untr_a_hek', \n",
    "        'untr_b_hek', \n",
    "        'tm_a_hek', \n",
    "        'tm_b_hek', \n",
    "        'tmisrib_a_hek', \n",
    "        'tmisrib_b_hek', \n",
    "        'isrib_a_hek', \n",
    "        'isrib_b_hek']\n",
    "    ].min(axis=1) >= 25] # Apply threshold to data\n",
    "\n",
    "data_threshold = data.loc[\n",
    "    data[[\n",
    "        'untr_a_hek', \n",
    "        'untr_b_hek', \n",
    "        'tm_a_hek', \n",
    "        'tm_b_hek', \n",
    "        'tmisrib_a_hek', \n",
    "        'tmisrib_b_hek', \n",
    "        'isrib_a_hek', \n",
    "        'isrib_b_hek']\n",
    "    ].min(axis=1) >= 10] # Apply threshold to data\n",
    "\n",
    "data_rpm = rpm(data_threshold)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Export processed XPRESSpipe data for DESeq2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export truth set for downsteam checks\n",
    "data_threshold_deseq.to_csv(\n",
    "    deseq_directory + 'ISRIBxpresspipe_thresholded_counts.tsv',\n",
    "    sep='\\t')\n",
    "\n",
    "# Define sample sets\n",
    "untr_mrna = ['untr_a_hek', 'untr_b_hek']\n",
    "untr_ribo = ['ribo_untr_a', 'ribo_untr_b']\n",
    "tm_mrna = ['tm_a_hek', 'tm_b_hek']\n",
    "tm_ribo = ['ribo_tm_a', 'ribo_tm_b']\n",
    "tmisrib_mrna = ['tmisrib_a_hek', 'tmisrib_b_hek']\n",
    "tmisrib_ribo = ['ribo_tmisrib_a', 'ribo_tmisrib_b']\n",
    "isrib_mrna = ['isrib_a_hek', 'isrib_b_hek']\n",
    "isrib_ribo = ['ribo_isrib_a', 'ribo_isrib_b']\n",
    "\n",
    "# Output count tables\n",
    "data_threshold_deseq[untr_ribo + tm_ribo].to_csv(\n",
    "    deseq_directory + 'tm_ribo_counts.tsv',\n",
    "    sep='\\t')\n",
    "data_threshold_deseq[untr_ribo + tmisrib_ribo].to_csv(\n",
    "    deseq_directory + 'tmisrib_ribo_counts.tsv',\n",
    "    sep='\\t')\n",
    "data_threshold_deseq[untr_ribo + isrib_ribo].to_csv(\n",
    "    deseq_directory + 'isrib_ribo_counts.tsv',\n",
    "    sep='\\t')\n",
    "\n",
    "data_threshold_deseq[untr_mrna + tm_mrna].to_csv(\n",
    "    deseq_directory + 'tm_rna_counts.tsv',\n",
    "    sep='\\t')\n",
    "data_threshold_deseq[untr_mrna + tmisrib_mrna].to_csv(\n",
    "    deseq_directory + 'tmisrib_rna_counts.tsv',\n",
    "    sep='\\t')\n",
    "data_threshold_deseq[untr_mrna + isrib_mrna].to_csv(\n",
    "    deseq_directory + 'isrib_rna_counts.tsv',\n",
    "    sep='\\t')\n",
    "\n",
    "data_threshold_deseq[untr_ribo + tm_ribo + untr_mrna + tm_mrna].to_csv(\n",
    "    deseq_directory + 'tm_counts.tsv',\n",
    "    sep='\\t')\n",
    "data_threshold_deseq[untr_ribo + tmisrib_ribo + untr_mrna + tmisrib_mrna].to_csv(\n",
    "    deseq_directory + 'tmisrib_counts.tsv',\n",
    "    sep='\\t')\n",
    "data_threshold_deseq[untr_ribo + isrib_ribo + untr_mrna + isrib_mrna].to_csv(\n",
    "    deseq_directory + 'isrib_counts.tsv',\n",
    "    sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Read in original count data and process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in original raw counts from Elife supplement table\n",
    "original = pd.read_csv(\n",
    "    str(original),\n",
    "    sep = '\\t',\n",
    "    index_col = 0)\n",
    "original = original.drop('size', axis = 1)\n",
    "\n",
    "# Clean up data\n",
    "original = original.groupby(level=0).sum() # Combine duplicate named genes\n",
    "original_rpm = rpm(original)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Parse through original data and XPRESSpipe-processed and return dataframes with shared genes only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clean NaNs\n",
    "data_rpm.shape\n",
    "original_rpm.shape\n",
    "\n",
    "data_rpm = data_rpm.dropna()\n",
    "original_rpm = original_rpm.dropna()\n",
    "\n",
    "data_rpm.shape\n",
    "original_rpm.shape\n",
    "\n",
    "# Make dataframes for each dataset for matching genes\n",
    "data_genes = data_rpm.index.tolist()\n",
    "original_genes = original_rpm.index.tolist()\n",
    "len(data_genes)\n",
    "len(original_genes)\n",
    "\n",
    "common_genes = list(set(data_genes).intersection(original_genes))\n",
    "len(common_genes)\n",
    "\n",
    "data_common = data_rpm.reindex(index = common_genes)\n",
    "original_common = original_rpm.reindex(index = common_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Over-representation analysis\n",
    "TopHat2 is .48% more likely to align a read that should be counted as ambiguous\n",
    "https://www.nature.com/articles/nmeth.4106, supplementary table 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tophat_overcount_rate = '0.48%'\n",
    "\n",
    "#RPF Untr A -> started with ~135,000,000 reads\n",
    "rpf_start_reads = 135000000 #approx\n",
    "\n",
    "rpf_test = pd.concat([data_common[['ribo_untr_a']], original_common[['ribo_untr_a']]], axis=1, sort=False)\n",
    "rpf_test.columns = ['xpresspipe','original']\n",
    "rpf_test = rpf_test[(rpf_test['xpresspipe'] >= 10) & (rpf_test['original'] >= 10)]\n",
    "\n",
    "rpf_test.shape\n",
    "rpf_test.sum()\n",
    "rpf_log = np.log10(rpf_test+1)\n",
    "\n",
    "plt.scatter(rpf_log['xpresspipe'].values, rpf_log['original'], s=1)\n",
    "x, y = [0, 4.9], [-0.05, 5.05]\n",
    "plt.plot(x, y, c='black')\n",
    "\n",
    "m = (y[1] - y[0]) / (x[1] - x[0])\n",
    "b = y[0]\n",
    "\n",
    "rpf_above = rpf_log.loc[rpf_log['original'] > ((rpf_log['xpresspipe'] * m) + b)]\n",
    "rpf_above.shape\n",
    "rpf_above['diff'] = 10**rpf_above['original'] - 10**rpf_above['xpresspipe']\n",
    "overcounts = rpf_above['diff'].sum()\n",
    "overcount_rate = overcounts / rpf_start_reads\n",
    "print('RPF_Untr_A')\n",
    "print('# Reads overcounted: ' + str(int(round(overcounts))))\n",
    "print('Overcount rate: ' + str(overcount_rate * 100) + '%')\n",
    "print('TopHat2 vs STAR overcount ambiguous rate increase: ' + tophat_overcount_rate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#RNA Untr A\n",
    "rna_start_reads = 48000000\n",
    "\n",
    "rna_test = pd.concat([data_common[['untr_a_hek']], original_common[['untr_a_hek']]], axis=1, sort=False)\n",
    "rna_test.columns = ['xpresspipe','original']\n",
    "rna_test.shape\n",
    "rna_test.sum()\n",
    "\n",
    "rna_log = np.log10(rna_test+1)\n",
    "\n",
    "plt.scatter(rna_log['xpresspipe'].values, rna_log['original'], s=1)\n",
    "x, y = [0, 4.9], [0.18, 5.03]\n",
    "plt.plot(x, y, c='black')\n",
    "\n",
    "m = (y[1] - y[0]) / (x[1] - x[0])\n",
    "b = y[0]\n",
    "\n",
    "rna_above = rna_log.loc[rna_log['original'] > ((rna_log['xpresspipe'] * m) + b)]\n",
    "print(rna_above.shape)\n",
    "rna_above['diff'] = 10**rna_above['original'] - 10**rna_above['xpresspipe']\n",
    "overcounts = rna_above['diff'].sum()\n",
    "\n",
    "rna_above['diff'].index.tolist()\n",
    "\n",
    "\n",
    "plt.scatter(rna_log['xpresspipe'].loc[rna_above['diff'].index.tolist()].values, rna_log['original'].loc[rna_above['diff'].index.tolist()], s=5)\n",
    "x, y = [0, 4.9], [0.18, 5.03]\n",
    "plt.plot(x, y, c='red')\n",
    "\n",
    "len(rna_above['diff'].index.tolist())\n",
    "\n",
    "overcount_rate = overcounts / rna_start_reads\n",
    "print('RNA_Untr_A')\n",
    "print('# Reads overcounted: ' + str(int(round(overcounts))))\n",
    "print('Overcount rate: ' + str(overcount_rate * 100) + '%')\n",
    "print('TopHat2 vs STAR overcount ambiguous rate increase: ' + tophat_overcount_rate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Some further analysis\n",
    "len(rpf_above['diff'].index.tolist())\n",
    "len(rna_above['diff'].index.tolist())\n",
    "\n",
    "common_above = list(set(rpf_above['diff'].index.tolist()).intersection(rna_above['diff'].index.tolist()))\n",
    "len(common_above)\n",
    "\n",
    "from urllib.request import Request, urlopen\n",
    "\n",
    "target = \"important paralog\"\n",
    "counter = 0\n",
    "gene_number = 0\n",
    "for gene in common_above:\n",
    "\n",
    "    gene_number += 1\n",
    "\n",
    "    url = \"https://www.genecards.org/cgi-bin/carddisp.pl?gene=\" + gene\n",
    "    req = Request(url, headers={'User-Agent': 'Mozilla/5.0'})\n",
    "    webpage = urlopen(req).read()\n",
    "    page = webpage.decode('utf-8').splitlines()\n",
    "\n",
    "    if any(target in s for s in page) == True:\n",
    "        counter += 1\n",
    "\n",
    "print(\n",
    "    'Number of genes above threshold with annotated paralogs:' \n",
    "    + str(counter) \n",
    "    + '/' \n",
    "    + len(common_above)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Process internal correlations (within-method)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run correlations between sample alignments\n",
    "make_internal_correlations(\n",
    "    data_rpm,\n",
    "    'internal_correlations_summary_htseq_protein_truncated_v98.png')\n",
    "\n",
    "make_internal_correlations_supplement(\n",
    "    data_rpm,\n",
    "    'internal_correlations_summary_htseq_protein_truncated_v98_all.png')\n",
    "\n",
    "make_internal_correlations_supplement(\n",
    "    original_rpm,\n",
    "    'internal_correlations_summary_ingolia.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### External correlations (between-methods)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_list = [\n",
    "    'isrib_comp_v98_truncated_count_table.tsv',\n",
    "    'isrib_comp_v98_normal_count_table.tsv',\n",
    "    'isrib_comp_v98_longest_truncated_count_table.tsv',\n",
    "    'isrib_comp_v72_normal_count_table.tsv',\n",
    "    'isrib_comp_v72_longest_truncated_count_table.tsv',\n",
    "    'isrib_comp_v72_truncated_count_table.tsv']\n",
    "\n",
    "file_list = [str(comparison_tables_directory) + str(f) for f in file_list]\n",
    "\n",
    "for file in file_list:\n",
    "\n",
    "    cycle_external_correlations_supplement(file, original)\n",
    "\n",
    "make_external_correlations(\n",
    "    data_common,\n",
    "    original_common,\n",
    "    'external_correlations_summary_htseq_protein_truncated_v98.png')\n",
    "\n",
    "make_external_correlations_supplement(\n",
    "    data_common,\n",
    "    original_common,\n",
    "    'external_correlations_summary_htseq_protein_truncated_v98_all.png')\n",
    "\n",
    "make_external_correlations_supplement(\n",
    "    data_common,\n",
    "    original_common,\n",
    "    '',\n",
    "    interactive=True)\n",
    "\n",
    "make_spearman_pearson_correlations(\n",
    "    data_common,\n",
    "    original_common,\n",
    "    'spearman_pearson_correlations_htseq_protein_truncated_v98.png')\n",
    "\n",
    "make_paralogs_analysis(\n",
    "    data_common,\n",
    "    original_common,\n",
    "    'highlight_paralogs_comparison_isrib.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Differential Translation Efficiency Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Import DESeq2 count data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import count data for sanity checks and DESeq2 output tables\n",
    "merged_data_XP, merged_data_split_XP, check_data_XP = parse_de_data(\n",
    "        directory=deseq_directory,\n",
    "        check_file='ISRIBxpresspipe_thresholded_counts.tsv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Run some sanity checks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Some of the genes in plotting cluster on a single y-axis coordinate\n",
    "# This is due to insufficient RPF counts after RNA thresholding of the data\n",
    "merged_data_split_XP[\n",
    "    (merged_data_split_XP['tm_ribo_log2FC'] > 2.7) \n",
    "    & (merged_data_split_XP['tm_ribo_log2FC'] < 2.75)\n",
    "][['tm_ribo_log2FC', 'tm_ribo_padj']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check fold changes of targets for comparison\n",
    "print('Raw fold changes:')\n",
    "print('+++++++++++++++++')\n",
    "print('ATF4')\n",
    "print(2**merged_data_split_XP.loc['ATF4']['tm_ribo_log2FC'])\n",
    "print('=====')\n",
    "print('ATF5')\n",
    "print(2**merged_data_split_XP.loc['ATF5']['tm_ribo_log2FC'])\n",
    "print('=====')\n",
    "print('PPP1R15A')\n",
    "print(2**merged_data_split_XP.loc['PPP1R15A']['tm_ribo_log2FC'])\n",
    "print('=====')\n",
    "print('DDIT3')\n",
    "print(2**merged_data_split_XP.loc['DDIT3']['tm_ribo_log2FC'])\n",
    "print('=====')\n",
    "\n",
    "print('\\nlog2FC comparisons:')\n",
    "print('+++++++++++++++++++++')\n",
    "te_sanity(\n",
    "    data=merged_data_XP[['tm_log2FC', 'tm_padj']],\n",
    "    check_data=check_data_XP,\n",
    "    gene_name='ATF4')\n",
    "print('=====')\n",
    "\n",
    "te_sanity(\n",
    "    data=merged_data_XP[['tm_log2FC', 'tm_padj']],\n",
    "    check_data=check_data_XP,\n",
    "    gene_name='DDIT3')\n",
    "print('=====')\n",
    "\n",
    "te_sanity(\n",
    "    data=merged_data_XP[['tm_log2FC', 'tm_padj']],\n",
    "    check_data=check_data_XP,\n",
    "    gene_name='UCP2')\n",
    "print('=====')\n",
    "\n",
    "te_sanity(\n",
    "    data=merged_data_XP[['tm_log2FC', 'tm_padj']],\n",
    "    check_data=check_data_XP,\n",
    "    gene_name='POMGNT1')\n",
    "print('=====')\n",
    "\n",
    "te_sanity(\n",
    "    data=merged_data_XP[['tm_log2FC', 'tm_padj']],\n",
    "    check_data=check_data_XP,\n",
    "    gene_name='NDUFA11')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Perform pattern analysis on TE data to pull out interesting genes across conditions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TE analysis\n",
    "data_de_plot = merged_data_XP.copy()\n",
    "\n",
    "data_de_plot['Untreated'] = 0\n",
    "data_de_plot = data_de_plot[['Untreated','tm_log2FC', 'tmisrib_log2FC','isrib_log2FC']]\n",
    "data_de_plot.columns = ['Untreated','Tm', 'Tm + ISRIB','ISRIB']\n",
    "\n",
    "down_strict_1_XP = merged_data_XP.loc[\n",
    "    (merged_data_XP['tm_log2FC'] - merged_data_XP['tmisrib_log2FC'] <= -1) \\\n",
    "    & (merged_data_XP['tm_padj'] <= 0.001) \\\n",
    "    & (merged_data_XP['tmisrib_padj'] <= 0.001) \\\n",
    "    & ((merged_data_split_XP['tm_ribo_padj'] <= 0.1))]\n",
    "\n",
    "down_strict_2_XP = merged_data_XP.loc[\n",
    "    (merged_data_XP['tm_log2FC'] - merged_data_XP['tmisrib_log2FC'] <= -0.58) \\\n",
    "    & (merged_data_split_XP['tm_ribo_log2FC'] <= -0.9) \\\n",
    "    & (merged_data_split_XP['tm_ribo_padj'] <= 0.1)]\n",
    "\n",
    "down_strict_XP = down_strict_1_XP.index.tolist() + down_strict_2_XP.index.tolist()\n",
    "\n",
    "# Sanity check count data\n",
    "print(check_data_XP.loc[down_strict_XP].min().min())\n",
    "\n",
    "merged_data_XP.loc[down_strict_XP][['tm_log2FC','tm_padj','tmisrib_log2FC','tmisrib_padj']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### False-discovery Rate check\n",
    "- Using the alpha-debt method from dataset decay theory (https://doi.org/10.1101/801696)\n",
    "- Number of PubMed citations for ISRIB paper == 199\n",
    "- In random sample of 20 (~10%), only 2 used for some sort of re-analysis\n",
    "- Will assume a 1/10 use-rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n = 199 * (2/20)\n",
    "alpha = 0.05\n",
    "from math import ceil\n",
    "for x in range(0, ceil(n)):\n",
    "    alpha = alpha / 2\n",
    "print('Dataset decay alpha: ' + str(alpha))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Make riboseq vs RNAseq plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tm\n",
    "rp_plot(\n",
    "    merged_data_split_XP,\n",
    "    'tm_ribo_log2FC',\n",
    "    'tm_rna_log2FC',\n",
    "    'tm_ribo_padj',\n",
    "    'tm_rna_padj',\n",
    "    'Tm',\n",
    "    'Untr',\n",
    "    merged_data_XP[['tm_log2FC','tm_padj']],\n",
    "    'tm_padj',\n",
    "    'Tm_vs_Untr_deseq.png',\n",
    "    isr,\n",
    "    uorf_targets,\n",
    "    down_strict_XP)\n",
    "\n",
    "# Tm + ISRIB\n",
    "rp_plot(\n",
    "    merged_data_split_XP,\n",
    "    'tmisrib_ribo_log2FC',\n",
    "    'tmisrib_rna_log2FC',\n",
    "    'tmisrib_ribo_padj',\n",
    "    'tmisrib_rna_padj',\n",
    "    'Tm + ISRIB',\n",
    "    'Untr',\n",
    "    merged_data_XP[['tmisrib_log2FC','tmisrib_padj']],\n",
    "    'tmisrib_padj',\n",
    "    'TmISRIB_vs_Untr_deseq.png',\n",
    "    isr,\n",
    "    uorf_targets,\n",
    "    down_strict_XP,\n",
    "    label_up=False,\n",
    "    label_down=False,\n",
    "    label_down_custom=False)\n",
    "\n",
    "# ISRIB\n",
    "rp_plot(\n",
    "    merged_data_split_XP,\n",
    "    'isrib_ribo_log2FC',\n",
    "    'isrib_rna_log2FC',\n",
    "    'isrib_ribo_padj',\n",
    "    'isrib_rna_padj',\n",
    "    'ISRIB',\n",
    "    'Untr',\n",
    "    merged_data_XP[['isrib_log2FC','isrib_padj']],\n",
    "    'isrib_padj',\n",
    "    'ISRIB_vs_Untr_deseq.png',\n",
    "    isr,\n",
    "    uorf_targets,\n",
    "    down_strict_XP,\n",
    "    label_up=False,\n",
    "    label_down=False,\n",
    "    label_down_custom=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Plot TEs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "te_plot(\n",
    "    data=data_de_plot,\n",
    "    list_up=isr,\n",
    "    list_down=down_strict_XP,\n",
    "    directory=str(deseq_plot_directory),\n",
    "    plot_all=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Comparing the methods and their outputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"XPRESSpipe vs Original w/ DESeq1\n",
    "\"\"\"\n",
    "# Tm\n",
    "tm_fc_XP = merged_data_split_XP.loc[(merged_data_split_XP['tm_ribo_log2FC'] >= 1) | (merged_data_split_XP['tm_ribo_log2FC'] <= -1)].index.tolist()\n",
    "tm_te_XP = merged_data_XP.loc[merged_data_XP['tm_padj'] < 0.1].index.tolist()\n",
    "tm_common_XP = list(set(tm_fc_XP).intersection(tm_te_XP))\n",
    "\n",
    "venn_plot(\n",
    "    directory=str(deseq_plot_directory),\n",
    "    figure_title='Tm',\n",
    "    gene_list_A=tm_common_XP,\n",
    "    gene_list_B=ingolia_tm,\n",
    "    label_A='This paper',\n",
    "    label_B='Original paper',\n",
    "    color_A='#d95f02',\n",
    "    color_B='#1b9e77',\n",
    "    output_name='tm_deseq2XP_venn.png',\n",
    "    dpi=3600)\n",
    "\n",
    "# Tm + ISRIB\n",
    "tmisrib_fc_XP = merged_data_split_XP.loc[(merged_data_split_XP['tmisrib_ribo_log2FC'] >= 1) | (merged_data_split_XP['tmisrib_ribo_log2FC'] <= -1)].index.tolist()\n",
    "tmisrib_te_XP = merged_data_XP.loc[merged_data_XP['tmisrib_padj'] < 0.1].index.tolist()\n",
    "tmisrib_common_XP = list(set(tmisrib_fc_XP).intersection(tmisrib_te_XP))\n",
    "\n",
    "venn_plot(\n",
    "    directory=str(deseq_plot_directory),\n",
    "    figure_title='Tm + ISRIB',\n",
    "    gene_list_A=tmisrib_common_XP,\n",
    "    gene_list_B=ingolia_tmisrib,\n",
    "    label_A='This paper',\n",
    "    label_B='Original paper',\n",
    "    color_A='#d95f02',\n",
    "    color_B='#1b9e77',\n",
    "    output_name='tm_isrib_deseq2XP_venn.png',\n",
    "    dpi=3600)\n",
    "\n",
    "# ISRIB\n",
    "isrib_fc_XP = merged_data_split_XP.loc[(merged_data_split_XP['isrib_ribo_log2FC'] >= 1) | (merged_data_split_XP['isrib_ribo_log2FC'] <= -1)].index.tolist()\n",
    "isrib_te_XP = merged_data_XP.loc[merged_data_XP['isrib_padj'] < 0.1].index.tolist()\n",
    "isrib_common_XP = list(set(isrib_fc_XP).intersection(isrib_te_XP))\n",
    "\n",
    "venn_plot(\n",
    "    directory=str(deseq_plot_directory),\n",
    "    figure_title='ISRIB',\n",
    "    gene_list_A=isrib_common_XP,\n",
    "    gene_list_B=ingolia_isrib,\n",
    "    label_A='This paper',\n",
    "    label_B='Original paper',\n",
    "    color_A='#d95f02',\n",
    "    color_B='#1b9e77',\n",
    "    output_name='isrib_deseq2XP_venn.png',\n",
    "    dpi=3600)\n",
    "\n",
    "\"\"\"Original data DESeq2 vs Original data DESeq1\n",
    "\"\"\"\n",
    "merged_data_OG2, merged_data_split_OG2, check_data_OG2 = parse_de_data(\n",
    "    directory=dir_og2,\n",
    "    check_file='ingolia_counts_deduplicated.txt')\n",
    "\n",
    "# Tm\n",
    "tm_fc_og2 = merged_data_split_OG2.loc[(merged_data_split_OG2['tm_ribo_log2FC'] >= 1) | (merged_data_split_OG2['tm_ribo_log2FC'] <= -1)].index.tolist()\n",
    "tm_te_og2 = merged_data_OG2.loc[merged_data_OG2['tm_padj'] < 0.1].index.tolist()\n",
    "tm_common_og2 = list(set(tm_fc_og2).intersection(tm_te_og2))\n",
    "\n",
    "venn_plot(\n",
    "    directory=str(deseq_plot_directory),\n",
    "    figure_title='Tm',\n",
    "    gene_list_A=tm_common_og2,\n",
    "    gene_list_B=ingolia_tm,\n",
    "    label_A='DESeq2',\n",
    "    label_B='DESeq1',\n",
    "    color_A='#7570b3',\n",
    "    color_B='#1b9e77',\n",
    "    output_name='tm_deseq2OG_venn.png',\n",
    "    dpi=3600)\n",
    "\n",
    "# Tm + ISRIB\n",
    "tmisrib_fc_og2 = merged_data_split_OG2.loc[(merged_data_split_OG2['tmisrib_ribo_log2FC'] >= 1) | (merged_data_split_OG2['tmisrib_ribo_log2FC'] <= -1)].index.tolist()\n",
    "tmisrib_te_og2 = merged_data_OG2.loc[merged_data_OG2['tmisrib_padj'] < 0.1].index.tolist()\n",
    "tmisrib_common_og2 = list(set(tmisrib_fc_og2).intersection(tmisrib_te_og2))\n",
    "\n",
    "venn_plot(\n",
    "    directory=str(deseq_plot_directory),\n",
    "    figure_title='Tm + ISRIB',\n",
    "    gene_list_A=tmisrib_common_og2,\n",
    "    gene_list_B=ingolia_tmisrib,\n",
    "    label_A='DESeq2',\n",
    "    label_B='DESeq1',\n",
    "    color_A='#7570b3',\n",
    "    color_B='#1b9e77',\n",
    "    output_name='tm_isrib_deseq2OG_venn.png',\n",
    "    dpi=3600)\n",
    "\n",
    "# ISRIB\n",
    "isrib_fc_og2 = merged_data_split_OG2.loc[(merged_data_split_OG2['isrib_ribo_log2FC'] >= 1) | (merged_data_split_OG2['isrib_ribo_log2FC'] <= -1)].index.tolist()\n",
    "isrib_te_og2 = merged_data_OG2.loc[merged_data_OG2['isrib_padj'] < 0.1].index.tolist()\n",
    "isrib_common_og2 = list(set(isrib_fc_og2).intersection(isrib_te_og2))\n",
    "\n",
    "venn_plot(\n",
    "    directory=str(deseq_plot_directory),\n",
    "    figure_title='ISRIB',\n",
    "    gene_list_A=isrib_common_og2,\n",
    "    gene_list_B=ingolia_isrib,\n",
    "    label_A='DESeq2',\n",
    "    label_B='DESeq1',\n",
    "    color_A='#7570b3',\n",
    "    color_B='#1b9e77',\n",
    "    output_name='isrib_deseq2OG_venn.png',\n",
    "    dpi=3600)\n",
    "\n",
    "\"\"\"XPRESSpipe DESeq2 vs Original w/ DESeq2\n",
    "\"\"\"\n",
    "# Tm\n",
    "venn_plot(\n",
    "    directory=str(deseq_plot_directory),\n",
    "    figure_title='Tm',\n",
    "    gene_list_A=tm_common_XP,\n",
    "    gene_list_B=tm_common_og2,\n",
    "    label_A='This paper',\n",
    "    label_B='Original paper',\n",
    "    color_A='#d95f02',\n",
    "    color_B='#7570b3',\n",
    "    output_name='tm_deseq2both_venn.png',\n",
    "    dpi=3600)\n",
    "\n",
    "# Tm + ISRIB\n",
    "venn_plot(\n",
    "    directory=str(deseq_plot_directory),\n",
    "    figure_title='Tm + ISRIB',\n",
    "    gene_list_A=tmisrib_common_XP,\n",
    "    gene_list_B=tmisrib_common_og2,\n",
    "    label_A='This paper',\n",
    "    label_B='Original paper',\n",
    "    color_A='#d95f02',\n",
    "    color_B='#7570b3',\n",
    "    output_name='tm_isrib_deseq2both_venn.png',\n",
    "    dpi=3600)\n",
    "\n",
    "# ISRIB\n",
    "venn_plot(\n",
    "    directory=str(deseq_plot_directory),\n",
    "    figure_title='ISRIB',\n",
    "    gene_list_A=isrib_common_XP,\n",
    "    gene_list_B=isrib_common_og2,\n",
    "    label_A='This paper',\n",
    "    label_B='Original paper',\n",
    "    color_A='#d95f02',\n",
    "    color_B='#7570b3',\n",
    "    output_name='isrib_deseq2both_venn.png',\n",
    "    dpi=3600)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
