{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import xpresstools as xp\n",
    "from scipy import stats\n",
    "import numpy as np\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt \n",
    "import matplotlib.patches as patches\n",
    "\n",
    "from rpy2.robjects.packages import importr\n",
    "from rpy2.robjects.vectors import FloatVector\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get data and clean up formatting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(10941, 16)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv('isrib_raw_counts.txt',sep='\\t', index_col=0)\n",
    "\n",
    "data.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10941 genes before thresholding\n",
      "10941 genes before thresholding\n"
     ]
    }
   ],
   "source": [
    "data_t = data.loc[data.max(axis=1) > 25]\n",
    "\n",
    "print(str(data.shape[0]) + ' genes before thresholding')\n",
    "print(str(data_t.shape[0]) + ' genes before thresholding')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_t.loc['UCP2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_t.sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TPM Normalize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>isrib_a_hek_mrna</th>\n",
       "      <th>isrib_b_hek_mrna</th>\n",
       "      <th>tm_a_hek_mrna</th>\n",
       "      <th>tm_b_hek_mrna</th>\n",
       "      <th>tmisrib_a_hek_mrna</th>\n",
       "      <th>tmisrib_b_hek_mrna</th>\n",
       "      <th>untr_a_hek_mrna</th>\n",
       "      <th>untr_b_hek_mrna</th>\n",
       "      <th>ribo_isrib_a_ribo</th>\n",
       "      <th>ribo_isrib_b_ribo</th>\n",
       "      <th>ribo_tm_a_ribo</th>\n",
       "      <th>ribo_tm_b_ribo</th>\n",
       "      <th>ribo_tmisrib_a_ribo</th>\n",
       "      <th>ribo_tmisrib_b_ribo</th>\n",
       "      <th>ribo_untr_a_ribo</th>\n",
       "      <th>ribo_untr_b_ribo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>LOC388312</th>\n",
       "      <td>3.854140</td>\n",
       "      <td>6.849915</td>\n",
       "      <td>3.273076</td>\n",
       "      <td>3.559601</td>\n",
       "      <td>3.070450</td>\n",
       "      <td>4.780258</td>\n",
       "      <td>6.019118</td>\n",
       "      <td>5.291535</td>\n",
       "      <td>0.648683</td>\n",
       "      <td>0.163487</td>\n",
       "      <td>0.467467</td>\n",
       "      <td>0.324734</td>\n",
       "      <td>2.667065</td>\n",
       "      <td>1.650422</td>\n",
       "      <td>1.225085</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMD11</th>\n",
       "      <td>49.928633</td>\n",
       "      <td>56.968458</td>\n",
       "      <td>58.125309</td>\n",
       "      <td>48.232600</td>\n",
       "      <td>51.799622</td>\n",
       "      <td>56.177904</td>\n",
       "      <td>52.358084</td>\n",
       "      <td>56.144087</td>\n",
       "      <td>37.191183</td>\n",
       "      <td>35.967101</td>\n",
       "      <td>37.163651</td>\n",
       "      <td>30.849745</td>\n",
       "      <td>38.005682</td>\n",
       "      <td>35.559092</td>\n",
       "      <td>37.487601</td>\n",
       "      <td>40.133321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NOC2L</th>\n",
       "      <td>318.229338</td>\n",
       "      <td>304.935375</td>\n",
       "      <td>317.714066</td>\n",
       "      <td>321.343023</td>\n",
       "      <td>323.250110</td>\n",
       "      <td>299.496971</td>\n",
       "      <td>276.219816</td>\n",
       "      <td>307.716233</td>\n",
       "      <td>414.292485</td>\n",
       "      <td>428.008497</td>\n",
       "      <td>462.792640</td>\n",
       "      <td>445.210528</td>\n",
       "      <td>407.894314</td>\n",
       "      <td>386.048713</td>\n",
       "      <td>410.893505</td>\n",
       "      <td>435.004384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLHL17</th>\n",
       "      <td>15.942125</td>\n",
       "      <td>17.010622</td>\n",
       "      <td>16.478243</td>\n",
       "      <td>12.547595</td>\n",
       "      <td>14.271905</td>\n",
       "      <td>15.012379</td>\n",
       "      <td>16.325828</td>\n",
       "      <td>13.991178</td>\n",
       "      <td>9.297796</td>\n",
       "      <td>8.174341</td>\n",
       "      <td>8.180678</td>\n",
       "      <td>9.417291</td>\n",
       "      <td>8.834654</td>\n",
       "      <td>9.602455</td>\n",
       "      <td>8.698103</td>\n",
       "      <td>8.332766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HES4</th>\n",
       "      <td>27.592139</td>\n",
       "      <td>29.112138</td>\n",
       "      <td>27.764711</td>\n",
       "      <td>27.408931</td>\n",
       "      <td>31.329958</td>\n",
       "      <td>33.264272</td>\n",
       "      <td>35.867349</td>\n",
       "      <td>39.462297</td>\n",
       "      <td>26.596021</td>\n",
       "      <td>25.994405</td>\n",
       "      <td>20.334828</td>\n",
       "      <td>18.509847</td>\n",
       "      <td>23.170131</td>\n",
       "      <td>27.306982</td>\n",
       "      <td>26.829361</td>\n",
       "      <td>27.974285</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           isrib_a_hek_mrna  isrib_b_hek_mrna  tm_a_hek_mrna  tm_b_hek_mrna  \\\n",
       "LOC388312          3.854140          6.849915       3.273076       3.559601   \n",
       "SAMD11            49.928633         56.968458      58.125309      48.232600   \n",
       "NOC2L            318.229338        304.935375     317.714066     321.343023   \n",
       "KLHL17            15.942125         17.010622      16.478243      12.547595   \n",
       "HES4              27.592139         29.112138      27.764711      27.408931   \n",
       "\n",
       "           tmisrib_a_hek_mrna  tmisrib_b_hek_mrna  untr_a_hek_mrna  \\\n",
       "LOC388312            3.070450            4.780258         6.019118   \n",
       "SAMD11              51.799622           56.177904        52.358084   \n",
       "NOC2L              323.250110          299.496971       276.219816   \n",
       "KLHL17              14.271905           15.012379        16.325828   \n",
       "HES4                31.329958           33.264272        35.867349   \n",
       "\n",
       "           untr_b_hek_mrna  ribo_isrib_a_ribo  ribo_isrib_b_ribo  \\\n",
       "LOC388312         5.291535           0.648683           0.163487   \n",
       "SAMD11           56.144087          37.191183          35.967101   \n",
       "NOC2L           307.716233         414.292485         428.008497   \n",
       "KLHL17           13.991178           9.297796           8.174341   \n",
       "HES4             39.462297          26.596021          25.994405   \n",
       "\n",
       "           ribo_tm_a_ribo  ribo_tm_b_ribo  ribo_tmisrib_a_ribo  \\\n",
       "LOC388312        0.467467        0.324734             2.667065   \n",
       "SAMD11          37.163651       30.849745            38.005682   \n",
       "NOC2L          462.792640      445.210528           407.894314   \n",
       "KLHL17           8.180678        9.417291             8.834654   \n",
       "HES4            20.334828       18.509847            23.170131   \n",
       "\n",
       "           ribo_tmisrib_b_ribo  ribo_untr_a_ribo  ribo_untr_b_ribo  \n",
       "LOC388312             1.650422          1.225085          0.000000  \n",
       "SAMD11               35.559092         37.487601         40.133321  \n",
       "NOC2L               386.048713        410.893505        435.004384  \n",
       "KLHL17                9.602455          8.698103          8.332766  \n",
       "HES4                 27.306982         26.829361         27.974285  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Create length dictionary\n",
    "len_dict = pd.read_csv('length_dict.txt',sep='\\t', header=None, index_col=0)\n",
    "del len_dict.index.name\n",
    "len_dict.columns = ['length']\n",
    "len_dict.length = len_dict.length / 1e3\n",
    "\n",
    "#Perform RPK calculations\n",
    "\"\"\"data_c = data_t.copy()\n",
    "data_rpk = data_c.div(len_dict.length, axis=0)\n",
    "data_rpk = data_rpk.dropna(axis=0)\"\"\"\n",
    "\n",
    "#Perform per-million calculations\n",
    "data_rpm = xp.rpm(data_t)\n",
    "\n",
    "data_rpm.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "isrib_a_hek_mrna        6.043992\n",
       "isrib_b_hek_mrna       14.841482\n",
       "tm_a_hek_mrna           4.966046\n",
       "tm_b_hek_mrna           4.538492\n",
       "tmisrib_a_hek_mrna      5.515437\n",
       "tmisrib_b_hek_mrna      7.387671\n",
       "untr_a_hek_mrna        11.790876\n",
       "untr_b_hek_mrna         8.699643\n",
       "ribo_isrib_a_ribo       6.703062\n",
       "ribo_isrib_b_ribo       7.029933\n",
       "ribo_tm_a_ribo         30.385375\n",
       "ribo_tm_b_ribo         31.823947\n",
       "ribo_tmisrib_a_ribo    16.002392\n",
       "ribo_tmisrib_b_ribo    11.702992\n",
       "ribo_untr_a_ribo       11.638307\n",
       "ribo_untr_b_ribo        9.183048\n",
       "Name: UCP2, dtype: float64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_rpm.loc['UCP2']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Check replicate and RPF to RNA correlations (Fig 1E)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAQ4AAAEUCAYAAADa/NhYAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi40LCBodHRwOi8vbWF0cGxvdGxpYi5vcmcv7US4rQAAIABJREFUeJztnXlc1NX6x9+AyqosA1qKlokezX0XUcsbJdF2s6wwy9S6lUu2mHm9RaX3es3M8kqLt9QWi257v8q0KA0X3EpTW05quFvKCCqboPD74/udcRgGZAAZluf9evly5jvne77PfIf5zHOe85zzeBUXFyMIguAO3p42QBCEuocIhyAIbiPCIQiC24hwCILgNiIcgiC4jQiHIAhu08jTBtRGlFLFwA7gDFAMBAAngPu11puVUt7AbOAaoAjYCdyrtT5ahWu2BxYDFiAbuFNr/auLdkOB5zA+OyvwoNb6R6XUNOA2h6YRQFOtdTOl1MXAK8BFZt/Paq3fM/sbBvzL7K8I+LvWeoVSyguYCdwK5ADrgIe11vlV7G+4ad8m837mnqO/ScA/gD/M805qrQefw77WwCKgBeBj9veG2d8g4AXTvnxgotZ6s8P9DQFSgbG240qpCOBN074i4G9a63VOn8tfgTe11s2cP7P6iHgcZTNUa91Da91Ta62A/wELzNfGAr2BXlrrrsAujC9zVXgbeFlrfSnwJPCh+eWwo5QKBj4CHtVadwPuB95TSvlqrWeb9vYALsf4Mt1qnvoGsF5r3Qn4CzBVKdXd7O8dYLR53l3A/5RSTc3H1wJ9zdcOA/+sQn83AlcBPYDOGGI8ubz+zNcGYghCD/PfYPN4efa9CCzTWncHrgAWKKUizdeWAlPNc54xr227v/HARqCj02fzIrDa/GxGAe8rpQIczmsPzKUBfZ8azButCkqpRkAb4Jh56CeML+8p8/lmjF8j5/PuVEptdfGvq1O7Vhh/rO8CaK2/BAKBnk5dtgeOa62/Mdv9iuEJRTu1mwt8afYDhsi9bp5zEliJ8UVuDIzXWv9ktvsZ8ALCzXM+0Vpnma99BNxc2f601h8BMVrrAqAp0BzDYyqvPzCEY6RSaotSaoXDvSvPvr9yVuTbAKeBPPO5DxBqPm6K4XXYeAAYDRyyHTA/+2uBV037tmJ4mHHm6wEYYvQwDQgZqpTNSqVUEYbLnw98DowB0Fqn2RoppUKBRAxXuwRa6zcxXNxz0Ro4pLUucjh2AIgEfnA49hsQpJS6Smv9lVKqL8av94UO9nTG+OK0czhvAzBGKfUUhijEA2u01hkYnpSNGcBvWut0pdQG4CGlVBKGYN7pcB23+zPvR6FSaiKGZ3AQ+Li8/pRSgcCvwCyt9Tql1C3Al0qpjuY5Lu2z3Uel1CpgEDBPa20TqbHAJ0qp+UAIcKXNWK21TQwc3gLhgLfTMNT22QAsNP9towEhHkfZDDVd3Wsw3Op1Wusjjg2UUu0wxsNrMNxZnF6vkMdB2Z/DGccnWusTwA3AdKXUjxhflm+BAodmk4EkrfVxh2OjgU4Yf9yLMETQfo5SqpFS6j/ACOAm81pvAe+b/a/F+AIXVLY/h/eQhPGL/zHwQXn9aa1ztNbDbPEEM+6RiTE8Kc8+27UuxxCTq5RSY5RSLTA8h8u01pEYw44PTIEqizI/G6XUeOC01npxOefXS8TjOAda6y1KqYeA15RS67XWe8AepPwfMEdrPbeMcyvqcewDLlBKeWmtbYuHWmH8stkxg7LZ5hfCduwXjBgLSikfjC9qb6f+/YExWuscs93LGF80m8f0AcaQYoDtl1kpFQa8o7X+t/m8v+06leyvO8Yv9xatdbFS6jXOxjhc9qeUugi4XmttG3Zg9ltYnn1KqZuBFVrrk1rro0qpT4BewElgry3oqbX+RCn1AoZobcY1R2zvS2udaR6zfTb/AAKUUluBJoC/+Thea33IZW/1BPE4KoDWOhlIw4jGo5QaiPGLeWdZouFm/weA3ZjBTHNmogjY7tS0GFimlOpjthsBFHLWTe4KZNrEzYGnMQKpKKU6YHgtHymlfIGvgHTgKgd3HqAP8LFSqrE5zv87RgC3sv11A5Y4BBVt3lKZ/WEEeP+plOpnvhaP4f1tPId99wOTzHOCzf6+Ne9TF/MaNrEJwBgCukRrfRr4ArjXPKcbcCmwSmvdT2vdxQy0xgN5ZgC3XosGiMfhDhOBbeaXegrGL99spdRs8/V0rfWNZZ59bm4DXlVKPY4RUxnhMFbfCtxtTgWPNNs1wZhJ+KuDl9Ie2OOi70eBt5RSozEChWO01vuVUrdjfAH9gM0OY/s7zBjKZRhfNm/gE+D5KvT3llIqyjx+GiPAPK68/sz3fguw0Hy/J4AbzQBrefbdZZ5jE9RXtdYfm/3dhzFjVQzkAsPNIWB5jMfwOHdgiPcdTkPBBoeXLKsXBMFdZKgiCILbiHAIguA2IhyCILiNBEeFMjEDiBFmYld19DcdYzalEUa25dMOgV232wmeQzwOoUYwp1JHYOSYdAGGms8r1U7wLOJx1GGUUpcDz2Kkb1+CsR7jLq31L07tgoH9QAet9R/msfUY+RO7MbJeg4CWwFbgVq11vsP5dwE3a62vdX5uTpM+A1yGsQ5kC/CAiynOGzEStmxJXkswMjffq2Q7wYOIx1H36QU8Z66WXQK85dzAzDn4GOMLiFKqE0Yq9grgHuANrXU0EAW0xUizryjTMHIvepsp+ocwthxwpjWGeNlwXO9RmXaCBxHhqPv8qLVebT5eDPRUSllctHsVY00IGIv1lpgJZo8BR5VSU4GXMbyOIDeufy1GZuYWM1HtrxiZlc64+ls74+JYRdsJHkSGKnWf0w6Pvcx/Z8wvsY27tdZrzMVn/YCRGMvVAZIx/g7ew0itbmP24Uix07EmDo99gMm2JfxKqSCMzFFn9uGwihcXa3HcbCd4EBGOuk8PpVQ3rfU24G/AWnOPih4u2r6GsU/FNq31PvPYMIzVoj8qpS4F+lNyaTzAUYw1Hn4YQnWdw2srgIlKqW/M117F2MXrHqc+PgWeVEr912x3F+YeHJVsJ3gQGarUff4A/qWU2o4xTLijnLZvYAjKaw7HpmMsFtuMsafIdxixDke+Mo//Cqym5OK7mRjrY7ZwduOeR5wvrLX+DGPh2kaMbRm/x1w5rJS6z1wtW247ofYga1XqMOasSpLWuounbREaFuJxCILgNuJxCILgNuJxCILgNp6YVfEF+mJsQiPz84JQc/hgTHVvAk6do225eEI4+mJE5gVB8AyDMTbYrjSeEI7DAJmZORQV1Vx8xWIJwmrNrrHruYPYVjnENvfw9vYiNDQQzO9gVfCEcJwBKCoqrlHhsF2ztiK2VQ6xrVJUOUQgwVFBENxGhEMQBLcR4RAEwW1EOARBcBsRDkEQ3EaEQxAEt6mQcCilmimldiilLnY6PkEptep8GCYIQsWwWq0kJc3HarWeu3E1cU7hMAvzrgE6OB2/FKPQryAIHiQ5eSkzZjxBcvLSGrtmRRLA7gEm4LAJrlmVfCHwBGf3sRQEoYY4kbaW/L17aX7bSBISRgHY/68JKrysXim1B7hca71HKTUPYxeodOAprfXlblzzYvM8QRAqwaHPPif9tSWEDxmEeuShynTRFmPXtkrjdsq5UupKoI3W+mFzB6pKYbVm12hKbkREU44ePVlj13MHsa1yNETbji37nIyPPiCoV29CR97l1jW8vb2wWNzZwL6cvipxTgLQ2dxF+zWgj1LKeXNbQRCqkeLiYjI++YiMjz6gaf8BXHjveLwaeW6vcbevrLUea3tsehxPaa1vrU6jBEE4S3FxMRnvvUvm1ytoNngILe64Cy9vz2ZSSHkEQajFFBcVceTttzj+3UpC/hJLxG0jPS4a4IZwaK0vdnFsFXB59ZkjCIKN4jNn+PONxZxYt5bQuHjCbxqBl5dzrSzPIB6HINRCik+f5vBr/yV780YsN9xI2LXX1xrRABEOQah1FBUWcHjhy+Rs3UL4iFsJG3a1p00qhQiHINQiik6d4tBLC8j9aQfNb7+DkKFXeNokl4hwCEItoSg/j4P/eYG8nb/RYsw4gmMGe9qkMhHhEIRawJmcHA7Of478PXu44J57adZvgKdNKhcRDkHwMKdPnuDgvLkUHD5Ey/snENSzt6dNOiciHILgQU5nZXFg3hwKjx6l5cQHCOzSzdMmVQgRDkHwEIVWKweem8Pp41m0mvwwAR07edqkCiPCIQgeoODIEQ489wxFublEPvwo/u2iPG2SW4hwCEINU3D4EPufm0Px6dNETnkMv4su9rRJbiPCIQg1yKn9+zkwbw54edH60Wn4tor0tEmVQoRDEGqI/PTfOfD8c3j7+hL5yFSaXHCBp02qNCIcglAD5O38jYPz5+ET1JTIR6bSOCLC0yZVCREOQTjP5Pz8E4eS5tMoNMwQjbAwT5tUZUQ4BOE8kr1tK4dfSqJxiwuIfPhRGgUHe9qkakGEQxDOEye/38Th/76Cb2RrIh+agk9Q9ez3WRsQ4RCE88CRVakcXvgyfm0vodXkh/EJCPC0SdWKCIcgVDNZqas48tYb+KuOtJo4GW8/P0+bVO2IcAhCNZKZ8jVH332b0N49sYy7H+8mTTxt0nnB87ueCkI94diyzzn67tsE9exNx78/Vm9FA8TjEIQqU1xcjPXTjzn2+f/RtP8ALhh7D96NGwP5njbtvCHCIQhVoETNk0FDaHGn52ue1AQVFg6lVDNgHXCtWT/2b8ADQDGwGbhXa11wfswUhNpHcVERR95ZyvFV37I3PIKe11zXIEQDKhjjUEr1B9YAHcznHYBHgYFAN7OfCefJRkGodRQXFfHn64s4vupb0ps35/bFC3n3f+942qwao6Iexz0YwvCW+fwUcL/W+gSAUmo70Kb6zROE2kfx6dP8sei/nNxk1jyJjiGxZSsSEkZ52rQaw6u4uOIV45VSe4DLtdZ7HI5FAJuAu8zKbufiYiDdDRsFoUbJyMhgyZIljBkzhvDw8BKvFRUWouc8x7GNm7ho9B1EDv+rh6ysEm2BPVXpoErBUaVUK+BLYFEFRcOO1ZpNUVHFRauqREQ05ejRkzV2PXcQ2yrH+bItKekVZsx4guzsU0ycONl+POPwYX6bM4vwkycJvOFGXv1xBwmX9sBisdSYbVXB29sLi6V60t4rLRxKqY7AcmCB1vq5arFGEGoBtiGH49Aj49BBfvjHNNp4e7OjzUUc2ruHGTOeACghLg2FSgmHUqop8BUwXWu9tHpNEgTPYrFYSojBmZwcfp/9LyK9vHgvNwc/fz+Gx8UDNKi4hiOV9TjuBloAU5RSU8xj/6e1TqweswShdnDm5En2PPtvmubm8k5eLj6qI3PnziYgILBBeho23BIOrfXF5sPnzX+CUG85mp7O3mf/jf+pU/x9wzrWH/mTKT17kZg4s8F6GjYkc1QQXFBotXLg2dk0ys3l4Q3r2GLNoG3bdgwfPoKoqPaeNs/jNIw0N0E4B1arlaSk+VitVgqOHGH/nFkEeXmxuW1bom8dyZAhQ0lP383y5cs8bWqtQDwOQQCSk5cyY8YTBOTnM/jQQU5lZ7M2MpKbJ0zGYrFgtVpJTl7a4IcoNkQ4BAFjdqRR5jH67PyNPOCeb74i/eQJCsOMGRbnmZaGjgiHIACn9+9jQHo6J06dIsVigfAIxo64TTyMMhDhEBo8eTt3kvHSArLy8pictpp8Xz8yM4/Rrl2Uy6xQQYRDaOD8sWE9mYv+S66XFxPWpuLfvDmH9+0lKqoDM2bM8rR5tRYRDqHBkr3tRzJfW0j68SzeKy7maH4e46+9gfDwCBISRom3UQ4iHEKDwmq1smjRQiwZGcQcOwYRESRnZdLyknZM6duPcePuFcGoAJLHIdRbHHMzbCQnL2Xbu28z0GplxzEr/9z5Gyu++5YlS14lICBQRKOCiHAIdRKr1cqzzz5bQhScseVmJCefXYd5Y7sonujVl59PHOeRtDU0DQ9nypRpTJkyTWZQ3ECGKkKdxCYKjntmOCZpWSyWUsvjd777DsUpX3HY15fJq1dRUFTE5s0b6dKlm8Q03ESEQ6iTJCSMIijIl+uuG2E/ZhMT2+uOInLsyy8oTvmK7w4f5HWrlZtvG0Va2lqGDr2iQe+rUVlEOIQ6icViYcyYMSxY8EoJDyM3N4fc3BwWLVrI3LmzobiYG8IjyPvma9ZYM1h4+BD7DuwnYdRoXnghCavVSmRkaxmmuInEOIQ6y5IlS0rEMCwWCwEBgcydO5uNGzcw/v5JtN23l7xvvuazvXuYvjYVn8ZNSiyLt6WSyzDFPUQ4BI/javajIufk5OSUCmomJIwiNnYYq1NXErJ5E+0zM/ktMJCvvIppFhLKrFlziIuLZ9Kk+9i1a+f5eDsNAhEOweO4mv2oyDlPP/10qSlUi8XCf+a/xKwhQ4m/sCUfHtjPGj9ftmzdQlZWJr/88jOJidNJSVlBYuL08/F2GgQS4xA8jqvNgStyTlCQL4MGXUFS0nz7uS8ueJ4OO39jcEgoS/f8zivbttI28xgAQ4YMJSFhFHHmfqGSUl553KqrUk1cDKRLeYSziG2VIyKiKU8++U9mzHiCKVOmsX3rFi47fpwhF7bkxZ+2k7x7J/7+/uTl5dG2bTvefvu9Gtu9qzbeN4fyCFWuqyJDFaFOk5AwisTEmfgUFXFVdjZDLmzJvG1b+ezPP2jVqjV5eXkAsntXNSNDFaFWY1tbArhcR2KxWLht+M389u9/EhbRnH9v+Z4v9u8FoFOnS2nTpg0dO15aIiFMqDpVqVYfC8wD/IH/aa0fP082Cg2Y5OSlRj6GydSpZwOaWmseuHc8d/j50bpxE2Zs+Z6Ug/vtr2/atIHExJmS2HUeqGy1en9gMXAD0Anoq5S6+nwZKTQ8bFO0cXHxREfHAJCbm4vVauWppx7nhhuu5q9XDePW4mIiGzchcfNGUg7ux8vLi86duwJng6FC9VPZavX9gJ1a63QApdRSYARGHVlBqDKO6eMxMYNJS1tLQEAAyclLeeml/xDm68sL0YNoFRjE3zemsf7InwAUFxdjsRgL12SJ/PmjQsKhtb4bQCllO9QSOOzQ5DAQ6c6Fq6v4rTtERDSt8WtWlIZmW3kV4QEmTbqPoCBfxowZA0Dz5mH2x6lffMK9wWFY/Px5dMM6fsg4CkDfvn0JDg4mJSWFa6+9mo4dL652u92hNn+mVaWywVEvF8eK3OlApmPP0hBtK6si/Fl8ueuu+8jIOLvitbjYl4KjR3gg/AICvL15OG0NOzKP0bhxYwoLC+natQePPjqd5OSlXHfdCI/e09r4mdaGavUHgQscnl8IHKq6OUJDoSJJX1arlXvvHUtq6kpWrfqW6feO50zy2/h5eTF53Wr08Sx7ngbArl27pIxBDVFZ4dgAKKVUFJAOjMQIlgpChXD1BXcuemQTDYCDW7/n9Fuvc6aomIfS1vD7yRMA5OfnAxAaGsbs2XNr8B00bColHFrrfKXUXcCHgB+wDPigGu0SGhA2wcjNzbFPvebm5thFo2NIKPMGxJB/5gyT161mf062/dzi4mKiojrw5pvJUtO1BqlstXq01t8A3avbIKHhYZtBmTJlGomJM4mLi+eRRwxvpFuYhWf7D+R4QQGT01ZzODe3xLlt27bjs89WyOxJDSOZo4LHsW3AY4tVfPTR+6SlraF3eASz+0WTceoUD6xN5Wh+XonzAgICePvt90Q0PIAIh+BxHDfgAVCqE0MiW/Nk914cyMnmkQ3ryMjPIzAwCG9vb06a8Y0xY8bI8MRDyCI3oVYwYEA0fn5+ALQ4cZyne/Rmz8kTTFq7moy8PLy9vcnJySY0NAwwskKfeuopD1rcsBGPQ6h2nHcbL6/dggXPs2XL9xw8eJD8/HyubNWax3v25uesTB5dv47s04UAFBUZaUKxsVcSGdmGhIRRhIeH17pciYaCCIdQ7Timi7vKqdi1ayfTpk2hsLCAtLS19uN/vaQdD3fuxo7jRs2TvDNnAMO7iIpqz65dO7n77vtkeFILkKGKUCnK2yfUtkeGq+Quq9XKnXcmkJq6krS0tVx4YUt8ff24qe0lTOnSnY1H/uTBNavsohEV1YGFCxcTGdma1NSVsqdGLUE8DsFtrFYrkybdR0rKCqCkV+FqmGKIzAts2JDGkSN/sm/fXkJDw2jT5iJ+/HELI6PaM/7Srnx3+CAztnxPgTksiYxszZtvJrssriR4FhEOwW2Sk5eSkrKC2Nhhpb7Itnomubk59r0zkpOX8uKL8+1tvLy8yMw8RnZ2NmNVJ8aqTqQc2M/MLZsJahaMH3D8eBbXX3+jfVgiqeS1CxEOwW0cf/2dg5+2XAzb/wBxcfF8+eVn/PzzT+Tk5GDb5/ae9h0YGdWBz/fuYc6PP1AEnDx5wh4IDQgIqIF3I1QGiXEIbuNcxMgx3uHv7w/A9u3b7PGP5cuXsWnTRnJycgBjafXDXbszMqoDH6bv5hlTNMCYPQkNDWPs2L8xbty9Nf3WhAoiHodQZRxnUcaNu5etW7eQkrLCvlfosWPHaNmyFYcOHcQbeKxHb65pcxFv7/qNl3/eAUBYWDj5+Xk0auRDZuYxIiNbS0ZoLUaEQ6gycXHxrFr1Lfv372PRooXMmDGLgQMHsX//PpYsedXezsfLiyd69iE2sjWL9S8s1r/YXzt2LMP+OCqqgwRBazkiHEKlcJw9Wb58GampK+2rWcHYVPiGG+Ltzxt7e/N0734lap44Eh0dQ8eOl7J79y5mz54r3kYtR4RDqBS24cm6dWuYMWMWubk5rFr1LZs3b2Tt2tVs3ryR33/fBYCvjw+z+g6gf/MWzNu2lY/2/I6Xl5c9SDpkyFAWLlwsYlGHEOEQyiUjI4OkpFdKzaAkJIxi1apvSUlZQU5ONj179sbLy9hRMi1tLcOHX0t+fj7+Po14pn80PSzhJWqeFBcXExwcwqhRo5k48UERjTqGCIdQLkuWLHGZPm6xWOjXr789A9QxdRyMnbmCGjXm2QED6RQSyswfNvH1wQMA+Pn5kZ+fz/HjWVgs4SIadRARDqFcxowZQ3b2qVLBys2bN/Lee+/SpUs3CgsLOXToIE2aNLZPwQY3acK8AYO4pFkzEr/fSOrhs1vShoSEMmrUaEAyQesqIhxCmatZrVYrn332PnFx8aVef+CB8ezbt8dlfxZfP56PjilV88RGZGTrEhXZhLqHCIdQ5mpW2/HY2GGkpKywB0KXL1/Go4/+nccee4jw8OYcO2bl+PEszpw5Q3M/f+YPHFSq5glA585dsVjCZVPheoAIh1BqAZnNA4mLiycoyJdBg66goKCAlJQVFBQUkJq6ktjYYWRlZZGVlWXvp2VAAPMHDqZp4yb2mic2rrnmejp1ulSqq9UTJOVcKJVCbvM0DM/iUaKi2tOvX38Adu7UtGlzMSNG3EajRmd/d9oEBfFizGUENGrE5HWrS4jG+PEP0Lt3X+bOnU1y8tKafXPCeaFKHodSahTwd/Ppl1rrKVU3SfA0rpawjxt3Lx9++D7p6bsBuP/+cfbFaJc0bcYLAwcBMGntanvNEzASuyZNeqhU3xXdJUyonVTa41BKBQD/AS7DKJMwWCkVW12GCZ7D1SK2RYsWEhwcbG9jEw0VHMKCmMGcLipmwppUu2j4+PgwZsw9LF68FIvFUqZXIx5I3aQqHocPhvAEAjlAYyCv3DOEWo8rTyA5eal9B3JHuoSGMXdADCcKCngwbTWHHGqetGoVyTPPPFfmdWRjnrqNly3ttzIopSYBczAEYxVwk9b6XB1ejFE2UqgFOFaNBxg9ejTLlhmxjZ9++ol58+aRmZlJTEyM3csA6BUewTP9ojman8eD69ZwxKHmiY+PD2vWrGHAgAE1/n6ECtEW2FOVDirtcSilugFjgYuA48BSYArwbEXOl2r1Z/GkbY5V43Nzc1i2bBkhISG89dZS/vjjMOvWraOoqLiEaAxo3oJ/9R3AgZxsHkpbw7FTp0r02a/fANq163ze35N8pu5RG6rVAwwDvtFaHwFQSr0OjKeCwiHUDmxDhbi4eKZNM2LbxhRrlsPjswy5sCVP9+7H7yeO8/D6tRwvKLC/1qVLN8LCLJKn0QCoynTsj0CsUipQKeUFXAdsqh6zhJrCFrT86KP3SU1dSdOmzcpse2WrSGb07sevWZlMXremhGg0bdqUHTu2cfnlf5HyBQ2ASnscWuuvlFI9ge+BQmAjUDqCJtQ6HBO8li9fRlxcPGvXrgbg2mtvYM+edHbs2GYvtQhwTZuLeKx7L7ZaM3hsQxp5Z07bX/Pz8+N///uY9evTJNjZQKhScLSSXAykS4zjLDVt25w5s5g7dzZt2lzEvn17iY4eRFraGgBatmxFUVERf/xx2N7+praX8FDXHqz/8w+mb1pvL18AEBgYyPvvf0qfPv1qzH4b8pm6h0OMo8rBUckcbQCUVTxp3z5jb4z95h4ZAIcOHSwhGiOj2vNQ1x6kHj7E351EA6B3734eEQ3Bs8halXqMbUiSm5tjr3USEBDI8OEj2Lhxg32rvwMH9hMSElIqEOpc8+SMk3caEdGcrl27YbVaJfuzgSEeRz3GcdVrYuJMAPsalNmz5xISEgJA48aNS4nG/Zd2YazqxOd79zDjh02lRKNPn36MHj2WF1+cL9mfDRDxOOopVquV3NwcpkyZZl+RumvXTjZu3MD+/ft45JEH7GJRWFhoP88LeKhrd4a3bceH6bt5YfuPOEpGYGAgjzwyzR4EDQgIlIBoA0SEox7hmC5uSxNPTJxpH0YsXfpGqd3IAby9vSkqKjJrnvTimjYX886u33jJrHniiHMgVMoyNkxEOOoRjkMT57Ugu3btLHNIUVRUhI+XF4/37MOVLmqe2OjRo5cEQgVAhKNe4SgWmZnHWLduDXFx8VgsFqZNm0Jm5jEaN25CYWFBifMca5689PN23tm1s1TfUVEdeOmlV0sdFxomEhytw5Q1zQqQmDidlJQV3H77LcyZM4uoqCgAzjgkbgE08fbm330HMOTCljy/fatL0QgMDOTNN5NZvnyZy2sJDQ8RjjqM854WRvnFJ7j33rGMG/c3LJZw0tN3M3fubL788gs6dry80nzIAAAcF0lEQVS0xGI1fx8fnu0/kH7NWzB76w98mP67y+tcemkXli9fJvtnCHZkqFKHcY5j5Jr7YaSmrqSwsACr9Ww91sOHD5VI7Apq1Jg5AwZyqVPNE0f8/QOIiRnIjBnPEBoaVuJaQsNGPI46wq5dO7n55ht4+ukn7MMFx121rFYrO3Zst7d3fGzDtrwguEkT5g8cRKeQUBK/31hKNHx9fQFISLidr7/+mqio9qV28BIaNuJx1HJsU6yrVn1rn0rV+ld7mQLbTl2LFi0kNXUlffr0Y/fuXWQ6bBbsSJivLy9EDyqz5gkYgdCbbrpFvAuhTEQ4ajm2OMb48Q8AUFholCkA7P875lKUJxrl1TxxpHv3HpKfIZSLCEctxzmOsWjRQmJiBjN8+AgGDhxEQsIoM0s0l5CQkDJFo2VAAC9ED6ZZkyY8sn4N24+VbhcU1JTs7JMcOFA63iEIjkiMo5bjGFuwZYMGBATag5Xp6bu57rphvPTSf0qtN7HROjCIpJjLCGxs1DxxJRoA2dkniYrqIDt4CedEPI46hKP3YRvCXHhhSw47FHR2pryaJzb8/f3Jy8sjOjrGXs5AEMpDhKOO4Fy2oFOnSwkODi5XNFRwCPOiYzh1pojJ61azPye7VJvu3Xvy8suvlQi0CsK5EOGohTiLhNVqZdKk+0hJWcHrr79Gq1aR7NmTzvHjx8vso7yaJ44EB4cQGhomwVDBLSTGUQtxzAh1FA1/f3/27dtLWtpajh49Uub5PS3hzIseROapfCauTS1TNEJCQkhNXSnZoILbiHDUQuLi4omNHUZcXDzJyUtJSVlB27btyMszih75+Phw+vRpl+f2j2jB3AExHM7NYcLa1BKFkpzJysoiNnaY5GsIbiNDFQ9jtVp58cXnyMk5Zd9wZ/nyZaSkrHCYbs3ggw/etZ9z5swZl30NuaAlT/dxXfMEwMvLi+LiYq699gY6duwEYL+mILiDCIeHMYYlTwPGbloTJ04ulbuxbNnn/Pln6QxPR65sFck/evbh16xMpqxfR/bpwlJt4uOvp3fvPhIEFapMlYRDKXUd8BRG4ekVWmuJsLlJQsIovLxOk5Nziri4eJKS5pOQMMoerHzqqcdJT99dbh/l1Txx5Oefd7BkyVvV/h6EhkelYxxKqUuAV4AbgK5AL6XU1dVlWEPBYrHw1FNPMXXq9FJL13ft2slbb71e7vnDL76Ev/fozaajR3h0wzqXotGixYUEBwczYMBA2U9DqBaqEhy9Efif1vqA1roQuBXYUD1mNTxsmwuPH/8ABw7sIz4+lri4v5SopuZMQrv2PNzNqHkybWMap8qIfQQEBHD8+HGSk9+SGRShWqjKUCUKKFBKrQAuAD4DnqjoydVVNdsdIiKa1vg1yyMjI4OkpCT787lzZxMaGkpmZuY5zx3ToSPjOl5aZs0TgCZNmlBQUEB8fBzh4eEATJx4H+Hh7t2H2nbfHBHbPENVhKMRMAS4HMgGPgVGA69X5GQpAQlJSa/YA6NTpkwjKqoDu3b9Zn/dNgvizP2XduH2qA58sW8vz2z9nqJSLQwuuSSKW25JKBEMLS7GrftQG++bDbHNPRxKQFaZqgjHH0CK1voogFLqE6AfFRSOhopjVmhCwigyMo6yY8dWDh06iJ+fX4nNhJ1Fwwt4sGt3bmrbjo/Tf2fe9q2UJ70DBw6SjFDhvFAV4fgceEMpFQKcBK4GPqkWq+oZzvVOZsx4gnXr1rBgwSsEBASQmpp6zj68gande3HtRWXXPLHRrFkIJ05kERYWVo3vQhDOUmnh0FpvUErNAdYAjYGvgSXVZVh9wLl2KxjTr+vWrSElZQWLFi1k40YjnmxboepqeFKRmic2oqNjeOKJp1m/Pk0yQoXzRpXyOLTWi4HF1WRLvcPmXUyZMo3ExJn2WMOIEbexdm0qy5cvY8eObTRp0sSeTu4sGo41T17+eQdvO8RAnFm4cAldu3YjMXE6M2bMkiQv4bwhmaPVhPOKVnC9e5fx/3/Jy8tjx45tNG3arMwpV18fH/7Vpz8DWlzA89u3llm+AIz1K0OGXG5fEAfwzjvvV9v7EwRHRDiqCcfyi7aApMViscc1rNYMXnxxPgBdunQjJyeHgoJTZYqGv48Pz/QfSA9LOP/e+j1f7Ntb5rW9vLw4c+aMacMsAB5++FF7Fqp4HkJ1I8JRTTh7FzYWLVrI3Lmzad68hf3Yjh3biIxszYED+132FdSoMc8OGEincmqeADRu3Jg77riLm2++lXnzniUuLp6oqPa88877JCXNLyVkglBdiHBUE7a9QW3Yhi7HzP09j5hlCGzBzxMnXG/C06xxE+ZFx9CuWTCJ328ktYwdvho1asSnn35Jnz79SEqab19Na7OhLCEThOpAhKOK2AQiLi6e5cuX2f+3zaQMGTIUOCsYtuDniROlhyhhvr48Hz2IyHJqntg4ffo069en0adPP5ci4SxkglCdiHBUEce8jJSUFfb/hwwZypQp0xg+fASxsYPt5RnLIsKseRLh58/UDev4voyaJwAXXHAhHTp0JC4uHhCREGoeEY5KYLVa7TMkw4ePIDc3h9zcXHr06Mlf/hLL7t27SE1dyf79+/jLX2KJiRlMSspXLtPHAS4MCGC+WfPk4TJqngBERDTn6NEjtG17CampK1m+fJkIhuARRDgqga2+CRib7wQEBDJ37mwSE2fywQfv2ffPSE/fza23Di93hWvrwCBeGDgIfx+j5ok+7ro2CkDLlq0YPXosw4ePsO9KLgieQISjEiQkjCI3N8f+ODPzGOvWrSEuLp7bb78FMGY8CgsLyxWNEjVP1qWy20Xcw5Eff9zCDTcMJyqqvXgagkcR4agEFouFqVOn25/bNhQeOHAQs2bNYdy4O84Z07DVPCk4U8TktNXsyy5d88RGeHgEAQGBXHHFleJlCLUCEY5KYotz2ARiypRpJCSMcjjmBWWsXbXVPDlZWMDkdWs4ZHovrggODuH6629k8eL/0rp1G0nmEmoFIhxu4LzK1RbnAEhMnAlgX7RWlmj0tITzTP+BWPPzmLxuTbnlC3x9fTl+PAt/f3/7WhdBqA2IcLiBY1p5QsIoVq36ltTUlQQHh7Bjx3ZiYwdzsIwsTzBqnszqN4BDOTk8mLYG66n8cq936tQpwFj45hzTcLU2RhBqChEON3BMtLJYLEyb9g82bkzj+PEsPvrovXLPtdU8ST95gofS1pSqeeJIYGAgOTk5hISEkpWVSUBAQKk2rtbGCEJNIcLhBrZEK6vVSlLSfL7+ejn5+We9BtueGs7Etork8Z590FlZTNmwlpOFpWueONK+vSI8PIKUlBXExg5j3Lh7S7WRlHLBk4hwOOGq4LMt2ctW9cz2a9+iRYsS57oSjWtaX8RjPXrxozWDqeXUPHEkJmYwEyc+WO5QRLJFBU8iwuGE8xDAOdnLlsMxZMhQUlNXltvX8Isv4eFuPdhw5E+mb1pfZvkCGwEBgYwfP8kuUCIMQm1FhMMJ5yFAXFw8X3+9HPBiwIBo+0Y5ffv2L3MXcjBqnkzo3JXVhw+R+P1GCovK2ovcmD05deoUnTt3kVquQp1AqtU7Yfulz8w8xsiRI1i69A3S0taSlraGefOeJSVlBd7e3mzatKFM0RjToSMTOncl5eB+Ht+8oVzRAGjVqhVDhgxl06YNUjBJqBOIx+GAYzxj48YNpKauJCcnmyFDhtK1azeuueY6Vq36htOny45TVLTmCYBSnSgoKOD333cTGXmRPYlMEGo74nE4YItnzJ07m3btooiNHUbHjpeSmrqS4uJiEhP/UaZoeAEPde3O7VEd+Dj9d2afQzTAqHuybFkK8fHxpKauJCAgUIYpQp2gWoRDKfWsUur16ujLkyQkjLJvvGOxWFiw4BW2b/8RgDffXMzmza5L43oDj3XvxU1t25G86zeeO0ehJBuffPIhmZnHeOONNyQzVKhTVHmoopS6ArgL+KLK1ngIxyHK7Nlz7Tt5jR07is2bNwKQXcYiNHdqnoBRADou7hpWrvyGzMxjJCZO5+uvl8sMilCnqJJwKKXCgH8Bs4Du1WKRB3BedzJ16nSSkuaTlra23PMae3vzVO++XHZhK175eQdLy6l5YmP8+AeYOnU6u3bttNc/EYS6hldZMwMVQSn1PvAK0Bq4XGt9VwVOuxhIr/RFq4mMjAyWLFnCmDFjALj55pv57rvvmDBhAuHh4WRkZPDJJ59w8OBBvL29KXKaGWni7c2/+g4gugI1TwCaNWvGHXfcwVNPPWWvHC8IHqItsKcqHVTa41BK3Q3s11p/o5S6y93za7pavZfXKRYseMWeiWlUin+CI0eOERAQSOfO3fnuu+/Yvv3nUoldzqLh7+PD7H7R9AyPYPbWH/h83x6X17QJTt++/XnzzXexWCwuq8XXxsrmNsS2ylEbbavOavVVCY7eClyllNoKzACuV0o9Xy1WVTNWq5XRo0czY8YT9jyJhIRR9qXwM2Y8QUBAAOPHP0Bubg4REc3L7CuwUSOeix5Ed0s4M3/YXKZowFnBueyyoTJbItQrqlJ0+krbY9PjuFxr/VB1GFXdJCcvZdmyZQwZMpTc3BysVqu9ytqCBc/Tp08/Vq36liNH/mRfORXTHGuePPn9Rr4ro+aJI0OGDHW5SE0Q6jINIgEsIWEUQUG+HDlyjLlzZ9vXnIwdO6pUALSsNPJQX19eqGDNE4DOnbty9dXXSAq5UC+pljwOrfXrFQyMegSLxWIPgtqyM5OSXighGrY9L1yJRoSfPy/GDKFlQCBTN6w7p2gAXH31NUydOl1EQ6iX1MvMUdt+GVar1f44KSnJ7m1YLBa2b98GGAFMoMzNhS8MCODFmCFYfP14ZP3acgsl2WjT5iKGDx9RfW9IEGoZ9VI4bEvjk5OXllgmn5g4kwEDorn55hto1SqSNm0uKjVj4kjrwCCSYoYQ1Lgxk9PWsO2YtULX37dvL8uXL6uW9yIItZF6GeNwXhpvq7RWXOzDzJlP2ocoYWFlDyPaNm3GC9GD8PaqWM0TWx2V7t17cuWVwyR9XKjX1EvhcN4EJyAg0O51REa2th8/VoYH0SE4hHkDYjhdXMTEtWvYm33u+fhCczvA4OCQEjVXBKE+Uq+GKo6xDcdjubk5PProowwZMpQDB/aX20fn0DDmDxxE3pnTjF+TWiHRsNG2bTtmz55bafsFoa5QrzwO5/IFyclLycg4yksv/YcJEyYAEBoaRmam66LOZ2ue5PNg2mr+dLGHqDNt2lxEbOwwwsLCZOpVaDDUK+FwjG3YRKRPn/4AvP3222RlZeHl5eXyXHdrnoCR3LVw4WIRC6HBUa+EwxHbpsLvvvs2AFlZWTRt2pSTJ0sPPQZfcCEz+vQn/eQJHk5bQ1Y5NU8AunTpRlxcvHgYQoOlXgmHzcvIzc3h2DEr77//vxLV4l3tqXFFy0ie6FXxmidgbGAsAVChIVNvhGPXrp2sWvUt48c/gNVqZcmSV0u1cc4KjW99EdPcqHkSFBTEiBEJsvZEaPDUm1mVxMTppKau5NNPP+btt984Z/sbL76E6T17s+noEaZsWFeuaPj5+QGGx7J69Xekp+8uNXsjCA2JOiMcZU212o7NmDGLtm3bcfDgfgrOEaNIaNeeR7r1YPXhQ0zbmHbOQkn5+fm0bduONm0uYteu37jjjttKLNEXhIZGnREOxzRyV8dCQ8No1qzZOfu5y82aJwCxscNYtiyFd9/9iKioDlitGcTGSnao0HCpMzEO2yxJRsZR5syZxV/+EsuqVd8yZsw95ObmsGjRQn78cUu5fdzXqTOj2qsK1TwBCApqyp13jmHSpIewWCxYLBY++2xFuTVdBaEhUGeEw2KxEBAQaN9U+LnnnqG4uJhNm9aTl5fHhAmTadeuPbt37yx1rhfwQJdujLgkio/3/M68bRUrXzB69FiefHJmKTtkR3KhoVNnhipQsu6JbYYkLy+Ptm3b8cMPm12KhjfwaPeejLgkind37+S5CopGdPQgJk58sBqtF4T6Q50SDovFwsKFixk58g77TEezZsEUFBS4LGXg4+XF4736cP1FbXld/0rST9vPeY2AgEB69OhFz569qt1+Qagv1JmhChizKMnJS5k48UFCQ8P47LNP2bdvDydOHC/VtpGXF0/17sflLStW86RRo0b4+wdw8uQJtm79ga1bfyA8PEKGJYLggjolHLZZlK+/XkFa2poy2zXx9uaffQcwsMUFvLD9Rz5I333Ovk+fPs3JkycICQkhKyuL6OgYmTURhDKoU0OVuLh4YmOHUVhYdp6Gv48Pc/oPZEDzFjyz9Ydziobjorc2bS5m+PBbAIiJGSyzJoJQBnXK41i69A1SUlYwfPgtbNnyPWecErcCGzXi2f4D6Rxm4Z9bNvPVOfbegLNBVttKVzA2+xFvQxDKpk4Jx6ZN6wHYuDGtlGg0bdyYeQMG0T44mCc3b2BVBWqe2IiOjimxPF7iGoJQPlUtOv0kcIv59Aut9dSqm1Q2ubnGxjpBQU257LKhfPedUaox1NeX5wfE0DqoKX/fuJ60I39UqL/u3Xtx5ZVXyfJ4QXCTqtSOjQWuAnoCxcBypdSNWuuPq8Mw2wyKY4ambY/Q3377lX1m6cUIP39eiB5Ec39/pm5YV6HyBQCtWrXm5ZdfJSqqfXWYKwgNiqoERw8Dj2itC7TWhcAvQJvqMcv12pRWrSIBoyZrbm4uF/gbNU/C/Spe88TGwYP7pYSBIFSSqtSO/cn2WCnVHqMI9cCKnn+uqtmTJt1HUJAvY8aMITy8KRkZGfTv35fvv99EcXExrQODeGHgIPx9GjE5bQ2/ZmVW6LqNGzemb9++xMTEMGnSfYSHN62oyeeViIjaYYcrxLbKUZttqypVDo4qpToDXwBTtNalc77LwGrNpqiovORvX+666z6Ki+Ho0ZMkJb3Ciy++CLhf88SRwsJCYmOvZuLEyfa+PU1ERNNaYYcrxLbKURtt8/b2OucPdoX7qsrJSqkY4Btgmtb63LvnVIGEhFH06dOPDsEhJMUMoai4mAlrKy4aUVFR+Pn5MXz4LTLVKghVpNLCoZRqDXwCjNRav1t9JrnGYrGgmjaz1zyZsDaVfS72EHWkT5/+9jUtmZmZ5Ofnc+LEcZlBEYQqUpWhyhTAD5inlLIde0Vr/UqVrXJB7q+/cH+YhT+ys5mw5rsSNU+aNPGloOBUifZXXhlHZGRrNm/eQHBwMMOGDePAgcPMmDHrfJgnCA2KqgRHJwPnNVPKNiV7U68+ZL/1Ok0iIug0dTqXPDiePx1Ww7oqlRIdHcOqVd8CcPz4cd555x1iY4cRGhp2Pk0WhAZBrV6rkpy8lJX/fYnjS17DOyKCFWFh7LUeZf/+fSXanTp11tuIiGhO37792b9/n70i/ciRdxIbG0tKygrZJ1QQqoFanXI+vFNnYvtF0ySyNV8FB5M462miojqUqP/aqFEjTp8+u0P50aNHOHr0CJs2bbAfi4pqz/z5z7FgwSsSGBWEaqDWCkeu/pXsd98msIOi1QMPMiInl9ONGjFgQDQzZz7JyZMn2bt3DydPnqBLl24cOLCPCy5oya+//mzvIzo6hpiYwSQkjCI8PFzWoAhCNVFrhyo+gYGEXhVHq8kP4+3nb9/rs0+ffnz66ZfExcXbq7TFxcWTlraF5s1bAIZgTJkyjcWLlzJ16nSZRRGEaqbWehy+ka2JGHHrOdu1bduO4cNHkJy8lNTUlcTGDmPBgldELAThPFJrheNcDB8+gk8++Yhdu35j+fJlJSrVi2gIwvmlzgrH8uXL2LXrN3thJClbIAg1R50VDvEwBMFz1FnhEA9DEDxHrZ1VEQSh9iLCIQiC24hwCILgNiIcgiC4jQiHIAhuI8IhCILbeGI61geM/Q9rGk9cs6KIbZVDbKs4Dvb4VLUvL1sJxBpkELC6pi8qCIKdwUDZVdsrgCeEwxfoi1GX5cw52gqCUH34ABcCm4BT52hbLp4QDkEQ6jgSHBUEwW1EOARBcBsRDkEQ3EaEQxAEtxHhEATBbUQ4BEFwGxEOQRDcRoRDEAS3qbNbB5aFUupJ4Bbz6Rda66lOrycC44BM89CrWusXa8i2b4EWQKF56F6t9QaH12OBeYA/8D+t9eM1ZNfdwESHQ22Bt7TWEx3a1Oh9U0o1A9YB12qt91Tk3iil2gBLgeaABm7XWmfXgG1/Ax4AioHNGJ9rgdM5dwLPAH+ah77QWv+jBmxbjJFinmM2eVpr/bHTOT2AV4FgIBW4T2t9mnKoV8Jh/nFdBfTE+BCXK6VudLpRfYHbtNZpNWybF9ARaOPqQ1FK+QOLgcuA/cAXSqmrtdZfnm/btNavAa+ZdnQGPgGecmpWY/dNKdUf4w+5g/m8ovfmJeAlrfW7SqkngCeAx86zbR2AR4HewEngdWAC8LzTqX2Bh7XWydVpT3m2OVx3iNb6cDmnLgXu1lqvV0otAu4BXi7vWvVtqHIYeERrXaC1LgR+Ado4tekDPKaU2qaUSlJK+dWQbQpDzL5USv2olJro9Ho/YKfWOt0UlqXAiBqyzZGXgela6wyn4zV53+7B+PIdMp+f894opRoDQ4APzEOvO7c5T7adAu7XWp/QWhcD2yn9NwfGF/hO87NfqpQKPd+2KaUCTVteNT+3p5VSJb7zSqmLAH+t9Xrz0OtU4L7VK+HQWv9kuwFKqfbArcAy2+tKqSBgCzAF6AWEYPwq1QShwDfAX4ErgPuUUlc6vN4SQ/hsHAYia8g2wO6x+Wut33c6XqP3TWt9t9bacQV1Re5NOHDCwZs7L/fP2Tat9V6tdQqAUioCY8j3qYtTD2N4cT0wvKak820bxrD4W2AsMABjyDLO6bRK/d3Vq6GKDdPd/gKYorXeaTtujnfjHdo9h+ECV/tY0xnTxbe5+TmmSxgPfG0ec7V5Q9H5tsuJezHiCCXw5H0zqci98ej9U0q1Ar4EFmmtVzm/rrW+0aHtHOD3822T1vp3wPG6C4A7MYYzNip13+qVxwGglIrB+GWfprV+w+m1NkqpsQ6HvDgbqDzfdg1SSl1RzrUPAhc4PL+Qs+7weUcp1QQjhvB/Ll7z2H0zqci9OQo0U0r5lNPmvKCU6gisBd7QWs908XqwUuohh0M1cv+UUl2VUjed47qV+rurV8KhlGqNEdgbqbV+10WTPGCOUqqtGaycAHzsot35IAR4Vinlp5RqCox2uvYGQCmlosw//pEYv2A1RTfgN611jovXPHnfoAL3xoxprcYYnoLxy3re75/5WX4FPK61fq6MZtnAVDN4CcZwpibunxfwglIq1IwB/c35ulrrvUC++YMLFbxv9Uo4MMbgfsA8pdRW8999SqllSqk+WuujGO74ZxjTdV5AWR92taK1/hxj+LQF+B5YrLVOM21sqbXOB+4CPgR+Bn7lbKCvJrgEOOB4oDbcN4Dy7o1S6jWl1PVm0/HA35RSP2OM52tiOvtujFjCFIe/uRmOtmmtz2CkCLyslPoFYwZmatldVg9a623AvzG8oZ+BrbZZHdtnaza9HXjetC0Q+M+5+paNfARBcJv65nEIglADiHAIguA2IhyCILiNCIcgCG4jwiEIgtuIcAiC4DYiHIIguM3/A4soB1Lqn4FdAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 288x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "data_c = data_rpm.copy()\n",
    "\n",
    "gene_a = data_c[['isrib_a_hek_mrna']].sum(axis=1).values.tolist()\n",
    "gene_a = [x + 1 for x in gene_a]\n",
    "gene_a = np.array(gene_a).astype(np.float)\n",
    "gene_a = np.ndarray.tolist(gene_a)\n",
    "\n",
    "gene_b = data_c[['isrib_b_hek_mrna']].sum(axis=1).values.tolist()\n",
    "gene_b = [x + 1 for x in gene_b]\n",
    "gene_b = np.array(gene_b).astype(np.float)\n",
    "gene_b = np.ndarray.tolist(gene_b)\n",
    "\n",
    "gene_a = np.log2(gene_a)\n",
    "gene_b = np.log2(gene_b)\n",
    "\n",
    "slope, intercept, r_value, p_value, std_err = stats.linregress(gene_a, gene_b)\n",
    "x = np.linspace(np.amin(gene_a), np.amax(gene_b), 100)\n",
    "y = (slope * x) + intercept\n",
    "\n",
    "fig, ax = plt.subplots(1,1, figsize=(4,4))\n",
    "\n",
    "ax.scatter(gene_a, gene_b, s=1,c='black')\n",
    "ax.set_title('R2 = ' + str(r_value ** 2) + '\\np-value ' + str(p_value))\n",
    "ax.plot(x, y, '-r')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_c = data_rpm.copy()\n",
    "\n",
    "gene_a = data_c[['ribo_isrib_a_ribo']].sum(axis=1).values.tolist()\n",
    "gene_a = [x + 1 for x in gene_a]\n",
    "gene_a = np.array(gene_a).astype(np.float)\n",
    "gene_a = np.ndarray.tolist(gene_a)\n",
    "\n",
    "gene_b = data_c[['ribo_isrib_b_ribo']].sum(axis=1).values.tolist()\n",
    "gene_b = [x + 1 for x in gene_b]\n",
    "gene_b = np.array(gene_b).astype(np.float)\n",
    "gene_b = np.ndarray.tolist(gene_b)\n",
    "\n",
    "gene_a = np.log2(gene_a)\n",
    "gene_b = np.log2(gene_b)\n",
    "\n",
    "slope, intercept, r_value, p_value, std_err = stats.linregress(gene_a, gene_b)\n",
    "x = np.linspace(np.amin(gene_a), np.amax(gene_b), 100)\n",
    "y = (slope * x) + intercept\n",
    "\n",
    "fig, ax = plt.subplots(1,1, figsize=(4,4))\n",
    "\n",
    "ax.scatter(gene_a, gene_b, s=1,c='black')\n",
    "ax.set_title('R2 = ' + str(r_value ** 2) + '\\np-value ' + str(p_value))\n",
    "ax.plot(x, y, '-r')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_c = data_rpm.copy()\n",
    "\n",
    "gene_a = data_c[['isrib_a_hek_mrna']].sum(axis=1).values.tolist()\n",
    "gene_a = [x + 1 for x in gene_a]\n",
    "gene_a = np.array(gene_a).astype(np.float)\n",
    "gene_a = np.ndarray.tolist(gene_a)\n",
    "\n",
    "gene_b = data_c[['ribo_isrib_a_ribo']].sum(axis=1).values.tolist()\n",
    "gene_b = [x + 1 for x in gene_b]\n",
    "gene_b = np.array(gene_b).astype(np.float)\n",
    "gene_b = np.ndarray.tolist(gene_b)\n",
    "\n",
    "gene_a = np.log2(gene_a)\n",
    "gene_b = np.log2(gene_b)\n",
    "\n",
    "slope, intercept, r_value, p_value, std_err = stats.linregress(gene_a, gene_b)\n",
    "x = np.linspace(np.amin(gene_a), np.amax(gene_b), 100)\n",
    "y = (slope * x) + intercept\n",
    "\n",
    "fig, ax = plt.subplots(1,1, figsize=(4,4))\n",
    "\n",
    "ax.scatter(gene_a, gene_b, s=1,c='black')\n",
    "ax.set_title('R = ' + str(r_value) + '\\np-value ' + str(p_value))\n",
    "ax.plot(x, y, '-r')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Perform Translation Efficiency Normalizations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_te = data_rpm.copy()\n",
    "\n",
    "data_te['isrib_a'] = data_te['ribo_isrib_a_ribo'] / data_te['isrib_a_hek_mrna']\n",
    "data_te['isrib_b'] = data_te['ribo_isrib_b_ribo'] / data_te['isrib_b_hek_mrna']\n",
    "data_te['tm_a'] = data_te['ribo_tm_a_ribo'] / data_te['tm_a_hek_mrna']\n",
    "data_te['tm_b'] = data_te['ribo_tm_b_ribo'] / data_te['tm_b_hek_mrna']\n",
    "data_te['tmisrib_a'] = data_te['ribo_tmisrib_a_ribo'] / data_te['tmisrib_a_hek_mrna']\n",
    "data_te['tmisrib_b'] = data_te['ribo_tmisrib_b_ribo'] / data_te['tmisrib_b_hek_mrna']\n",
    "data_te['untr_a'] = data_te['ribo_untr_a_ribo'] / data_te['untr_a_hek_mrna']\n",
    "data_te['untr_b'] = data_te['ribo_untr_b_ribo'] / data_te['untr_b_hek_mrna']\n",
    "\n",
    "data_te = data_te[['untr_a','untr_b','tm_a','tm_b','isrib_a','isrib_b','tmisrib_a','tmisrib_b']]\n",
    "\n",
    "data_te.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Remove samples with 0 log-fold change and log2 transform\n",
    "data_logte = data_te.copy()\n",
    "\n",
    "data_logte = data_logte[(data_logte != 0).all(1)]\n",
    "\n",
    "data_logte = np.log2(data_logte)\n",
    "\n",
    "data_logte.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "xp.check_samples(data_logte)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test accuracy of XPRESSpipe counts table vs Ingolia table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "xpresspipe_table = pd.read_csv('isrib_counts_xpresspipe.tsv',sep='\\t',index_col=0)\n",
    "xpresspipe_table = xp.convert_names_gtf(xpresspipe_table, '/Users/jordan/Desktop/Homo_sapiens.GRCh38.95.gtf')\n",
    "\n",
    "xpresspipe_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "xpresspipe_table_deduped = xpresspipe_table[~xpresspipe_table.index.duplicated(keep='first')]\n",
    "data_deduped = data[~data.index.duplicated(keep='first')]\n",
    "\n",
    "xpresspipe_table_deduped = xpresspipe_table_deduped.dropna()\n",
    "data_deduped = data_deduped.dropna()\n",
    "\n",
    "xpresspipe_table_deduped = xpresspipe_table_deduped.reindex(labels=data_deduped.index.tolist())\n",
    "data_deduped = data_deduped.reindex(labels=xpresspipe_table_deduped.index.tolist())\n",
    "\n",
    "xpresspipe_table_deduped = xpresspipe_table_deduped.dropna()\n",
    "data_deduped = data_deduped.dropna()\n",
    "\n",
    "data_deduped = data_deduped.reindex(labels=xpresspipe_table_deduped.index.tolist())\n",
    "\n",
    "print(xpresspipe_table_deduped.shape)\n",
    "print(data_deduped.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_a = data_deduped.sum(axis=1).values.tolist()\n",
    "gene_a = [x + 1 for x in gene_a]\n",
    "gene_a = np.array(gene_a).astype(np.float)\n",
    "gene_a = np.ndarray.tolist(gene_a)\n",
    "\n",
    "gene_b = xpresspipe_table_deduped.sum(axis=1).values.tolist()\n",
    "gene_b = [x + 1 for x in gene_b]\n",
    "gene_b = np.array(gene_b).astype(np.float)\n",
    "gene_b = np.ndarray.tolist(gene_b)\n",
    "\n",
    "gene_a = np.log2(gene_a)\n",
    "gene_b = np.log2(gene_b)\n",
    "\n",
    "slope, intercept, r_value, p_value, std_err = stats.linregress(gene_a, gene_b)\n",
    "x = np.linspace(np.amin(gene_a), np.amax(gene_b), 100)\n",
    "y = (slope * x) + intercept\n",
    "\n",
    "fig, ax = plt.subplots(1,1, figsize=(4,4))\n",
    "\n",
    "ax.scatter(gene_a, gene_b, s=1,c='black')\n",
    "ax.set_title('R = ' + str(r_value) + '\\np-value ' + str(p_value))\n",
    "ax.plot(x, y, '-r')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_a = data_deduped.sum(axis=1).values.tolist()\n",
    "gene_a = [x + 1 for x in gene_a]\n",
    "gene_a = np.array(gene_a).astype(np.float)\n",
    "gene_a = np.ndarray.tolist(gene_a)\n",
    "\n",
    "gene_b = xpresspipe_table_deduped[xpresspipe_table_deduped.columns[1::2]].sum(axis=1).values.tolist()\n",
    "gene_b = [x + 1 for x in gene_b]\n",
    "gene_b = np.array(gene_b).astype(np.float)\n",
    "gene_b = np.ndarray.tolist(gene_b)\n",
    "\n",
    "gene_a = np.log2(gene_a)\n",
    "gene_b = np.log2(gene_b)\n",
    "\n",
    "slope, intercept, r_value, p_value, std_err = stats.linregress(gene_a, gene_b)\n",
    "x = np.linspace(np.amin(gene_a), np.amax(gene_b), 100)\n",
    "y = (slope * x) + intercept\n",
    "\n",
    "fig, ax = plt.subplots(1,1, figsize=(4,4))\n",
    "\n",
    "ax.scatter(gene_a, gene_b, s=1,c='black')\n",
    "ax.set_title('R = ' + str(r_value) + '\\np-value ' + str(p_value))\n",
    "ax.plot(x, y, '-r')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_a = data_deduped.sum(axis=1).values.tolist()\n",
    "gene_a = [x + 1 for x in gene_a]\n",
    "gene_a = np.array(gene_a).astype(np.float)\n",
    "gene_a = np.ndarray.tolist(gene_a)\n",
    "\n",
    "gene_b = xpresspipe_table_deduped[xpresspipe_table_deduped.columns[::2]].sum(axis=1).values.tolist()\n",
    "gene_b = [x + 1 for x in gene_b]\n",
    "gene_b = np.array(gene_b).astype(np.float)\n",
    "gene_b = np.ndarray.tolist(gene_b)\n",
    "\n",
    "gene_a = np.log2(gene_a)\n",
    "gene_b = np.log2(gene_b)\n",
    "\n",
    "slope, intercept, r_value, p_value, std_err = stats.linregress(gene_a, gene_b)\n",
    "x = np.linspace(np.amin(gene_a), np.amax(gene_b), 100)\n",
    "y = (slope * x) + intercept\n",
    "\n",
    "fig, ax = plt.subplots(1,1, figsize=(4,4))\n",
    "\n",
    "ax.scatter(gene_a, gene_b, s=1,c='black')\n",
    "ax.set_title('R = ' + str(r_value) + '\\np-value ' + str(p_value))\n",
    "ax.plot(x, y, '-r')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_a = data_deduped['isrib_a_hek_mrna'].values.tolist()\n",
    "gene_a = [x + 1 for x in gene_a]\n",
    "gene_a = np.array(gene_a).astype(np.float)\n",
    "gene_a = np.ndarray.tolist(gene_a)\n",
    "\n",
    "gene_b = xpresspipe_table_deduped['trimmed_SRR1795413_1_final'].values.tolist()\n",
    "gene_b = [x + 1 for x in gene_b]\n",
    "gene_b = np.array(gene_b).astype(np.float)\n",
    "gene_b = np.ndarray.tolist(gene_b)\n",
    "\n",
    "gene_a = np.log2(gene_a)\n",
    "gene_b = np.log2(gene_b)\n",
    "\n",
    "slope, intercept, r_value, p_value, std_err = stats.linregress(gene_a, gene_b)\n",
    "x = np.linspace(np.amin(gene_a), np.amax(gene_b), 100)\n",
    "y = (slope * x) + intercept\n",
    "\n",
    "fig, ax = plt.subplots(1,1, figsize=(4,4))\n",
    "\n",
    "ax.scatter(gene_a, gene_b, s=1,c='black')\n",
    "ax.set_title('R = ' + str(r_value) + '\\np-value ' + str(p_value))\n",
    "ax.plot(x, y, '-r')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_a = xpresspipe_table_deduped['trimmed_SRR1795429_1_final'].values.tolist()\n",
    "gene_a = [x + 1 for x in gene_a]\n",
    "gene_a = np.array(gene_a).astype(np.float)\n",
    "gene_a = np.ndarray.tolist(gene_a)\n",
    "\n",
    "gene_b = xpresspipe_table_deduped['trimmed_SRR1795413_1_final'].values.tolist()\n",
    "gene_b = [x + 1 for x in gene_b]\n",
    "gene_b = np.array(gene_b).astype(np.float)\n",
    "gene_b = np.ndarray.tolist(gene_b)\n",
    "\n",
    "gene_a = np.log2(gene_a)\n",
    "gene_b = np.log2(gene_b)\n",
    "\n",
    "slope, intercept, r_value, p_value, std_err = stats.linregress(gene_a, gene_b)\n",
    "x = np.linspace(np.amin(gene_a), np.amax(gene_b), 100)\n",
    "y = (slope * x) + intercept\n",
    "\n",
    "fig, ax = plt.subplots(1,1, figsize=(4,4))\n",
    "\n",
    "ax.scatter(gene_a, gene_b, s=1,c='black')\n",
    "ax.set_title('R = ' + str(r_value) + '\\np-value ' + str(p_value))\n",
    "ax.plot(x, y, '-r')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reproduce Figure 1 from DOI: 10.7554/eLife.05033 (original paper)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Tm treatment vs Untreated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_fig1a = data_rpm.copy()\n",
    "data_fig1a = data_fig1a[['ribo_tm_a_ribo','ribo_tm_b_ribo','tm_a_hek_mrna','tm_b_hek_mrna',\n",
    "                         'ribo_untr_a_ribo','ribo_untr_b_ribo','untr_a_hek_mrna','untr_b_hek_mrna']]\n",
    "\n",
    "data_fig1a['ribo_tm'] = data_fig1a['ribo_tm_a_ribo'] + data_fig1a['ribo_tm_b_ribo']\n",
    "data_fig1a['rna_tm'] = data_fig1a['tm_a_hek_mrna'] + data_fig1a['tm_b_hek_mrna']\n",
    "data_fig1a['ribo_untr'] = data_fig1a['ribo_untr_a_ribo'] + data_fig1a['ribo_untr_b_ribo']\n",
    "data_fig1a['rna_untr'] = data_fig1a['untr_a_hek_mrna'] + data_fig1a['untr_b_hek_mrna']\n",
    "\n",
    "data_fig1a['ribo_fc'] = data_fig1a['ribo_tm'] / data_fig1a['ribo_untr']\n",
    "data_fig1a['rna_fc'] = data_fig1a['rna_tm'] / data_fig1a['rna_untr']\n",
    "\n",
    "data_fig1a_plot = data_fig1a[['ribo_fc','rna_fc']]\n",
    "\n",
    "data_fig1a_plot = data_fig1a_plot[(data_fig1a_plot != 0).all(1)]\n",
    "data_fig1a_plot = data_fig1a_plot.replace([np.inf, -np.inf], np.nan).dropna()\n",
    "\n",
    "#Remove duplicate indices\n",
    "data_fig1a_plot = data_fig1a_plot[~data_fig1a_plot.index.duplicated(keep='first')]\n",
    "\n",
    "data_fig1a_plot.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_fig1a = data_fig1a[(data_fig1a != 0).all(1)]\n",
    "data_fig1a = data_fig1a.replace([np.inf, -np.inf], np.nan).dropna()\n",
    "\n",
    "data_signif = xp.calculate_p(data_fig1a, 'ribo_tm_', 'ribo_untr_')\n",
    "data_signif['p_value'] = 10 ** -data_signif['-log$_1$$_0$(P-Value)']\n",
    "\n",
    "#Use R's padjust BH\n",
    "stats = importr('stats')\n",
    "p_adjust = stats.p_adjust(FloatVector(data_signif['p_value'].values.tolist()), method = 'BH')\n",
    "data_signif['fdr'] = np.asarray(p_adjust).tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_list = data_signif.loc[data_signif['p_value'] < 0.05].index.tolist()\n",
    "data_fig1a_threshold = data_fig1a_plot.reindex(labels=sig_list, axis=0)\n",
    "print(data_fig1a_threshold.shape)\n",
    "\n",
    "data_fig1a_orf = data_fig1a_plot.loc[['SLC35A4','PTP4A1','UCP2','PNRC2','C7orf31','BCL2L11','SAT1']]\n",
    "print(data_fig1a_orf.shape)\n",
    "\n",
    "data_fig1a_isr = data_fig1a_plot.loc[['ATF5','ATF4','DDIT3','PPP1R15A']]\n",
    "print(data_fig1a_isr.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.close()\n",
    "plt.clf()\n",
    "\n",
    "#Prepare plot\n",
    "fig, ax = plt.subplots(1,1, figsize=(7,7))\n",
    "plt.grid(False)\n",
    "\n",
    "for index, row in data_fig1a_orf.iterrows():\n",
    "    if index == 'BCL2L11' or index == 'PNRC2':\n",
    "        ax.text(row[1], row[0] - 0.3, str(index), horizontalalignment='right', size='medium', color='green', weight='semibold')\n",
    "    else:\n",
    "        ax.text(row[1] - 0.05, row[0] - 0.1, str(index), horizontalalignment='right', size='medium', color='green', weight='semibold')\n",
    "for index, row in data_fig1a_isr.iterrows():\n",
    "    ax.text(row[1] + 0.1, row[0] , str(index), horizontalalignment='left', size='medium', color='#be00be', weight='semibold')\n",
    "    \n",
    "    \n",
    "ax.axhline(1, ls='-', color='black')\n",
    "ax.axvline(1, ls='-', color='black')\n",
    "\n",
    "rect = patches.Rectangle((.5,.5),1.5,1.5,linewidth=1.5,edgecolor='gray',facecolor='none')\n",
    "ax.add_patch(rect)\n",
    "\n",
    "ax.set_ylabel(\"Ribo-Seq (Tm/UT)\", fontsize=16)\n",
    "ax.set_xlabel(\"mRNA-Seq (Tm/UT)\", fontsize=16)\n",
    "ax.set_xlim(.125,8.5)\n",
    "ax.set_ylim(.125,8.5)\n",
    "ax.set_facecolor(\"#FFFFFF\")\n",
    "\n",
    "x = [.125,.250,.500,1,2,4,8]\n",
    "ticks = [\"1/8\",\"1/4\",\"1/2\",\"1\",\"2\",\"4\",\"8\"]\n",
    "\n",
    "ax.set_xscale('log', basex=2)\n",
    "ax.set_xticks(x)\n",
    "ax.set_xticklabels(ticks, fontsize=12)\n",
    "ax.set_xticklabels([\"\"]*len(x), minor=True)\n",
    "\n",
    "ax.set_yscale('log', basey=2)\n",
    "ax.set_yticks(x)\n",
    "ax.set_yticklabels(ticks, fontsize=12)\n",
    "ax.set_yticklabels([\"\"]*len(x), minor=True)\n",
    "\n",
    "#Prep data\n",
    "#All\n",
    "ribo_all = data_fig1a[['ribo_fc']].sum(axis=1).values.tolist()\n",
    "ribo_all = np.array(ribo_all).astype(np.float)\n",
    "ribo_all = np.ndarray.tolist(ribo_all)\n",
    "\n",
    "rna_all = data_fig1a[['rna_fc']].sum(axis=1).values.tolist()\n",
    "rna_all = np.array(rna_all).astype(np.float)\n",
    "rna_all = np.ndarray.tolist(rna_all)\n",
    "\n",
    "#ORFs\n",
    "ribo_orf = data_fig1a_orf[['ribo_fc']].sum(axis=1).values.tolist()\n",
    "ribo_orf = np.array(ribo_orf).astype(np.float)\n",
    "ribo_orf = np.ndarray.tolist(ribo_orf)\n",
    "\n",
    "rna_orf = data_fig1a_orf[['rna_fc']].sum(axis=1).values.tolist()\n",
    "rna_orf = np.array(rna_orf).astype(np.float)\n",
    "rna_orf = np.ndarray.tolist(rna_orf)\n",
    "\n",
    "#ISR\n",
    "ribo_isr = data_fig1a_isr[['ribo_fc']].sum(axis=1).values.tolist()\n",
    "ribo_isr = np.array(ribo_isr).astype(np.float)\n",
    "ribo_isr = np.ndarray.tolist(ribo_isr)\n",
    "\n",
    "rna_isr = data_fig1a_isr[['rna_fc']].sum(axis=1).values.tolist()\n",
    "rna_isr = np.array(rna_isr).astype(np.float)\n",
    "rna_isr = np.ndarray.tolist(rna_isr)\n",
    "\n",
    "#Plot data\n",
    "ax.scatter(rna_all, ribo_all, s=.5,c='gray')\n",
    "#ax.scatter(rna_thres, ribo_thres, s=5,c='black',alpha=0.7)\n",
    "ax.scatter(rna_orf, ribo_orf, s=25,c='green')\n",
    "ax.scatter(rna_isr, ribo_isr, s=25,c='#be00be')\n",
    "\n",
    "plt.savefig('./plots/tm_1a.pdf', dpi=1800, bbox_inches='tight')\n",
    "\n",
    "plt.show"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Tm + ISRIB treament vs untreated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_fig1b = data_tpm.copy()\n",
    "data_fig1b = data_fig1b[['ribo_tmisrib_a_ribo','ribo_tmisrib_b_ribo','tmisrib_a_hek_mrna','tmisrib_b_hek_mrna',\n",
    "                         'ribo_untr_a_ribo','ribo_untr_b_ribo','untr_a_hek_mrna','untr_b_hek_mrna']]\n",
    "\n",
    "data_fig1b['ribo_tmisrib'] = data_fig1b['ribo_tmisrib_a_ribo'] + data_fig1b['ribo_tmisrib_b_ribo']\n",
    "data_fig1b['rna_tmisrib'] = data_fig1b['tmisrib_a_hek_mrna'] + data_fig1b['tmisrib_b_hek_mrna']\n",
    "data_fig1b['ribo_untr'] = data_fig1b['ribo_untr_a_ribo'] + data_fig1b['ribo_untr_b_ribo']\n",
    "data_fig1b['rna_untr'] = data_fig1b['untr_a_hek_mrna'] + data_fig1b['untr_b_hek_mrna']\n",
    "\n",
    "data_fig1b['ribo_fc'] = data_fig1b['ribo_tmisrib'] / data_fig1b['ribo_untr']\n",
    "data_fig1b['rna_fc'] = data_fig1b['rna_tmisrib'] / data_fig1b['rna_untr']\n",
    "\n",
    "data_fig1b_plot = data_fig1b[['ribo_fc','rna_fc']]\n",
    "\n",
    "data_fig1b_plot = data_fig1b_plot[(data_fig1b_plot != 0).all(1)]\n",
    "data_fig1b_plot = data_fig1b_plot.replace([np.inf, -np.inf], np.nan).dropna()\n",
    "\n",
    "#Remove duplicate indices\n",
    "data_fig1b_plot = data_fig1b_plot[~data_fig1b_plot.index.duplicated(keep='first')]\n",
    "\n",
    "data_fig1b_plot.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_fig1b = data_fig1b[(data_fig1b != 0).all(1)]\n",
    "data_fig1b = data_fig1b.replace([np.inf, -np.inf], np.nan).dropna()\n",
    "\n",
    "data_signif_1b = xp.calculate_p(data_fig1b, 'ribo_tmisrib_', 'ribo_untr_')\n",
    "data_signif_1b['p_value'] = 10 ** -data_signif_1b['-log$_1$$_0$(P-Value)']\n",
    "\n",
    "#Use R's padjust BH\n",
    "stats = importr('stats')\n",
    "p_adjust_1b = stats.p_adjust(FloatVector(data_signif_1b['p_value'].values.tolist()), method = 'BH')\n",
    "data_signif_1b['fdr'] = np.asarray(p_adjust_1b).tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_list = data_signif_1b.loc[data_signif_1b['p_value'] < 0.05].index.tolist()\n",
    "data_fig1b_threshold = data_fig1b_plot.reindex(labels=sig_list, axis=0)\n",
    "print(data_fig1b_threshold.shape)\n",
    "\n",
    "data_fig1b_orf = data_fig1b_plot.loc[['SLC35A4','PTP4A1','UCP2','PNRC2','C7orf31','BCL2L11','SAT1']]\n",
    "print(data_fig1b_orf.shape)\n",
    "\n",
    "data_fig1b_isr = data_fig1b_plot.loc[['ATF5','ATF4','DDIT3','PPP1R15A']]\n",
    "print(data_fig1b_isr.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.close()\n",
    "plt.clf()\n",
    "\n",
    "#Prepare plot\n",
    "fig, ax = plt.subplots(1,1, figsize=(7,7))\n",
    "plt.grid(False)\n",
    "\n",
    "for index, row in data_fig1b_orf.iterrows():\n",
    "    ax.text(row[1] - 0.1, row[0] - 0.05, str(index), horizontalalignment='right', size='medium', color='green', weight='semibold')\n",
    "for index, row in data_fig1b_isr.iterrows():\n",
    "    if index == 'DDIT3':\n",
    "        ax.text(row[1], row[0], str(index), horizontalalignment='right', size='medium', color='#be00be', weight='semibold')\n",
    "    else:\n",
    "        ax.text(row[1] + 0.1, row[0] , str(index), horizontalalignment='left', size='medium', color='#be00be', weight='semibold')\n",
    "    \n",
    "ax.axhline(1, ls='-', color='black')\n",
    "ax.axvline(1, ls='-', color='black')\n",
    "\n",
    "rect = patches.Rectangle((.5,.5),1.5,1.5,linewidth=1.5,edgecolor='gray',facecolor='none')\n",
    "ax.add_patch(rect)\n",
    "\n",
    "ax.set_ylabel(\"Ribo-Seq (Tm + ISRIB/UT)\", fontsize=16)\n",
    "ax.set_xlabel(\"mRNA-Seq (Tm + ISRIB/UT)\", fontsize=16)\n",
    "ax.set_xlim(.125,8.5)\n",
    "ax.set_ylim(.125,8.5)\n",
    "ax.set_facecolor(\"#FFFFFF\")\n",
    "\n",
    "x = [.125,.250,.500,1,2,4,8]\n",
    "ticks = [\"1/8\",\"1/4\",\"1/2\",\"1\",\"2\",\"4\",\"8\"]\n",
    "\n",
    "ax.set_xscale('log', basex=2)\n",
    "ax.set_xticks(x)\n",
    "ax.set_xticklabels(ticks, fontsize=12)\n",
    "ax.set_xticklabels([\"\"]*len(xm), minor=True)\n",
    "\n",
    "ax.set_yscale('log', basey=2)\n",
    "ax.set_yticks(x)\n",
    "ax.set_yticklabels(ticks, fontsize=12)\n",
    "ax.set_yticklabels([\"\"]*len(xm), minor=True)\n",
    "\n",
    "#Prep data\n",
    "#All\n",
    "ribo_all = data_fig1b[['ribo_fc']].sum(axis=1).values.tolist()\n",
    "ribo_all = np.array(ribo_all).astype(np.float)\n",
    "ribo_all = np.ndarray.tolist(ribo_all)\n",
    "\n",
    "rna_all = data_fig1b[['rna_fc']].sum(axis=1).values.tolist()\n",
    "rna_all = np.array(rna_all).astype(np.float)\n",
    "rna_all = np.ndarray.tolist(rna_all)\n",
    "\n",
    "#ORFs\n",
    "ribo_orf = data_fig1b_orf[['ribo_fc']].sum(axis=1).values.tolist()\n",
    "ribo_orf = np.array(ribo_orf).astype(np.float)\n",
    "ribo_orf = np.ndarray.tolist(ribo_orf)\n",
    "\n",
    "rna_orf = data_fig1b_orf[['rna_fc']].sum(axis=1).values.tolist()\n",
    "rna_orf = np.array(rna_orf).astype(np.float)\n",
    "rna_orf = np.ndarray.tolist(rna_orf)\n",
    "\n",
    "#ISR\n",
    "ribo_isr = data_fig1b_isr[['ribo_fc']].sum(axis=1).values.tolist()\n",
    "ribo_isr = np.array(ribo_isr).astype(np.float)\n",
    "ribo_isr = np.ndarray.tolist(ribo_isr)\n",
    "\n",
    "rna_isr = data_fig1b_isr[['rna_fc']].sum(axis=1).values.tolist()\n",
    "rna_isr = np.array(rna_isr).astype(np.float)\n",
    "rna_isr = np.ndarray.tolist(rna_isr)\n",
    "\n",
    "#Plot data\n",
    "ax.scatter(rna_all, ribo_all, s=.5,c='gray')\n",
    "#ax.scatter(rna_thres, ribo_thres, s=5,c='black',alpha=0.7)\n",
    "ax.scatter(rna_orf, ribo_orf, s=25,c='green')\n",
    "ax.scatter(rna_isr, ribo_isr, s=25,c='#be00be')\n",
    "\n",
    "plt.savefig('./plots/tmisrib_1b.pdf', dpi=1800, bbox_inches='tight')\n",
    "\n",
    "plt.show"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### ISRIB treament vs untreated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_fig1c = data_tpm.copy()\n",
    "data_fig1c = data_fig1c[['ribo_isrib_a_ribo','ribo_isrib_b_ribo','isrib_a_hek_mrna','isrib_b_hek_mrna',\n",
    "                         'ribo_untr_a_ribo','ribo_untr_b_ribo','untr_a_hek_mrna','untr_b_hek_mrna']]\n",
    "\n",
    "data_fig1c['ribo_isrib'] = data_fig1c['ribo_isrib_a_ribo'] + data_fig1c['ribo_isrib_b_ribo']\n",
    "data_fig1c['rna_isrib'] = data_fig1c['isrib_a_hek_mrna'] + data_fig1c['isrib_b_hek_mrna']\n",
    "data_fig1c['ribo_untr'] = data_fig1c['ribo_untr_a_ribo'] + data_fig1c['ribo_untr_b_ribo']\n",
    "data_fig1c['rna_untr'] = data_fig1c['untr_a_hek_mrna'] + data_fig1c['untr_b_hek_mrna']\n",
    "\n",
    "data_fig1c['ribo_fc'] = data_fig1c['ribo_isrib'] / data_fig1c['ribo_untr']\n",
    "data_fig1c['rna_fc'] = data_fig1c['rna_isrib'] / data_fig1c['rna_untr']\n",
    "\n",
    "data_fig1c_plot = data_fig1c[['ribo_fc','rna_fc']]\n",
    "\n",
    "data_fig1c_plot = data_fig1c_plot[(data_fig1c_plot != 0).all(1)]\n",
    "data_fig1c_plot = data_fig1c_plot.replace([np.inf, -np.inf], np.nan).dropna()\n",
    "\n",
    "#Remove duplicate indices\n",
    "data_fig1c_plot = data_fig1c_plot[~data_fig1c_plot.index.duplicated(keep='first')]\n",
    "\n",
    "data_fig1c_plot.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_fig1c = data_fig1c[(data_fig1c != 0).all(1)]\n",
    "data_fig1c = data_fig1c.replace([np.inf, -np.inf], np.nan).dropna()\n",
    "\n",
    "data_signif_1c = xp.calculate_p(data_fig1c, 'ribo_isrib_', 'ribo_untr_')\n",
    "data_signif_1c['p_value'] = 10 ** -data_signif_1c['-log$_1$$_0$(P-Value)']\n",
    "\n",
    "#Use R's padjust BH\n",
    "stats = importr('stats')\n",
    "p_adjust_1c = stats.p_adjust(FloatVector(data_signif_1c['p_value'].values.tolist()), method = 'BH')\n",
    "data_signif_1c['fdr'] = np.asarray(p_adjust_1c).tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_list = data_signif_1c.loc[data_signif_1c['p_value'] < 0.05].index.tolist()\n",
    "data_fig1c_threshold = data_fig1c_plot.reindex(labels=sig_list, axis=0)\n",
    "print(data_fig1c_threshold.shape)\n",
    "\n",
    "data_fig1c_orf = data_fig1c_plot.loc[['SLC35A4','PTP4A1','UCP2','PNRC2','C7orf31','BCL2L11','SAT1']]\n",
    "print(data_fig1c_orf.shape)\n",
    "\n",
    "data_fig1c_isr = data_fig1c_plot.loc[['ATF5','ATF4','DDIT3','PPP1R15A']]\n",
    "print(data_fig1c_isr.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.close()\n",
    "plt.clf()\n",
    "\n",
    "#Prepare plot\n",
    "fig, ax = plt.subplots(1,1, figsize=(7,7))\n",
    "plt.grid(False)\n",
    "\n",
    "for index, row in data_fig1c_orf.iterrows():\n",
    "    ax.text(row[1] + 0.1, row[0], str(index), horizontalalignment='left', size='medium', color='green', weight='semibold')\n",
    "for index, row in data_fig1c_isr.iterrows():\n",
    "    ax.text(row[1] - 0.1, row[0] - 0.05, str(index), horizontalalignment='right', size='medium', color='#be00be', weight='semibold')\n",
    "\n",
    "ax.axhline(1, ls='-', color='black')\n",
    "ax.axvline(1, ls='-', color='black')\n",
    "\n",
    "rect = patches.Rectangle((.5,.5),1.5,1.5,linewidth=1.5,edgecolor='gray',facecolor='none')\n",
    "ax.add_patch(rect)\n",
    "\n",
    "ax.set_ylabel(\"Ribo-Seq (ISRIB/UT)\", fontsize=16)\n",
    "ax.set_xlabel(\"mRNA-Seq (ISRIB/UT)\", fontsize=16)\n",
    "ax.set_xlim(.125,8.5)\n",
    "ax.set_ylim(.125,8.5)\n",
    "ax.set_facecolor(\"#FFFFFF\")\n",
    "\n",
    "x = [.125,.250,.500,1,2,4,8]\n",
    "ticks = [\"1/8\",\"1/4\",\"1/2\",\"1\",\"2\",\"4\",\"8\"]\n",
    "\n",
    "ax.set_xscale('log', basex=2)\n",
    "ax.set_xticks(x)\n",
    "ax.set_xticklabels(ticks, fontsize=12)\n",
    "ax.set_xticklabels([\"\"]*len(xm), minor=True)\n",
    "\n",
    "ax.set_yscale('log', basey=2)\n",
    "ax.set_yticks(x)\n",
    "ax.set_yticklabels(ticks, fontsize=12)\n",
    "ax.set_yticklabels([\"\"]*len(xm), minor=True)\n",
    "\n",
    "#Prep data\n",
    "#All\n",
    "ribo_all = data_fig1c[['ribo_fc']].sum(axis=1).values.tolist()\n",
    "ribo_all = np.array(ribo_all).astype(np.float)\n",
    "ribo_all = np.ndarray.tolist(ribo_all)\n",
    "\n",
    "rna_all = data_fig1c[['rna_fc']].sum(axis=1).values.tolist()\n",
    "rna_all = np.array(rna_all).astype(np.float)\n",
    "rna_all = np.ndarray.tolist(rna_all)\n",
    "\n",
    "#ORFs\n",
    "ribo_orf = data_fig1c_orf[['ribo_fc']].sum(axis=1).values.tolist()\n",
    "ribo_orf = np.array(ribo_orf).astype(np.float)\n",
    "ribo_orf = np.ndarray.tolist(ribo_orf)\n",
    "\n",
    "rna_orf = data_fig1c_orf[['rna_fc']].sum(axis=1).values.tolist()\n",
    "rna_orf = np.array(rna_orf).astype(np.float)\n",
    "rna_orf = np.ndarray.tolist(rna_orf)\n",
    "\n",
    "#ISR\n",
    "ribo_isr = data_fig1c_isr[['ribo_fc']].sum(axis=1).values.tolist()\n",
    "ribo_isr = np.array(ribo_isr).astype(np.float)\n",
    "ribo_isr = np.ndarray.tolist(ribo_isr)\n",
    "\n",
    "rna_isr = data_fig1c_isr[['rna_fc']].sum(axis=1).values.tolist()\n",
    "rna_isr = np.array(rna_isr).astype(np.float)\n",
    "rna_isr = np.ndarray.tolist(rna_isr)\n",
    "\n",
    "#Plot data\n",
    "ax.scatter(rna_all, ribo_all, s=.5,c='gray')\n",
    "#ax.scatter(rna_thres, ribo_thres, s=5,c='black',alpha=0.7)\n",
    "ax.scatter(rna_orf, ribo_orf, s=25,c='green')\n",
    "ax.scatter(rna_isr, ribo_isr, s=25,c='#be00be')\n",
    "\n",
    "plt.savefig('./plots/isrib_1c.pdf', dpi=1800, bbox_inches='tight')\n",
    "\n",
    "plt.show"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Metabolism analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mitocarta = pd.read_csv('./human_mitocarta_gene_list.txt',sep='\\t', header=None)\n",
    "mitocarta = mitocarta[0].tolist()\n",
    "\n",
    "enzymes = pd.read_csv('./reactome_metabolism_enzymes_gene_list.txt',sep='\\t', header=None)\n",
    "enzymes = enzymes[0].tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Mitocarta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_fig1a_mitocarta = data_fig1a_plot.reindex(labels=mitocarta)\n",
    "print(data_fig1a_mitocarta.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.close()\n",
    "plt.clf()\n",
    "\n",
    "#Prepare plot\n",
    "fig, ax = plt.subplots(1,1, figsize=(7,7))\n",
    "plt.grid(False)\n",
    "\n",
    "ax.axhline(1, ls='-', color='black')\n",
    "ax.axvline(1, ls='-', color='black')\n",
    "\n",
    "ax.axhline(2, ls='--', color='gray')\n",
    "ax.axhline(0.5, ls='--', color='gray')\n",
    "\n",
    "ax.set_ylabel(\"Ribo-Seq (Tm/UT)\", fontsize=16)\n",
    "ax.set_xlabel(\"mRNA-Seq (Tm/UT)\", fontsize=16)\n",
    "ax.set_xlim(.125,8.5)\n",
    "ax.set_ylim(.125,8.5)\n",
    "ax.set_facecolor(\"#FFFFFF\")\n",
    "\n",
    "for index, row in data_fig1a_mitocarta.iterrows():\n",
    "    if index == 'ACCS':\n",
    "        ax.text(row[1] + 0.05, row[0], str(index), horizontalalignment='left', size='medium', color='#be00be', weight='semibold')\n",
    "    elif index == 'CPT1B':\n",
    "        ax.text(row[1] + 0.05, row[0] + 0.05, str(index), horizontalalignment='left', size='medium', color='#be00be', weight='semibold')\n",
    "    elif index == 'NGRN':\n",
    "        ax.text(row[1] + 0.05, row[0] - 0.05, str(index), horizontalalignment='left', size='medium', color='#be00be', weight='semibold')\n",
    "    elif index == 'CPS1':\n",
    "        ax.text(row[1] - 0.05, row[0] - 0.025, str(index), horizontalalignment='right', size='medium', color='#be00be', weight='semibold')\n",
    "    elif index in ['SLMO1', 'UCP2', 'OBSCN']:\n",
    "        ax.text(row[1] + 0.05, row[0] - 0.05, str(index), horizontalalignment='left', size='medium', color='#be00be', weight='semibold')\n",
    "    elif index == 'MTHFD2L':\n",
    "        ax.text(row[1] + 0.05, row[0] + 0.05, str(index), horizontalalignment='left', size='medium', color='#be00be', weight='semibold')\n",
    "    else:\n",
    "        pass\n",
    "\n",
    "x = [.125,.250,.500,1,2,4,8]\n",
    "ticks = [\"1/8\",\"1/4\",\"1/2\",\"1\",\"2\",\"4\",\"8\"]\n",
    "\n",
    "ax.set_xscale('log', basex=2)\n",
    "ax.set_xticks(x)\n",
    "ax.set_xticklabels(ticks, fontsize=12)\n",
    "ax.set_xticklabels([\"\"]*len(xm), minor=True)\n",
    "\n",
    "ax.set_yscale('log', basey=2)\n",
    "ax.set_yticks(x)\n",
    "ax.set_yticklabels(ticks, fontsize=12)\n",
    "ax.set_yticklabels([\"\"]*len(xm), minor=True)\n",
    "\n",
    "#Prep data\n",
    "#All\n",
    "ribo_all = data_fig1a[['ribo_fc']].sum(axis=1).values.tolist()\n",
    "ribo_all = np.array(ribo_all).astype(np.float)\n",
    "ribo_all = np.ndarray.tolist(ribo_all)\n",
    "\n",
    "rna_all = data_fig1a[['rna_fc']].sum(axis=1).values.tolist()\n",
    "rna_all = np.array(rna_all).astype(np.float)\n",
    "rna_all = np.ndarray.tolist(rna_all)\n",
    "\n",
    "#ORFs\n",
    "ribo_mito = data_fig1a_mitocarta[['ribo_fc']].sum(axis=1).values.tolist()\n",
    "ribo_mito = np.array(ribo_mito).astype(np.float)\n",
    "ribo_mito = np.ndarray.tolist(ribo_mito)\n",
    "\n",
    "rna_mito = data_fig1a_mitocarta[['rna_fc']].sum(axis=1).values.tolist()\n",
    "rna_mito = np.array(rna_mito).astype(np.float)\n",
    "rna_mito = np.ndarray.tolist(rna_mito)\n",
    "\n",
    "#Plot data\n",
    "ax.scatter(rna_all, ribo_all, s=.5,c='gray')\n",
    "ax.scatter(rna_mito, ribo_mito, s=10,c='#002c80')\n",
    "\n",
    "plt.savefig('./plots/tm_mitocarta.pdf', dpi=1800, bbox_inches='tight')\n",
    "\n",
    "plt.show"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_fig1a_mitocarta.loc[data_fig1a_mitocarta['ribo_fc'] > 2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_fig1a_mitocarta.loc[data_fig1a_mitocarta['ribo_fc'] < .5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_fig1b_mitocarta = data_fig1b_plot.reindex(labels=mitocarta)\n",
    "print(data_fig1b_mitocarta.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.close()\n",
    "plt.clf()\n",
    "\n",
    "#Prepare plot\n",
    "fig, ax = plt.subplots(1,1, figsize=(7,7))\n",
    "plt.grid(False)\n",
    "\n",
    "ax.axhline(1, ls='-', color='black')\n",
    "ax.axvline(1, ls='-', color='black')\n",
    "\n",
    "ax.axhline(2, ls='--', color='gray')\n",
    "ax.axhline(0.5, ls='--', color='gray')\n",
    "\n",
    "ax.set_ylabel(\"Ribo-Seq (Tm + ISRIB/UT)\", fontsize=16)\n",
    "ax.set_xlabel(\"mRNA-Seq (Tm + ISRIB/UT)\", fontsize=16)\n",
    "ax.set_xlim(.125,8.5)\n",
    "ax.set_ylim(.125,8.5)\n",
    "ax.set_facecolor(\"#FFFFFF\")\n",
    "\n",
    "for index, row in data_fig1b_mitocarta.iterrows():\n",
    "    if index == 'ACCS':\n",
    "        ax.text(row[1] + 0.05, row[0], str(index), horizontalalignment='left', size='medium', color='#be00be', weight='semibold')\n",
    "    elif index == 'CPT1B':\n",
    "        ax.text(row[1] + 0.05, row[0] + 0.05, str(index), horizontalalignment='left', size='medium', color='#be00be', weight='semibold')\n",
    "    elif index == 'NGRN':\n",
    "        ax.text(row[1] + 0.05, row[0] - 0.05, str(index), horizontalalignment='left', size='medium', color='#be00be', weight='semibold')\n",
    "    elif index == 'MTHFD2L':\n",
    "        ax.text(row[1] - 0.05, row[0] - 0.05, str(index), horizontalalignment='right', size='medium', color='#be00be', weight='semibold')\n",
    "    elif index ==  'UCP2':\n",
    "        ax.text(row[1] - 0.05, row[0] - 0.05, str(index), horizontalalignment='right', size='medium', color='#be00be', weight='semibold')\n",
    "    elif index == 'CPS1':\n",
    "        ax.text(row[1] - 0.05, row[0] - 0.025, str(index), horizontalalignment='right', size='medium', color='#be00be', weight='semibold')\n",
    "    elif index in ['SLMO1', 'OBSCN']:\n",
    "        ax.text(row[1] + 0.05, row[0] - 0.05, str(index), horizontalalignment='left', size='medium', color='#be00be', weight='semibold')\n",
    "    else:\n",
    "        pass\n",
    "\n",
    "x = [.125,.250,.500,1,2,4,8]\n",
    "ticks = [\"1/8\",\"1/4\",\"1/2\",\"1\",\"2\",\"4\",\"8\"]\n",
    "\n",
    "ax.set_xscale('log', basex=2)\n",
    "ax.set_xticks(x)\n",
    "ax.set_xticklabels(ticks, fontsize=12)\n",
    "ax.set_xticklabels([\"\"]*len(xm), minor=True)\n",
    "\n",
    "ax.set_yscale('log', basey=2)\n",
    "ax.set_yticks(x)\n",
    "ax.set_yticklabels(ticks, fontsize=12)\n",
    "ax.set_yticklabels([\"\"]*len(xm), minor=True)\n",
    "\n",
    "#Prep data\n",
    "#All\n",
    "ribo_all = data_fig1b[['ribo_fc']].sum(axis=1).values.tolist()\n",
    "ribo_all = np.array(ribo_all).astype(np.float)\n",
    "ribo_all = np.ndarray.tolist(ribo_all)\n",
    "\n",
    "rna_all = data_fig1b[['rna_fc']].sum(axis=1).values.tolist()\n",
    "rna_all = np.array(rna_all).astype(np.float)\n",
    "rna_all = np.ndarray.tolist(rna_all)\n",
    "\n",
    "#ORFs\n",
    "ribo_mito = data_fig1b_mitocarta[['ribo_fc']].sum(axis=1).values.tolist()\n",
    "ribo_mito = np.array(ribo_mito).astype(np.float)\n",
    "ribo_mito = np.ndarray.tolist(ribo_mito)\n",
    "\n",
    "rna_mito = data_fig1b_mitocarta[['rna_fc']].sum(axis=1).values.tolist()\n",
    "rna_mito = np.array(rna_mito).astype(np.float)\n",
    "rna_mito = np.ndarray.tolist(rna_mito)\n",
    "\n",
    "#Plot data\n",
    "ax.scatter(rna_all, ribo_all, s=.5,c='gray')\n",
    "ax.scatter(rna_mito, ribo_mito, s=10,c='#002c80')\n",
    "\n",
    "plt.savefig('./plots/tmisrib_mitocarta.pdf', dpi=1800, bbox_inches='tight')\n",
    "\n",
    "plt.show"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_fig1b_mitocarta.loc[data_fig1b_mitocarta['ribo_fc'] > 2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_fig1b_mitocarta.loc[data_fig1b_mitocarta['ribo_fc'] < .5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_fig1c_mitocarta = data_fig1c_plot.reindex(labels=mitocarta)\n",
    "print(data_fig1c_mitocarta.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.close()\n",
    "plt.clf()\n",
    "\n",
    "#Prepare plot\n",
    "fig, ax = plt.subplots(1,1, figsize=(7,7))\n",
    "plt.grid(False)\n",
    "\n",
    "ax.axhline(1, ls='-', color='black')\n",
    "ax.axvline(1, ls='-', color='black')\n",
    "\n",
    "ax.axhline(2, ls='--', color='gray')\n",
    "ax.axhline(0.5, ls='--', color='gray')\n",
    "\n",
    "ax.set_ylabel(\"Ribo-Seq (ISRIB/UT)\", fontsize=16)\n",
    "ax.set_xlabel(\"mRNA-Seq (ISRIB/UT)\", fontsize=16)\n",
    "ax.set_xlim(.125,8.5)\n",
    "ax.set_ylim(.125,8.5)\n",
    "ax.set_facecolor(\"#FFFFFF\")\n",
    "\n",
    "for index, row in data_fig1c_mitocarta.iterrows():\n",
    "    if index == 'ACCS':\n",
    "        ax.text(row[1] + 0.05, row[0], str(index), horizontalalignment='left', size='medium', color='#be00be', weight='semibold')\n",
    "    elif index == 'CPT1B':\n",
    "        ax.text(row[1] + 0.05, row[0] + 0.05, str(index), horizontalalignment='left', size='medium', color='#be00be', weight='semibold')\n",
    "    elif index == 'NGRN':\n",
    "        ax.text(row[1] + 0.05, row[0] - 0.05, str(index), horizontalalignment='left', size='medium', color='#be00be', weight='semibold')\n",
    "    elif index == 'MTHFD2L':\n",
    "        ax.text(row[1] - 0.05, row[0] - 0.05, str(index), horizontalalignment='right', size='medium', color='#be00be', weight='semibold')\n",
    "    elif index ==  'UCP2':\n",
    "        ax.text(row[1] - 0.05, row[0] - 0.05, str(index), horizontalalignment='right', size='medium', color='#be00be', weight='semibold')\n",
    "    elif index == 'CPS1':\n",
    "        ax.text(row[1] - 0.05, row[0] - 0.025, str(index), horizontalalignment='right', size='medium', color='#be00be', weight='semibold')\n",
    "    elif index in ['SLMO1', 'OBSCN']:\n",
    "        ax.text(row[1] + 0.05, row[0] - 0.05, str(index), horizontalalignment='left', size='medium', color='#be00be', weight='semibold')\n",
    "    else:\n",
    "        pass\n",
    "\n",
    "x = [.125,.250,.500,1,2,4,8]\n",
    "ticks = [\"1/8\",\"1/4\",\"1/2\",\"1\",\"2\",\"4\",\"8\"]\n",
    "\n",
    "ax.set_xscale('log', basex=2)\n",
    "ax.set_xticks(x)\n",
    "ax.set_xticklabels(ticks, fontsize=12)\n",
    "ax.set_xticklabels([\"\"]*len(xm), minor=True)\n",
    "\n",
    "ax.set_yscale('log', basey=2)\n",
    "ax.set_yticks(x)\n",
    "ax.set_yticklabels(ticks, fontsize=12)\n",
    "ax.set_yticklabels([\"\"]*len(xm), minor=True)\n",
    "\n",
    "#Prep data\n",
    "#All\n",
    "ribo_all = data_fig1c[['ribo_fc']].sum(axis=1).values.tolist()\n",
    "ribo_all = np.array(ribo_all).astype(np.float)\n",
    "ribo_all = np.ndarray.tolist(ribo_all)\n",
    "\n",
    "rna_all = data_fig1c[['rna_fc']].sum(axis=1).values.tolist()\n",
    "rna_all = np.array(rna_all).astype(np.float)\n",
    "rna_all = np.ndarray.tolist(rna_all)\n",
    "\n",
    "#ORFs\n",
    "ribo_mito = data_fig1c_mitocarta[['ribo_fc']].sum(axis=1).values.tolist()\n",
    "ribo_mito = np.array(ribo_mito).astype(np.float)\n",
    "ribo_mito = np.ndarray.tolist(ribo_mito)\n",
    "\n",
    "rna_mito = data_fig1c_mitocarta[['rna_fc']].sum(axis=1).values.tolist()\n",
    "rna_mito = np.array(rna_mito).astype(np.float)\n",
    "rna_mito = np.ndarray.tolist(rna_mito)\n",
    "\n",
    "#Plot data\n",
    "ax.scatter(rna_all, ribo_all, s=.5,c='gray')\n",
    "ax.scatter(rna_mito, ribo_mito, s=10,c='#002c80')\n",
    "\n",
    "plt.savefig('./plots/isrib_mitocarta_1hr.pdf', dpi=1800, bbox_inches='tight')\n",
    "\n",
    "plt.show"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_fig1c_mitocarta.loc[data_fig1c_mitocarta['ribo_fc'] > 2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_fig1c_mitocarta.loc[data_fig1c_mitocarta['ribo_fc'] < 0.5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Metabolic Enzymes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_fig1a_enzymes = data_fig1a_plot.reindex(labels=enzymes)\n",
    "print(data_fig1a_enzymes.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.close()\n",
    "plt.clf()\n",
    "\n",
    "#Prepare plot\n",
    "fig, ax = plt.subplots(1,1, figsize=(7,7))\n",
    "plt.grid(False)\n",
    "\n",
    "ax.axhline(1, ls='-', color='black')\n",
    "ax.axvline(1, ls='-', color='black')\n",
    "\n",
    "ax.axhline(2, ls='--', color='gray')\n",
    "ax.axhline(0.5, ls='--', color='gray')\n",
    "\n",
    "ax.set_ylabel(\"Ribo-Seq (Tm/UT)\", fontsize=16)\n",
    "ax.set_xlabel(\"mRNA-Seq (Tm/UT)\", fontsize=16)\n",
    "ax.set_xlim(.125,8.5)\n",
    "ax.set_ylim(.125,8.5)\n",
    "ax.set_facecolor(\"#FFFFFF\")\n",
    "\n",
    "\n",
    "\n",
    "x = [.125,.250,.500,1,2,4,8]\n",
    "ticks = [\"1/8\",\"1/4\",\"1/2\",\"1\",\"2\",\"4\",\"8\"]\n",
    "\n",
    "ax.set_xscale('log', basex=2)\n",
    "ax.set_xticks(x)\n",
    "ax.set_xticklabels(ticks, fontsize=12)\n",
    "ax.set_xticklabels([\"\"]*len(xm), minor=True)\n",
    "\n",
    "ax.set_yscale('log', basey=2)\n",
    "ax.set_yticks(x)\n",
    "ax.set_yticklabels(ticks, fontsize=12)\n",
    "ax.set_yticklabels([\"\"]*len(xm), minor=True)\n",
    "\n",
    "#Prep data\n",
    "#All\n",
    "ribo_all = data_fig1a[['ribo_fc']].sum(axis=1).values.tolist()\n",
    "ribo_all = np.array(ribo_all).astype(np.float)\n",
    "ribo_all = np.ndarray.tolist(ribo_all)\n",
    "\n",
    "rna_all = data_fig1a[['rna_fc']].sum(axis=1).values.tolist()\n",
    "rna_all = np.array(rna_all).astype(np.float)\n",
    "rna_all = np.ndarray.tolist(rna_all)\n",
    "\n",
    "#ORFs\n",
    "ribo_mito = data_fig1a_enzymes[['ribo_fc']].sum(axis=1).values.tolist()\n",
    "ribo_mito = np.array(ribo_mito).astype(np.float)\n",
    "ribo_mito = np.ndarray.tolist(ribo_mito)\n",
    "\n",
    "rna_mito = data_fig1a_enzymes[['rna_fc']].sum(axis=1).values.tolist()\n",
    "rna_mito = np.array(rna_mito).astype(np.float)\n",
    "rna_mito = np.ndarray.tolist(rna_mito)\n",
    "\n",
    "#Plot data\n",
    "ax.scatter(rna_all, ribo_all, s=.5,c='gray')\n",
    "ax.scatter(rna_mito, ribo_mito, s=10,c='#ffa500')\n",
    "\n",
    "for index, row in data_fig1a_enzymes.iterrows():\n",
    "    if index in [ 'CPS1', 'CPT1B', 'UCP2', 'ST3GAL3', 'OAZ1', 'SAT1', 'ABCA1', 'PLCD1', 'ADRA2C', 'PLCD1']:\n",
    "        ax.text(row[1] + 0.025, row[0], str(index), horizontalalignment='left', size='medium', color='#016a86', weight='semibold')\n",
    "    else:\n",
    "        pass\n",
    "\n",
    "plt.savefig('./plots/tm_enzymes.pdf', dpi=1800, bbox_inches='tight')\n",
    "\n",
    "plt.show"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_fig1a_enzymes.loc[data_fig1a_enzymes['ribo_fc'] > 2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_fig1a_enzymes.loc[data_fig1a_enzymes['ribo_fc'] < 0.5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_fig1b_enzymes = data_fig1b_plot.reindex(labels=enzymes)\n",
    "print(data_fig1b_enzymes.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.close()\n",
    "plt.clf()\n",
    "\n",
    "#Prepare plot\n",
    "fig, ax = plt.subplots(1,1, figsize=(7,7))\n",
    "plt.grid(False)\n",
    "\n",
    "ax.axhline(1, ls='-', color='black')\n",
    "ax.axvline(1, ls='-', color='black')\n",
    "\n",
    "ax.axhline(2, ls='--', color='gray')\n",
    "ax.axhline(0.5, ls='--', color='gray')\n",
    "\n",
    "ax.set_ylabel(\"Ribo-Seq (Tm + ISRIB/UT)\", fontsize=16)\n",
    "ax.set_xlabel(\"mRNA-Seq (Tm + ISRIB/UT)\", fontsize=16)\n",
    "ax.set_xlim(.125,8.5)\n",
    "ax.set_ylim(.125,8.5)\n",
    "ax.set_facecolor(\"#FFFFFF\")\n",
    "\n",
    "x = [.125,.250,.500,1,2,4,8]\n",
    "ticks = [\"1/8\",\"1/4\",\"1/2\",\"1\",\"2\",\"4\",\"8\"]\n",
    "\n",
    "ax.set_xscale('log', basex=2)\n",
    "ax.set_xticks(x)\n",
    "ax.set_xticklabels(ticks, fontsize=12)\n",
    "ax.set_xticklabels([\"\"]*len(xm), minor=True)\n",
    "\n",
    "ax.set_yscale('log', basey=2)\n",
    "ax.set_yticks(x)\n",
    "ax.set_yticklabels(ticks, fontsize=12)\n",
    "ax.set_yticklabels([\"\"]*len(xm), minor=True)\n",
    "\n",
    "#Prep data\n",
    "#All\n",
    "ribo_all = data_fig1b[['ribo_fc']].sum(axis=1).values.tolist()\n",
    "ribo_all = np.array(ribo_all).astype(np.float)\n",
    "ribo_all = np.ndarray.tolist(ribo_all)\n",
    "\n",
    "rna_all = data_fig1b[['rna_fc']].sum(axis=1).values.tolist()\n",
    "rna_all = np.array(rna_all).astype(np.float)\n",
    "rna_all = np.ndarray.tolist(rna_all)\n",
    "\n",
    "#ORFs\n",
    "ribo_mito = data_fig1b_enzymes[['ribo_fc']].sum(axis=1).values.tolist()\n",
    "ribo_mito = np.array(ribo_mito).astype(np.float)\n",
    "ribo_mito = np.ndarray.tolist(ribo_mito)\n",
    "\n",
    "rna_mito = data_fig1b_enzymes[['rna_fc']].sum(axis=1).values.tolist()\n",
    "rna_mito = np.array(rna_mito).astype(np.float)\n",
    "rna_mito = np.ndarray.tolist(rna_mito)\n",
    "\n",
    "#Plot data\n",
    "ax.scatter(rna_all, ribo_all, s=.5,c='gray')\n",
    "ax.scatter(rna_mito, ribo_mito, s=10,c='#ffa500')\n",
    "\n",
    "for index, row in data_fig1b_enzymes.iterrows():\n",
    "    if index in [ 'CPS1', 'CPT1B', 'UCP2', 'ST3GAL3', 'OAZ1', 'SAT1', 'ABCA1', 'PLCD1', 'ADRA2C', 'PLCD1']:\n",
    "        ax.text(row[1] + 0.025, row[0], str(index), horizontalalignment='left', size='medium', color='#016a86', weight='semibold')\n",
    "    else:\n",
    "        pass\n",
    "\n",
    "plt.savefig('./plots/tmisrib_enzymes.pdf', dpi=1800, bbox_inches='tight')\n",
    "\n",
    "plt.show"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_fig1b_enzymes.loc['CHST2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_fig1b_enzymes.loc[data_fig1b_enzymes['ribo_fc'] > 2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_fig1b_enzymes.loc[data_fig1b_enzymes['ribo_fc'] < 0.5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_fig1c_enzymes = data_fig1c_plot.reindex(labels=enzymes)\n",
    "print(data_fig1c_enzymes.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.close()\n",
    "plt.clf()\n",
    "\n",
    "#Prepare plot\n",
    "fig, ax = plt.subplots(1,1, figsize=(7,7))\n",
    "plt.grid(False)\n",
    "\n",
    "ax.axhline(1, ls='-', color='black')\n",
    "ax.axvline(1, ls='-', color='black')\n",
    "\n",
    "ax.axhline(2, ls='--', color='gray')\n",
    "ax.axhline(0.5, ls='--', color='gray')\n",
    "\n",
    "ax.set_ylabel(\"Ribo-Seq (ISRIB/UT)\", fontsize=16)\n",
    "ax.set_xlabel(\"mRNA-Seq (ISRIB/UT)\", fontsize=16)\n",
    "ax.set_xlim(.125,8.5)\n",
    "ax.set_ylim(.125,8.5)\n",
    "ax.set_facecolor(\"#FFFFFF\")\n",
    "\n",
    "x = [.125,.250,.500,1,2,4,8]\n",
    "ticks = [\"1/8\",\"1/4\",\"1/2\",\"1\",\"2\",\"4\",\"8\"]\n",
    "\n",
    "ax.set_xscale('log', basex=2)\n",
    "ax.set_xticks(x)\n",
    "ax.set_xticklabels(ticks, fontsize=12)\n",
    "ax.set_xticklabels([\"\"]*len(xm), minor=True)\n",
    "\n",
    "ax.set_yscale('log', basey=2)\n",
    "ax.set_yticks(x)\n",
    "ax.set_yticklabels(ticks, fontsize=12)\n",
    "ax.set_yticklabels([\"\"]*len(xm), minor=True)\n",
    "\n",
    "#Prep data\n",
    "#All\n",
    "ribo_all = data_fig1c[['ribo_fc']].sum(axis=1).values.tolist()\n",
    "ribo_all = np.array(ribo_all).astype(np.float)\n",
    "ribo_all = np.ndarray.tolist(ribo_all)\n",
    "\n",
    "rna_all = data_fig1c[['rna_fc']].sum(axis=1).values.tolist()\n",
    "rna_all = np.array(rna_all).astype(np.float)\n",
    "rna_all = np.ndarray.tolist(rna_all)\n",
    "\n",
    "#ORFs\n",
    "ribo_mito = data_fig1c_enzymes[['ribo_fc']].sum(axis=1).values.tolist()\n",
    "ribo_mito = np.array(ribo_mito).astype(np.float)\n",
    "ribo_mito = np.ndarray.tolist(ribo_mito)\n",
    "\n",
    "rna_mito = data_fig1c_enzymes[['rna_fc']].sum(axis=1).values.tolist()\n",
    "rna_mito = np.array(rna_mito).astype(np.float)\n",
    "rna_mito = np.ndarray.tolist(rna_mito)\n",
    "\n",
    "#Plot data\n",
    "ax.scatter(rna_all, ribo_all, s=.5,c='gray')\n",
    "ax.scatter(rna_mito, ribo_mito, s=10,c='#ffa500')\n",
    "\n",
    "for index, row in data_fig1c_enzymes.iterrows():\n",
    "    if index in [ 'CPS1', 'CPT1B', 'UCP2', 'ST3GAL3', 'OAZ1', 'SAT1', 'ABCA1', 'PLCD1', 'ADRA2C', 'PLCD1']:\n",
    "        ax.text(row[1] + 0.025, row[0], str(index), horizontalalignment='left', size='medium', color='#016a86', weight='semibold')\n",
    "    else:\n",
    "        pass\n",
    "    \n",
    "plt.savefig('./plots/isrib_enzymes.pdf', dpi=1800, bbox_inches='tight')\n",
    "\n",
    "plt.show"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_fig1c_enzymes.loc[data_fig1c_enzymes['ribo_fc'] > 2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_fig1c_enzymes.loc[data_fig1c_enzymes['ribo_fc'] < 0.5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Tracking analysis of gene groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_logte_avg = data_logte.copy()\n",
    "\n",
    "data_logte_avg['untr'] = (data_logte_avg['untr_a'] + data_logte_avg['untr_b']) / 2\n",
    "data_logte_avg['tm'] = (data_logte_avg['tm_a'] + data_logte_avg['tm_b']) / 2\n",
    "data_logte_avg['tmisrib'] = (data_logte_avg['tmisrib_a'] + data_logte_avg['tmisrib_b']) / 2\n",
    "data_logte_avg['isrib'] = (data_logte_avg['isrib_a'] + data_logte_avg['isrib_b']) / 2\n",
    "\n",
    "data_logte_avg = data_logte_avg[['untr','tm','tmisrib','isrib']]\n",
    "\n",
    "data_logte_avg = data_logte_avg[~data_logte_avg.index.duplicated(keep='first')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_logte_avg_mitocarta = data_logte_avg.reindex(labels=mitocarta)\n",
    "print(data_logte_avg_mitocarta.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_logte_avg_enzymes = data_logte_avg.reindex(labels=enzymes)\n",
    "print(data_logte_avg_enzymes.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_tpm.loc['NPIP']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_tpm_c = data_tpm.copy()\n",
    "\n",
    "data_tpm_c['idx'] = data_tpm_c.index\n",
    "data_tpm_c = data_tpm_c.sort_values(by=['idx'], axis=0)\n",
    "\n",
    "data_tpm_c = data_tpm_c.groupby('idx').mean()\n",
    "\n",
    "data_tpm_c.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(data_tpm_c.loc['NPIP']['ribo_isrib_a_ribo'] + data_tpm_c.loc['NPIP']['ribo_isrib_b_ribo']) / (data_tpm_c.loc['NPIP']['ribo_untr_a_ribo'] + data_tpm_c.loc['NPIP']['ribo_untr_b_ribo'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.log(0.18118091020321184)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_tpm_c.loc['ST3GAL3']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### TM\n",
    "#Footprint\n",
    "statistic, p_value = stats.ttest_ind([1.511040,1.549846], [0.967716,0.530006])\n",
    "print(p_value)\n",
    "\n",
    "#mRNA\n",
    "statistic, p_value = stats.ttest_ind([6.674015,5.682717], [8.467867,6.772637])\n",
    "print(p_value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### TM ISRIB\n",
    "#Footprint\n",
    "statistic, p_value = stats.ttest_ind([1.082089,0.724264], [0.967716,0.530006])\n",
    "print(p_value)\n",
    "\n",
    "#mRNA\n",
    "statistic, p_value = stats.ttest_ind([6.043531,5.682717], [8.467867,6.772637])\n",
    "print(p_value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_tpm_c.loc['ADRA2C']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### TM\n",
    "#Footprint\n",
    "statistic, p_value = stats.ttest_ind([0.790908,0.000000], [0.324174,0.388382])\n",
    "print(p_value)\n",
    "\n",
    "#mRNA\n",
    "statistic, p_value = stats.ttest_ind([2.579677,3.118664], [2.319682,3.015436])\n",
    "print(p_value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### TM ISRIB \n",
    "#Footprint\n",
    "statistic, p_value = stats.ttest_ind([0.792941,1.010918], [0.324174,0.388382])\n",
    "print(p_value)\n",
    "\n",
    "#mRNA\n",
    "statistic, p_value = stats.ttest_ind([3.317482,2.993154], [2.319682,3.015436])\n",
    "print(p_value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_tpm_c.loc['UCP2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### TM\n",
    "#Footprint\n",
    "statistic, p_value = stats.ttest_ind([31.412700,32.384651], [11.761092,9.153570])\n",
    "print(p_value)\n",
    "\n",
    "### TM ISRIB \n",
    "#Footprint\n",
    "statistic, p_value = stats.ttest_ind([16.611926,12.045261], [11.761092,9.153570])\n",
    "print(p_value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_tpm_c.loc['SLMO1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### TM\n",
    "#Footprint\n",
    "statistic, p_value = stats.ttest_ind([10.534073,5.637184], [3.754487,3.052298])\n",
    "print(p_value)\n",
    "\n",
    "### TM ISRIB \n",
    "#Footprint\n",
    "statistic, p_value = stats.ttest_ind([4.591804,2.927037], [3.754487,3.052298])\n",
    "print(p_value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_tpm_c.loc['MTHFD2L']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### TM\n",
    "#Footprint\n",
    "statistic, p_value = stats.ttest_ind([8.545669,11.394674], [9.413403,6.444502])\n",
    "print(p_value)\n",
    "\n",
    "### TM ISRIB \n",
    "#Footprint\n",
    "statistic, p_value = stats.ttest_ind([3.365857,4.505673], [9.413403,6.444502])\n",
    "print(p_value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
